{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "# ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "28 October 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n",
    "\n",
    "## STEPS\n",
    "\n",
    "### imports\n",
    "- pandas\n",
    "- pathlib\n",
    "\n",
    "### Load CDE for validation\n",
    "- check all columns\n",
    "\n",
    "\n",
    "### Team Hardy\n",
    "- load .csv files with tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streamlit NOT successfully\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# local helpers\n",
    "from utils.qcutils import validate_table, force_enum_string, reorder_table_to_CDE\n",
    "from utils.io import ReportCollector, get_dtypes_dict, read_meta_table\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load CDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>ClinPath field</th>\n",
       "      <th>team_Hafler type</th>\n",
       "      <th>ClinPath description</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Project Description:   Brief description of th...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table                Field   \n",
       "0  STUDY         project_name  \\\n",
       "1  STUDY      project_dataset   \n",
       "2  STUDY  project_description   \n",
       "3  STUDY       ASAP_team_name   \n",
       "4  STUDY        ASAP_lab_name   \n",
       "\n",
       "                                         Description DataType  Required   \n",
       "0  Project Name:   A Title of the overall project...   String  Required  \\\n",
       "1  Dataset Name:   A unique name is required for ...   String  Required   \n",
       "2  Project Description:   Brief description of th...   String  Required   \n",
       "3  ASAP Team Name:   Name of the ASAP CRN Team. i...     Enum  Required   \n",
       "4  Lab Name. :   Lab name that is submitting data...   String  Required   \n",
       "\n",
       "                                          Validation  Unnamed: 6   \n",
       "0                                                NaN         NaN  \\\n",
       "1                                                NaN         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "  ClinPath field team_Hafler type ClinPath description Unnamed: 10  \n",
       "0            NaN              NaN                  NaN        bard  \n",
       "1            NaN              NaN                  NaN         NaN  \n",
       "2            NaN              NaN                  NaN         NaN  \n",
       "3            NaN              NaN                  NaN         NaN  \n",
       "4            NaN              NaN                  NaN         NaN  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDE_path = Path.cwd() / \"ASAP_CDE.csv\" \n",
    "CDE = pd.read_csv(CDE_path )\n",
    "\n",
    "CDE.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean Team Hardy tables\n",
    "\n",
    "### Load Tables from csv\n",
    "\n",
    "The metadata path below has copies of the raw meta-tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # AS UPLOADED FROM Team Hardy.  This is the raw meta-data\n",
    "# Samples with proper batch: \"Projects/ASAP/team-hardy/metadata/23102023_SAMPLE.csv\"\n",
    "# All other tables transferred directly to the raw bucket \"Projects/ASAP/team-hardy/hardy-metadata-20232009\"\n",
    "\n",
    "# Initialize the data types dictionary\n",
    "dtypes_dict = get_dtypes_dict(CDE)\n",
    "    \n",
    "\n",
    "\n",
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-hardy\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "\n",
    "# the unmolested Hardy meta-tables don't have index columns \n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\", dtype=dtypes_dict)\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\", dtype=dtypes_dict)\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\", dtype=dtypes_dict)\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\", dtype=dtypes_dict)\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\", dtype=dtypes_dict)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PDD | PD (with dementia)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>borah</td>\n",
       "      <td>P4/11</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bovon</td>\n",
       "      <td>P95/10</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>81</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59.5</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>davof</td>\n",
       "      <td>P80/11</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dudug</td>\n",
       "      <td>P82/10</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84.0</td>\n",
       "      <td>No PD nor other neurological disorder</td>\n",
       "      <td>clinpath info: Control | Control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id biobank_name organism     sex   \n",
       "0      babom             P2/14      QSBB_UK    Human  Female  \\\n",
       "1      borah             P4/11      QSBB_UK    Human    Male   \n",
       "2      bovon            P95/10      QSBB_UK    Human    Male   \n",
       "3      davof            P80/11      QSBB_UK    Human    Male   \n",
       "4      dudug            P82/10      QSBB_UK    Human  Female   \n",
       "\n",
       "   age_at_collection race  ethnicity  duration_pmi   \n",
       "0                 78  NaN        NaN          46.0  \\\n",
       "1                 63  NaN        NaN          37.0   \n",
       "2                 81  NaN        NaN          59.5   \n",
       "3                 80  NaN        NaN         100.0   \n",
       "4                 87  NaN        NaN          84.0   \n",
       "\n",
       "                       primary_diagnosis   \n",
       "0                          Idiopathic PD  \\\n",
       "1                          Idiopathic PD   \n",
       "2                          Idiopathic PD   \n",
       "3                          Idiopathic PD   \n",
       "4  No PD nor other neurological disorder   \n",
       "\n",
       "                    primary_diagnosis_text  \n",
       "0  clinpath info: PDD | PD (with dementia)  \n",
       "1                   clinpath info: PD | PD  \n",
       "2                   clinpath info: PD | NA  \n",
       "3                   clinpath info: PD | PD  \n",
       "4         clinpath info: Control | Control  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT.head()\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\", dtype=dtypes_dict)\n",
    "# CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\", dtype=dtypes_dict)\n",
    "# STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\", dtype=dtypes_dict)\n",
    "# PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\", dtype=dtypes_dict)\n",
    "# SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\", dtype=dtypes_dict)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SUBJECT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SUBJECT['sex'] = SUBJECT['sex'].replace({'F':\"Female\", 'M':\"Male\"})\n",
    "SUBJECT['race'] = SUBJECT['race'].replace({'W':\"White\", 'B':\"Black or African American\"})\n",
    "\n",
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace({'Normal control':\"Healthy Control\", \"Idiopathic Parkinson's disease\":\"Idiopathic PD\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- ethnicity: 64/64 empty rows\n",
      "\n",
      "\t- duration_pmi: 1/64 empty rows\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in SUBJECT. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "subject_report = ReportCollector(destination=\"print\")\n",
    "\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE, subject_report)\n",
    "print(subject_report.get_log())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PDD | PD (with dementia)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>borah</td>\n",
       "      <td>P4/11</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bovon</td>\n",
       "      <td>P95/10</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>81</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59.5</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>davof</td>\n",
       "      <td>P80/11</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dudug</td>\n",
       "      <td>P82/10</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84.0</td>\n",
       "      <td>No PD nor other neurological disorder</td>\n",
       "      <td>clinpath info: Control | Control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id biobank_name organism     sex   \n",
       "0      babom             P2/14      QSBB_UK    Human  Female  \\\n",
       "1      borah             P4/11      QSBB_UK    Human    Male   \n",
       "2      bovon            P95/10      QSBB_UK    Human    Male   \n",
       "3      davof            P80/11      QSBB_UK    Human    Male   \n",
       "4      dudug            P82/10      QSBB_UK    Human  Female   \n",
       "\n",
       "   age_at_collection  race  ethnicity  duration_pmi   \n",
       "0                 78   NaN        NaN          46.0  \\\n",
       "1                 63   NaN        NaN          37.0   \n",
       "2                 81   NaN        NaN          59.5   \n",
       "3                 80   NaN        NaN         100.0   \n",
       "4                 87   NaN        NaN          84.0   \n",
       "\n",
       "                       primary_diagnosis   \n",
       "0                          Idiopathic PD  \\\n",
       "1                          Idiopathic PD   \n",
       "2                          Idiopathic PD   \n",
       "3                          Idiopathic PD   \n",
       "4  No PD nor other neurological disorder   \n",
       "\n",
       "                    primary_diagnosis_text  \n",
       "0  clinpath info: PDD | PD (with dementia)  \n",
       "1                   clinpath info: PD | PD  \n",
       "2                   clinpath info: PD | NA  \n",
       "3                   clinpath info: PD | PD  \n",
       "4         clinpath info: Control | Control  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SAMPLE\n",
    "\n",
    "`source_subject_id` are in the CLINPATH table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\", dtype=dtypes_dict)\n",
    "SAMPLE_OG = SAMPLE.copy()\n",
    "# SAMPLE: source_subject_id -> source_sample_id\n",
    "\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "#SAMPLE[['sample_id','source_sample_id','subject_id', 'batch']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "# fix batch to be BATCH_1, BATCH_2, etc\n",
    "SAMPLE['batch'] = \"BATCH_\" + SAMPLE['batch']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 3616/3616 empty rows\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- pm_PH: 3616/3616 empty rows\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "\t- sequencing_length:190\n",
      "\t>\t change to: 25, 50, 100, 150\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)     \n",
    "\n",
    "sample_report = ReportCollector(destination=\"print\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE, sample_report)\n",
    "print(sample_report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([''], dtype=object)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.source_sample_id.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CLINPATH\n",
    "\n",
    "For some reason there's some duplicate 'sample_id' which also have NULL entries for some of the Enum fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH_og = CLINPATH.copy()\n",
    "\n",
    "CLINPATH.drop_duplicates(subset=['sample_id'], inplace=True)\n",
    "\n",
    "# # CLINPATH.rename(columns={\"subject_id\":\"SUBJECT_ID\"}, inplace=True)\n",
    "# CLINPATH['source_sample_id']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚨⚠️❗ **Missing Required Fields in CLINPATH: source_sample_id**\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 128/128 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 128/128 empty rows\n",
      "\n",
      "\t- path_year_death: 128/128 empty rows\n",
      "\n",
      "\t- brain_weight: 128/128 empty rows\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 128/128 empty rows\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "\t- path_autopsy_dx_main:Control brain, Pathological ageing, Control brain / Path ageing, Argyrophilic grain disease, Control brain, Cerebrovascular disease (small vessel), Cerebrovascular disease (small vessel), Control brain, Alzheimer`s disease (intermediate level AD pathological change), Control brain / Path ageing, CAA\n",
      "\t>\t change to: Lewy body disease nos, Parkinson's disease, Parkinson's disease with dementia, Dementia with Lewy bodies, Multiple system atrophy (SND>OPCA), Multiple system atrophy (OPCA<SND), Multiple system atrophy (SND=OPCA), Progressive supranuclear palsy, Corticobasal degeneration, Globular glial tauoapathy (GGT), Chronic traumatic encephalopathy (CTE), FTLD-Tau (Pick's), FTLD-Tau (MAPT), FTLD-Tau (AGD), FTLD-TDP43, Type A, FTLD-TDP43, Type B, FTLD-TDP43, Type C, FTLD-TDP43, Type D, FTLD-TDP43, Type E, Motor neurone disease-TDP43 (MND or ALS), FTLD-MND-TDP43, Huntington's disease, Spinocerebellar ataxia, nos, Prion disease, nos, Alzheimer's disease (high level neuropathological change), Alzheimer's disease (intermediate level neuropathological change), Control, Low level AD neuropathological change, Control, Limbic predominant age-related TDP43 proteinopathy (LATE), Control, Argyrophilic grain disease, Control, Primary age-related tauopathy (PART), Control, Ageing-related tau astrogliopathy (ARTAG), Control, Cerebrovascular disease (atherosclerosis), Control, Cerebrovascular disease (hyaline arteriolosclerosis), Control, Cerebrovascular disease (cerebral amyloid angiopathy), Control, no misfolded protein or significant vascular pathology, Other neurological disorder\n",
      "\t- path_braak_nft:2, 1, 3, 4, 6\n",
      "\t>\t change to: 0, I, II, III, IV, V, VI, I/II, III/IV, IV/V, V/VI\n",
      "\t- path_thal:At least 4\n",
      "\t>\t change to: 0, 1, 2, 3, 4, 5, 1/2, 3, 4/5\n",
      "\t- path_mckeith:Diffuse neocortical, Limbic transitional, Diffuse Neocortical\n",
      "\t>\t change to: Neocortical, Limbic (transitional), Brainstem, Amygdala Predominant, Olfactory bulb only, Limbic, transitional (brainstem and limbic involvement), Diffuse, neocortical (brainstem, limbic and neocortical involvement), Olfactory Bulb-Only, Limbic (amygdala) predominant, Absent, Present, but extent unknown\n",
      "\t- path_nia_aa_a:1, 2, 0, 3\n",
      "\t>\t change to: A0, A1, A2, A3\n",
      "\t- path_nia_aa_b:1, 2, 0, 3\n",
      "\t>\t change to: B0, B1, B2, B3\n",
      "\t- path_nia_aa_c:0, 1, 2, 3\n",
      "\t>\t change to: C0, C1, C2, C3\n",
      "\t- path_ad_level:No evidence\n",
      "\t>\t change to: No evidence of Alzheimer's disease neuropathological change, Low level Alzheimer's disease neuropathological change, At least low level Alzheimer's disease neuropathological change, Intermediate level Alzheimer's disease neuropathological change, At least intermediate level Alzheimer's disease neuropathological change, High level Alzheimer's disease neuropathological change, Unknown\n",
      "🚨⚠️❗ **Found unexpected NULL (nan):**\n",
      "- family_historyhx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- APOE_e4_status \n",
      "- cognitive_status \n",
      "- path_cerad \n",
      "- known_pathogenic_mutation \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "clinpath_report = ReportCollector(destination=\"print\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE, clinpath_report)\n",
    "\n",
    "print(clinpath_report.get_log())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚨⚠️❗ **Missing Required Fields in CLINPATH: source_sample_id**\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 138/138 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 10/138 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 138/138 empty rows\n",
      "\n",
      "\t- path_year_death: 138/138 empty rows\n",
      "\n",
      "\t- brain_weight: 138/138 empty rows\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 138/138 empty rows\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "\t- path_autopsy_dx_main:Control brain, Pathological ageing, Control brain / Path ageing, Argyrophilic grain disease, Control brain, Cerebrovascular disease (small vessel), Cerebrovascular disease (small vessel), Control brain, Alzheimer`s disease (intermediate level AD pathological change), Control brain / Path ageing, CAA\n",
      "\t>\t change to: Lewy body disease nos, Parkinson's disease, Parkinson's disease with dementia, Dementia with Lewy bodies, Multiple system atrophy (SND>OPCA), Multiple system atrophy (OPCA<SND), Multiple system atrophy (SND=OPCA), Progressive supranuclear palsy, Corticobasal degeneration, Globular glial tauoapathy (GGT), Chronic traumatic encephalopathy (CTE), FTLD-Tau (Pick's), FTLD-Tau (MAPT), FTLD-Tau (AGD), FTLD-TDP43, Type A, FTLD-TDP43, Type B, FTLD-TDP43, Type C, FTLD-TDP43, Type D, FTLD-TDP43, Type E, Motor neurone disease-TDP43 (MND or ALS), FTLD-MND-TDP43, Huntington's disease, Spinocerebellar ataxia, nos, Prion disease, nos, Alzheimer's disease (high level neuropathological change), Alzheimer's disease (intermediate level neuropathological change), Control, Low level AD neuropathological change, Control, Limbic predominant age-related TDP43 proteinopathy (LATE), Control, Argyrophilic grain disease, Control, Primary age-related tauopathy (PART), Control, Ageing-related tau astrogliopathy (ARTAG), Control, Cerebrovascular disease (atherosclerosis), Control, Cerebrovascular disease (hyaline arteriolosclerosis), Control, Cerebrovascular disease (cerebral amyloid angiopathy), Control, no misfolded protein or significant vascular pathology, Other neurological disorder\n",
      "\t- path_braak_nft:2, 1, 3, 4, 6\n",
      "\t>\t change to: 0, I, II, III, IV, V, VI, I/II, III/IV, IV/V, V/VI\n",
      "\t- path_thal:At least 4\n",
      "\t>\t change to: 0, 1, 2, 3, 4, 5, 1/2, 3, 4/5\n",
      "\t- path_mckeith:Diffuse neocortical, Limbic transitional, Diffuse Neocortical\n",
      "\t>\t change to: Neocortical, Limbic (transitional), Brainstem, Amygdala Predominant, Olfactory bulb only, Limbic, transitional (brainstem and limbic involvement), Diffuse, neocortical (brainstem, limbic and neocortical involvement), Olfactory Bulb-Only, Limbic (amygdala) predominant, Absent, Present, but extent unknown\n",
      "\t- path_nia_aa_a:1, 2, 0, 3\n",
      "\t>\t change to: A0, A1, A2, A3\n",
      "\t- path_nia_aa_b:1, 2, 0, 3\n",
      "\t>\t change to: B0, B1, B2, B3\n",
      "\t- path_nia_aa_c:0, 1, 2, 3\n",
      "\t>\t change to: C0, C1, C2, C3\n",
      "\t- path_ad_level:No evidence\n",
      "\t>\t change to: No evidence of Alzheimer's disease neuropathological change, Low level Alzheimer's disease neuropathological change, At least low level Alzheimer's disease neuropathological change, Intermediate level Alzheimer's disease neuropathological change, At least intermediate level Alzheimer's disease neuropathological change, High level Alzheimer's disease neuropathological change, Unknown\n",
      "🚨⚠️❗ **Found unexpected NULL (nan):**\n",
      "- family_historyhx_dementia_mci \n",
      "- hx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- APOE_e4_status \n",
      "- cognitive_status \n",
      "- path_autopsy_dx_main \n",
      "- path_braak_nft \n",
      "- path_braak_asyn \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- known_pathogenic_mutation \n",
      "- path_mckeith \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "clinpath_report2 = ReportCollector(destination=\"print\")\n",
    "validate_table(CLINPATH_og, \"CLINPATH\", CDE, clinpath_report2)\n",
    "\n",
    "print(clinpath_report2.get_log())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['6', '0', '5'], dtype=object)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replace 'path_braak_asyn' with with string of the numeric.\n",
    "# ???:  convert nan to \"\"?? else \"\"\n",
    "# CLINPATH['path_braak_asyn'] = CLINPATH['path_braak_asyn'].apply(lambda val: str(int(float(val))))\n",
    "CLINPATH['path_braak_asyn'].apply(lambda val: str(int(float(val)))).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['2', '1', '3', '0', '4', '6'], dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_braak_nft'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_braak_nft'] = CLINPATH['path_braak_nft'].replace({\"0\":\"0\", \n",
    "                                                                \"1\":\"I\", \n",
    "                                                                \"2\": \"II\", \n",
    "                                                                \"3\":\"III\", \n",
    "                                                                \"4\":\"IV\", \n",
    "                                                                \"5\":\"V\", \n",
    "                                                                \"6\":\"VI\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# code family_history as \"Not Reported\" (currently empty)\n",
    "CLINPATH['family_history'] = \"Not Reported\"\n",
    "\n",
    "# check APOE_e4_status ? currently empty\n",
    "# `path_autopsy_dx_main`  actually seems good parser might be wrong\n",
    "\n",
    "# code \"at least 4\" as \"4/5\" \n",
    "CLINPATH['path_thal'] = CLINPATH['path_thal'].replace({'At least 4':\"4/5\"})\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace({'Diffuse neocortical': \"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\", \n",
    "                                                        'Limbic transitional': \"Limbic (transitional)\" ,\n",
    "                                                        'Diffuse Neocortical':\"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\"})\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_a'] = CLINPATH['path_nia_aa_a'].replace({\"0\":\"A0\", \n",
    "                                                                    \"1\":\"A1\", \n",
    "                                                                    \"2\": \"A2\", \n",
    "                                                                    \"3\":\"A3\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '0', '3'], dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_nia_aa_b'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_b'] = CLINPATH['path_nia_aa_b'].replace({\"0\":\"B0\", \n",
    "                                                                \"1\":\"B1\", \n",
    "                                                                \"2\": \"B2\", \n",
    "                                                                \"3\":\"B3\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_c'] = CLINPATH['path_nia_aa_c'].replace({\"0\":\"C0\", \n",
    "                                                                \"1\":\"C1\", \n",
    "                                                                \"2\": \"C2\", \n",
    "                                                                \"3\":\"C3\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['path_ad_level'] = CLINPATH['path_ad_level'].replace({\"No evidence\": \"No evidence of Alzheimer\\'s disease neuropathological change\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['source_sample_id'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32m/Users/ergonyc/Projects/ASAP/meta-clean/hardy_clean.ipynb Cell 31\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/ergonyc/Projects/ASAP/meta-clean/hardy_clean.ipynb#X44sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m CLINPATH[[\u001b[39m'\u001b[39;49m\u001b[39msource_sample_id\u001b[39;49m\u001b[39m'\u001b[39;49m,\u001b[39m'\u001b[39;49m\u001b[39mpath_autopsy_dx_main\u001b[39;49m\u001b[39m'\u001b[39;49m,\u001b[39m'\u001b[39;49m\u001b[39mpath_ad_level\u001b[39;49m\u001b[39m'\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mcause_death\u001b[39;49m\u001b[39m\"\u001b[39;49m]]\u001b[39m.\u001b[39mdrop_duplicates()\u001b[39m.\u001b[39mhead()\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/frame.py:3767\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3765\u001b[0m     \u001b[39mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3766\u001b[0m         key \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39m(key)\n\u001b[0;32m-> 3767\u001b[0m     indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49m_get_indexer_strict(key, \u001b[39m\"\u001b[39;49m\u001b[39mcolumns\u001b[39;49m\u001b[39m\"\u001b[39;49m)[\u001b[39m1\u001b[39m]\n\u001b[1;32m   3769\u001b[0m \u001b[39m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3770\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mgetattr\u001b[39m(indexer, \u001b[39m\"\u001b[39m\u001b[39mdtype\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m) \u001b[39m==\u001b[39m \u001b[39mbool\u001b[39m:\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/indexes/base.py:5876\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   5873\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   5874\u001b[0m     keyarr, indexer, new_indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 5876\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_raise_if_missing(keyarr, indexer, axis_name)\n\u001b[1;32m   5878\u001b[0m keyarr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtake(indexer)\n\u001b[1;32m   5879\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(key, Index):\n\u001b[1;32m   5880\u001b[0m     \u001b[39m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/miniforge3/envs/scverse10/lib/python3.10/site-packages/pandas/core/indexes/base.py:5938\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   5935\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mNone of [\u001b[39m\u001b[39m{\u001b[39;00mkey\u001b[39m}\u001b[39;00m\u001b[39m] are in the [\u001b[39m\u001b[39m{\u001b[39;00maxis_name\u001b[39m}\u001b[39;00m\u001b[39m]\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m   5937\u001b[0m not_found \u001b[39m=\u001b[39m \u001b[39mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[39m.\u001b[39mnonzero()[\u001b[39m0\u001b[39m]]\u001b[39m.\u001b[39munique())\n\u001b[0;32m-> 5938\u001b[0m \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mnot_found\u001b[39m}\u001b[39;00m\u001b[39m not in index\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['source_sample_id'] not in index\""
     ]
    }
   ],
   "source": [
    "CLINPATH[['source_sample_id','path_autopsy_dx_main','path_ad_level', \"cause_death\"]].drop_duplicates().head(50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"Parkinson's disease with dementia\", \"Parkinson's disease\",\n",
       "       'Control brain', 'Pathological ageing',\n",
       "       'Control brain / Path ageing', 'Argyrophilic grain disease',\n",
       "       'Control brain, Cerebrovascular disease (small vessel)',\n",
       "       'Cerebrovascular disease (small vessel)',\n",
       "       'Control brain, Alzheimer`s disease (intermediate level AD pathological change)',\n",
       "       'Control brain / Path ageing, CAA'], dtype=object)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_autopsy_dx_main'].unique()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_autopsy_map = { \"Parkinson's disease with dementia\": \"Parkinson's disease with dementia\", \n",
    "       \"Parkinson's disease\": \"Parkinson's disease\",\n",
    "       'Control brain':\"Control, no misfolded protein or significant vascular pathology\", \n",
    "       'Pathological ageing': 'Control, no misfolded protein or significant vascular pathology',\n",
    "       'Control brain / Path ageing': 'Control, no misfolded protein or significant vascular pathology',\n",
    "       'Argyrophilic grain disease': \"Control, Argyrophilic grain disease\",\n",
    "       'Control brain, Cerebrovascular disease (small vessel)':\"Control, Cerebrovascular disease (atherosclerosis)\",\n",
    "       'Cerebrovascular disease (small vessel)':\"Control, Cerebrovascular disease (atherosclerosis)\",\n",
    "       \"Control brain, Alzheimer`s disease (intermediate level AD pathological change)\":\"Alzheimer's disease (intermediate level neuropathological change)\",\n",
    "       'Control brain / Path ageing, CAA':\"Control, Cerebrovascular disease (cerebral amyloid angiopathy)\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['path_autopsy_dx_main'] = CLINPATH['path_autopsy_dx_main'].replace(path_autopsy_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"Parkinson's disease with dementia\", \"Parkinson's disease\",\n",
       "       'Control, no misfolded protein or significant vascular pathology',\n",
       "       'Control, Argyrophilic grain disease',\n",
       "       'Control, Cerebrovascular disease (atherosclerosis)',\n",
       "       \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       "       'Control, Cerebrovascular disease (cerebral amyloid angiopathy)'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_autopsy_dx_main'].unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚨⚠️❗ **Missing Required Fields in CLINPATH: source_sample_id**\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 128/128 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 128/128 empty rows\n",
      "\n",
      "\t- path_year_death: 128/128 empty rows\n",
      "\n",
      "\t- brain_weight: 128/128 empty rows\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 128/128 empty rows\n",
      "## Enum fields have valid values in CLINPATH. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "clinpath_report = ReportCollector(destination=\"print\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE, clinpath_report)\n",
    "\n",
    "print(clinpath_report.get_log())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0]).reset_index(drop=True)\n",
    "# # STUDY[[\"Unnamed: 1\"]].transpose().reset_index().drop(columns=[\"index\"]), tmp\n",
    "# STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚨⚠️❗ **Missing Required Fields in STUDY: submittor_email**\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in STUDY. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "study_report = ReportCollector(destination=\"print\")\n",
    "validate_table(STUDY, \"STUDY\", CDE, study_report)\n",
    "print(study_report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submitter_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  project_dataset   \n",
       "0  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq  \\\n",
       "\n",
       "0                                project_description ASAP_team_name   \n",
       "0  Genetic analysis has identified many risk gene...     TEAM-HARDY  \\\n",
       "\n",
       "0 ASAP_lab_name PI_full_name              PI_email   \n",
       "0     Ryten Lab   Mina Ryten  mina.ryten@ucl.ac.uk  \\\n",
       "\n",
       "0                                  contributor_names           submitter_name   \n",
       "0  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne  \\\n",
       "\n",
       "0                       submitter_email  ... other_funding_source   \n",
       "0  aine.fairbrother-browne.18@ucl.ac.uk  ...                  NaN  \\\n",
       "\n",
       "0 publication_DOI publication_PMID number_of_brain_samples   \n",
       "0             NaN              NaN                     128  \\\n",
       "\n",
       "0                                      brain_regions   \n",
       "0  Inferior Parietal Lobule (IPL), Anterior Cingu...  \\\n",
       "\n",
       "0                                   types_of_samples            PI_ORCHID   \n",
       "0  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957  \\\n",
       "\n",
       "0                               PI_google_scholar_id DUA_version   \n",
       "0  https://scholar.google.co.uk/citations?user=lt...         NaN  \\\n",
       "\n",
       "0 metadata_version_date  \n",
       "0    Version 1, 09/2023  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [project_name, project_dataset, project_description, ASAP_team_name, ASAP_lab_name, PI_full_name, PI_email, contributor_names, submitter_name, submittor_email, ASAP_grant_id, other_funding_source, publication_DOI, publication_PMID, number_of_brain_samples, brain_regions, types_of_samples, PI_ORCHID, PI_google_scholar_id, DUA_version, metadata_version_date]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 21 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### PROTOCOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_collection_summary</th>\n",
       "      <th>cell_extraction_summary</th>\n",
       "      <th>lib_prep_summary</th>\n",
       "      <th>data_processing_summary</th>\n",
       "      <th>github_url</th>\n",
       "      <th>protocols_io_DOI</th>\n",
       "      <th>other_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>This dataset contains cortical regions only, p...</td>\n",
       "      <td>From protocols.io: This protocol is used to is...</td>\n",
       "      <td>'Nuclei were extracted from homogenised post-m...</td>\n",
       "      <td>Cell ranger was used to convert raw sequencing...</td>\n",
       "      <td>Raw to fastq to mapped: https://github.com/RHR...</td>\n",
       "      <td>Nuclear extraction protocol: 10.17504/protocol...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0                          sample_collection_summary   \n",
       "0  This dataset contains cortical regions only, p...  \\\n",
       "\n",
       "0                            cell_extraction_summary   \n",
       "0  From protocols.io: This protocol is used to is...  \\\n",
       "\n",
       "0                                   lib_prep_summary   \n",
       "0  'Nuclei were extracted from homogenised post-m...  \\\n",
       "\n",
       "0                            data_processing_summary   \n",
       "0  Cell ranger was used to convert raw sequencing...  \\\n",
       "\n",
       "0                                         github_url   \n",
       "0  Raw to fastq to mapped: https://github.com/RHR...  \\\n",
       "\n",
       "0                                   protocols_io_DOI other_reference  \n",
       "0  Nuclear extraction protocol: 10.17504/protocol...             NaN  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = PROTOCOL[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "PROTOCOL = tmp.drop([0]).reset_index(drop=True)\n",
    "PROTOCOL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "protocol_report = ReportCollector(destination=\"print\")\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE, protocol_report)\n",
    "print(protocol_report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### export clean tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = data_path / \"team-hardy\"\n",
    "data_path = Path.home() / (\"Projects/ASAP/team-hardy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'hardy'"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path.name.split(\"-\")[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # write the clean metadata\n",
    "# STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "# PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "# CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "# SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "# SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Hardy\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure cleaned files are correct\n",
    "\n",
    "SUBJECT = read_meta_table(f\"{export_root}/SUBJECT.csv\", dtypes_dict)\n",
    "CLINPATH = read_meta_table(f\"{export_root}/CLINPATH.csv\", dtypes_dict)\n",
    "STUDY = read_meta_table(f\"{export_root}/STUDY.csv\", dtypes_dict)\n",
    "PROTOCOL = read_meta_table(f\"{export_root}/PROTOCOL.csv\", dtypes_dict)\n",
    "SAMPLE = read_meta_table(f\"{export_root}/SAMPLE.csv\", dtypes_dict)\n",
    "\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{export_root}/SUBJECT.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# CLINPATH = pd.read_csv(f\"{export_root}/CLINPATH.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# STUDY = pd.read_csv(f\"{export_root}/STUDY.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# PROTOCOL = pd.read_csv(f\"{export_root}/PROTOCOL.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# SAMPLE = pd.read_csv(f\"{export_root}/SAMPLE.csv\",header=0,index_col=0, dtype=dtypes_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Unnamed: 0'"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJ = pd.read_csv(f\"{export_root}/SUBJECT.csv\", dtype=dtypes_dict)\n",
    "SUBJ.columns[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>46.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PDD | PD (with dementia)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>borah</td>\n",
       "      <td>P4/11</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>63</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bovon</td>\n",
       "      <td>P95/10</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>81</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59.5</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>davof</td>\n",
       "      <td>P80/11</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>80</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>clinpath info: PD | PD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dudug</td>\n",
       "      <td>P82/10</td>\n",
       "      <td>QSBB_UK</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>87</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84.0</td>\n",
       "      <td>No PD nor other neurological disorder</td>\n",
       "      <td>clinpath info: Control | Control</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id biobank_name organism     sex   \n",
       "0      babom             P2/14      QSBB_UK    Human  Female  \\\n",
       "1      borah             P4/11      QSBB_UK    Human    Male   \n",
       "2      bovon            P95/10      QSBB_UK    Human    Male   \n",
       "3      davof            P80/11      QSBB_UK    Human    Male   \n",
       "4      dudug            P82/10      QSBB_UK    Human  Female   \n",
       "\n",
       "   age_at_collection race  ethnicity  duration_pmi   \n",
       "0                 78  NaN        NaN          46.0  \\\n",
       "1                 63  NaN        NaN          37.0   \n",
       "2                 81  NaN        NaN          59.5   \n",
       "3                 80  NaN        NaN         100.0   \n",
       "4                 87  NaN        NaN          84.0   \n",
       "\n",
       "                       primary_diagnosis   \n",
       "0                          Idiopathic PD  \\\n",
       "1                          Idiopathic PD   \n",
       "2                          Idiopathic PD   \n",
       "3                          Idiopathic PD   \n",
       "4  No PD nor other neurological disorder   \n",
       "\n",
       "                    primary_diagnosis_text  \n",
       "0  clinpath info: PDD | PD (with dementia)  \n",
       "1                   clinpath info: PD | PD  \n",
       "2                   clinpath info: PD | NA  \n",
       "3                   clinpath info: PD | PD  \n",
       "4         clinpath info: Control | Control  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- ethnicity: 64/64 empty rows\n",
      "\n",
      "\t- duration_pmi: 1/64 empty rows\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in SUBJECT. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table, table_name = SUBJECT, \"SUBJECT\"\n",
    "\n",
    "report = ReportCollector(destination=\"print\")\n",
    "validate_table(table, table_name, CDE, report)\n",
    "print(report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 3616/3616 empty rows\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- DV200: 3616/3616 empty rows\n",
      "\n",
      "\t- pm_PH: 3616/3616 empty rows\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "\t- sequencing_length:190\n",
      "\t>\t change to: 25, 50, 100, 150\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table, table_name = SAMPLE, \"SAMPLE\"\n",
    "\n",
    "report = ReportCollector(destination=\"print\")\n",
    "validate_table(table, table_name, CDE, report)\n",
    "print(report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "🚨⚠️❗ **Missing Required Fields in CLINPATH: source_sample_id**\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 128/128 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 128/128 empty rows\n",
      "\n",
      "\t- path_year_death: 128/128 empty rows\n",
      "\n",
      "\t- brain_weight: 128/128 empty rows\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 128/128 empty rows\n",
      "## Enum fields have valid values in CLINPATH. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table, table_name = CLINPATH, \"CLINPATH\"\n",
    "\n",
    "report = ReportCollector(destination=\"print\")\n",
    "validate_table(table, table_name, CDE, report)\n",
    "print(report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['6', '0', '5'], dtype=object)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_braak_asyn'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "      <th>dig_slide_avail</th>\n",
       "      <th>quant_path_avail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>libat_IPL</td>\n",
       "      <td>P27/11</td>\n",
       "      <td>0</td>\n",
       "      <td>MDGAP-QSBB_000088_s1</td>\n",
       "      <td>Left</td>\n",
       "      <td>Parietal lobe</td>\n",
       "      <td>Inferior parietal lobule</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low level Alzheimer's disease neuropathologica...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>libat_ACG</td>\n",
       "      <td>P27/11</td>\n",
       "      <td>0</td>\n",
       "      <td>MDGAP-QSBB_000088_s1</td>\n",
       "      <td>Left</td>\n",
       "      <td>Cingulate gyrus</td>\n",
       "      <td>Anterior cingulate gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low level Alzheimer's disease neuropathologica...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rijof_IPL</td>\n",
       "      <td>P8/18</td>\n",
       "      <td>0</td>\n",
       "      <td>MDGAP-QSBB_000583_s1</td>\n",
       "      <td>Right</td>\n",
       "      <td>Parietal lobe</td>\n",
       "      <td>Inferior parietal lobule</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low level Alzheimer's disease neuropathologica...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rijof_ACG</td>\n",
       "      <td>P8/18</td>\n",
       "      <td>0</td>\n",
       "      <td>MDGAP-QSBB_000583_s1</td>\n",
       "      <td>Right</td>\n",
       "      <td>Cingulate gyrus</td>\n",
       "      <td>Anterior cingulate gyrus</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low level Alzheimer's disease neuropathologica...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gotar_IPL</td>\n",
       "      <td>P41/09</td>\n",
       "      <td>0</td>\n",
       "      <td>MDGAP-QSBB_000406_s1</td>\n",
       "      <td>Right</td>\n",
       "      <td>Parietal lobe</td>\n",
       "      <td>Inferior parietal lobule</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A1</td>\n",
       "      <td>B1</td>\n",
       "      <td>C0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low level Alzheimer's disease neuropathologica...</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_subject_id time_from_baseline                GP2_id   \n",
       "0  libat_IPL            P27/11                  0  MDGAP-QSBB_000088_s1  \\\n",
       "1  libat_ACG            P27/11                  0  MDGAP-QSBB_000088_s1   \n",
       "2  rijof_IPL             P8/18                  0  MDGAP-QSBB_000583_s1   \n",
       "3  rijof_ACG             P8/18                  0  MDGAP-QSBB_000583_s1   \n",
       "4  gotar_IPL            P41/09                  0  MDGAP-QSBB_000406_s1   \n",
       "\n",
       "  hemisphere   region_level_1            region_level_2 region_level_3   \n",
       "0       Left    Parietal lobe  Inferior parietal lobule    Grey matter  \\\n",
       "1       Left  Cingulate gyrus  Anterior cingulate gyrus    Grey matter   \n",
       "2      Right    Parietal lobe  Inferior parietal lobule    Grey matter   \n",
       "3      Right  Cingulate gyrus  Anterior cingulate gyrus    Grey matter   \n",
       "4      Right    Parietal lobe  Inferior parietal lobule    Grey matter   \n",
       "\n",
       "  AMPPD_id family_history  ... path_nia_ri  path_nia_aa_a path_nia_aa_b   \n",
       "0      NaN   Not Reported  ...         NaN             A1            B1  \\\n",
       "1      NaN   Not Reported  ...         NaN             A1            B1   \n",
       "2      NaN   Not Reported  ...         NaN             A1            B1   \n",
       "3      NaN   Not Reported  ...         NaN             A1            B1   \n",
       "4      NaN   Not Reported  ...         NaN             A1            B1   \n",
       "\n",
       "   path_nia_aa_c TDP43 arteriolosclerosis_severity_scale   \n",
       "0             C0   NaN                               NaN  \\\n",
       "1             C0   NaN                               NaN   \n",
       "2             C0   NaN                               NaN   \n",
       "3             C0   NaN                               NaN   \n",
       "4             C0   NaN                               NaN   \n",
       "\n",
       "  amyloid_angiopathy_severity_scale   \n",
       "0                               NaN  \\\n",
       "1                               NaN   \n",
       "2                               NaN   \n",
       "3                               NaN   \n",
       "4                               NaN   \n",
       "\n",
       "                                       path_ad_level  dig_slide_avail   \n",
       "0  Low level Alzheimer's disease neuropathologica...              Yes  \\\n",
       "1  Low level Alzheimer's disease neuropathologica...              Yes   \n",
       "2  Low level Alzheimer's disease neuropathologica...              Yes   \n",
       "3  Low level Alzheimer's disease neuropathologica...              Yes   \n",
       "4  Low level Alzheimer's disease neuropathologica...              Yes   \n",
       "\n",
       "  quant_path_avail  \n",
       "0              Yes  \n",
       "1              Yes  \n",
       "2              Yes  \n",
       "3              Yes  \n",
       "4              Yes  \n",
       "\n",
       "[5 rows x 54 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_2</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0  babom_ACG              NaN      babom       NaN                1  \\\n",
       "1  babom_ACG              NaN      babom       NaN                1   \n",
       "2  babom_ACG              NaN      babom       NaN                1   \n",
       "3  babom_ACG              NaN      babom       NaN                1   \n",
       "4  babom_ACG              NaN      babom       NaN                1   \n",
       "\n",
       "   repeated_sample    batch tissue brain_region  source_RIN  ...   \n",
       "0                0  BATCH_2  brain          ACG         NaN  ...  \\\n",
       "1                0  BATCH_2  brain          ACG         NaN  ...   \n",
       "2                0  BATCH_2  brain          ACG         NaN  ...   \n",
       "3                0  BATCH_2  brain          ACG         NaN  ...   \n",
       "4                0  BATCH_2  brain          ACG         NaN  ...   \n",
       "\n",
       "    sex_ontology_term_id self_reported_ethnicity_ontology_term_id   \n",
       "0  PATO:0000383 (female)                                      NaN  \\\n",
       "1  PATO:0000383 (female)                                      NaN   \n",
       "2  PATO:0000383 (female)                                      NaN   \n",
       "3  PATO:0000383 (female)                                      NaN   \n",
       "4  PATO:0000383 (female)                                      NaN   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id   \n",
       "0             MONDO:0005180          UBERON:0009835  \\\n",
       "1             MONDO:0005180          UBERON:0009835   \n",
       "2             MONDO:0005180          UBERON:0009835   \n",
       "3             MONDO:0005180          UBERON:0009835   \n",
       "4             MONDO:0005180          UBERON:0009835   \n",
       "\n",
       "  cell_type_ontology_term_id assay_ontology_term_id suspension_type DV200   \n",
       "0                        NaN            EFO:0008913         nucleus   NaN  \\\n",
       "1                        NaN            EFO:0008913         nucleus   NaN   \n",
       "2                        NaN            EFO:0008913         nucleus   NaN   \n",
       "3                        NaN            EFO:0008913         nucleus   NaN   \n",
       "4                        NaN            EFO:0008913         nucleus   NaN   \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN      NaN  \n",
       "1   NaN      NaN  \n",
       "2   NaN      NaN  \n",
       "3   NaN      NaN  \n",
       "4   NaN      NaN  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scverse10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
