{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "# ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "28 October 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n",
    "\n",
    "## STEPS\n",
    "\n",
    "### imports\n",
    "- pandas\n",
    "- pathlib\n",
    "\n",
    "### Load CDE for validation\n",
    "- check all columns\n",
    "\n",
    "### Team Lee\n",
    "- load .tsv, csv tables\n",
    "- fix format\n",
    "- load additional metadata\n",
    "\n",
    "- add batch columns\n",
    "- add missing columns\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "# local helpers\n",
    "from utils.qcutils import validate_table, force_enum_string, reorder_table_to_CDE\n",
    "from utils.io import ReportCollector, get_dtypes_dict, read_meta_table\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load CDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>ClinPath field</th>\n",
       "      <th>team_Hafler type</th>\n",
       "      <th>ClinPath description</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Project Description:   Brief description of th...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table                Field   \n",
       "0  STUDY         project_name  \\\n",
       "1  STUDY      project_dataset   \n",
       "2  STUDY  project_description   \n",
       "3  STUDY       ASAP_team_name   \n",
       "4  STUDY        ASAP_lab_name   \n",
       "\n",
       "                                         Description DataType  Required   \n",
       "0  Project Name:   A Title of the overall project...   String  Required  \\\n",
       "1  Dataset Name:   A unique name is required for ...   String  Required   \n",
       "2  Project Description:   Brief description of th...   String  Required   \n",
       "3  ASAP Team Name:   Name of the ASAP CRN Team. i...     Enum  Required   \n",
       "4  Lab Name. :   Lab name that is submitting data...   String  Required   \n",
       "\n",
       "                                          Validation  Unnamed: 6   \n",
       "0                                                NaN         NaN  \\\n",
       "1                                                NaN         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "  ClinPath field team_Hafler type ClinPath description Unnamed: 10  \n",
       "0            NaN              NaN                  NaN        bard  \n",
       "1            NaN              NaN                  NaN         NaN  \n",
       "2            NaN              NaN                  NaN         NaN  \n",
       "3            NaN              NaN                  NaN         NaN  \n",
       "4            NaN              NaN                  NaN         NaN  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDE_path = Path.cwd() / \"ASAP_CDE.csv\" \n",
    "CDE = pd.read_csv(CDE_path )\n",
    "\n",
    "CDE.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "## Clean Team Lee tables\n",
    "\n",
    "### Load Tables from csv\n",
    "\n",
    "The metadata path below has copies of the raw meta-tables\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Initialize the data types dictionary\n",
    "dtypes_dict = get_dtypes_dict(CDE)\n",
    "    \n",
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-lee\")\n",
    "# NOTE:  ogmetadata is the original metadata folder\n",
    "metadata_path = data_path / \"metadata/ogmetadata\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\", delimiter=\"\\t\", dtype=dtypes_dict)\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\", dtype=dtypes_dict)\n",
    "\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\",delimiter=\",\", dtype=dtypes_dict)\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.tsv\",delimiter=\"\\t\", dtype=dtypes_dict)\n",
    "\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.tsv\",delimiter=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Team-Lee-Bras-Lab-Info</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>Data type</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Note</th>\n",
       "      <th>Required/Optional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name/Title</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unique and clear title.</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset name</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A Dataset name is required for each submission...</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Brief description of the goals and objectives ...</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team e.g. \"Scherzer\"</td>\n",
       "      <td>Enum</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bras</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>ASAP Lab under the above team e.g. \"Dong\"</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Unnamed: 0           Unnamed: 1   \n",
       "0  Is senescence a component of human PD and does...         project_name  \\\n",
       "1    Human snRNA-seq PD Senesence Jose Bras Team Lee      project_dataset   \n",
       "2  Characterize the neuropathological progression...  project_description   \n",
       "3                                           TEAM-LEE       ASAP_team_name   \n",
       "4                                               Bras        ASAP_lab_name   \n",
       "\n",
       "                                          Unnamed: 2 Team-Lee-Bras-Lab-Info   \n",
       "0                                 Project Name/Title                 String  \\\n",
       "1                                       Dataset name                 String   \n",
       "2  Brief description of the goals and objectives ...                 String   \n",
       "3                          ASAP Team e.g. \"Scherzer\"                   Enum   \n",
       "4          ASAP Lab under the above team e.g. \"Dong\"                 String   \n",
       "\n",
       "                                          Field   \n",
       "0                                           NaN  \\\n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]   \n",
       "4                                           NaN   \n",
       "\n",
       "                                         Description Data type  Validation   \n",
       "0                            Unique and clear title.  Required         NaN  \\\n",
       "1  A Dataset name is required for each submission...  Required         NaN   \n",
       "2                                                NaN  Required         NaN   \n",
       "3                                                NaN  Required         NaN   \n",
       "4                                                NaN  Required         NaN   \n",
       "\n",
       "   Note  Required/Optional  \n",
       "0   NaN                NaN  \n",
       "1   NaN                NaN  \n",
       "2   NaN                NaN  \n",
       "3   NaN                NaN  \n",
       "4   NaN                NaN  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# STUDY = pd.read_csv(metadata_path / \"STUDY.tsv\",delimiter=\"\\t\")\n",
    "STUDY.to_csv(data_path / \"STUDY_.csv\")\n",
    "STUDY = pd.read_csv(data_path / \"STUDY_.csv\")\n",
    "\n",
    "\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name   \n",
       "1  Is senescence a component of human PD and does...  \\\n",
       "\n",
       "0                                  project_dataset   \n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee  \\\n",
       "\n",
       "0                                project_description ASAP_team_name   \n",
       "1  Characterize the neuropathological progression...       TEAM-LEE  \\\n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email   \n",
       "1          Bras   Jose, Bras  jose.bras@vai.org  \\\n",
       "\n",
       "0                                       submitter_id    submitter_name   \n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra  \\\n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI   \n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN  \\\n",
       "\n",
       "0 publication_PMID number_of_brain_samples   \n",
       "1              NaN                      75  \\\n",
       "\n",
       "0                                      brain_regions   \n",
       "1  hippocampus; middle frontal gyrus; substantia ...  \\\n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id   \n",
       "1  human PD and control postmortem brains       NaN                  NaN  \\\n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"Unnamed: 1\",\"Unnamed: 0\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "# STUDY[[\"Unnamed: 1\"]].transpose().reset_index().drop(columns=[\"index\"]), tmp\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Need to rename submitter_id to contributor_names\n",
    "STUDY = STUDY.rename(columns={\"submitter_id\": \"contributor_names\"})\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *STUDY* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in STUDY. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "study_report = ReportCollector(destination=\"print\")\n",
    "validate_table(STUDY, \"STUDY\", CDE, study_report)\n",
    "print(study_report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        project_name   \n",
       "1  Is senescence a component of human PD and does...  \\\n",
       "\n",
       "                                   project_dataset   \n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee  \\\n",
       "\n",
       "                                 project_description ASAP_team_name   \n",
       "1  Characterize the neuropathological progression...       TEAM-LEE  \\\n",
       "\n",
       "  ASAP_lab_name PI_full_name           PI_email   \n",
       "1          Bras   Jose, Bras  jose.bras@vai.org  \\\n",
       "\n",
       "                                   contributor_names    submitter_name   \n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra  \\\n",
       "\n",
       "          submittor_email  ... other_funding_source publication_DOI   \n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN  \\\n",
       "\n",
       "  publication_PMID number_of_brain_samples   \n",
       "1              NaN                      75  \\\n",
       "\n",
       "                                       brain_regions   \n",
       "1  hippocampus; middle frontal gyrus; substantia ...  \\\n",
       "\n",
       "                         types_of_samples PI_ORCHID PI_google_scholar_id   \n",
       "1  human PD and control postmortem brains       NaN                  NaN  \\\n",
       "\n",
       "  DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SAMPLE\n",
    "\n",
    "`batch` must be collected additional metadata from covar.csv .. i.e. batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\", dtype=dtypes_dict)\n",
    "\n",
    "metadata_path = Path.home() / (\"Projects/ASAP/team-lee/metadata\")\n",
    "HIP_covar = pd.read_csv(f\"{metadata_path}/HIP/covar.csv\")\n",
    "HIP_cases = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "HIP_control = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{metadata_path}/MFG/covar.csv\") # includes 'PMI' ?\n",
    "MFG_cases = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "MFG_control = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "\n",
    "SN_covar = pd.read_csv(f\"{metadata_path}/SN/covar.csv\")\n",
    "SN_cases = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "SN_control = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_controls.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hippocampus samples\n",
    "# HIP_cases[\"GROUPcv\"]=\"PD\"\n",
    "# HIP_control[\"GROUPcv\"]=\"HC\"\n",
    "\n",
    "HIP_meta = pd.concat([HIP_cases, HIP_control], axis=0, ignore_index=True)\n",
    "HIP_meta[\"GROUPcv\"]= HIP_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "HIP_meta['MERGE_ID'] = \"HIP_\" + HIP_meta['GROUPcv'] +\"_\" + HIP_meta['CaseID'].str.replace('-','')\n",
    "HIP_covar['MERGE_ID'] = HIP_covar['COUNT_ID']\n",
    "# the fastqs follow COUNT_ID insteald of SEQ_ID naming convention\n",
    "HIP_covar['SEQ_ID'] = HIP_covar['COUNT_ID']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# there's a bug in the meta table... skip for now\n",
    "HIP_TABLE = pd.merge(HIP_covar, HIP_meta, on='MERGE_ID', how='outer')\n",
    "\n",
    "# HIP_TABLE = HIP_covar\n",
    "HIP_TABLE['subdir']=\"HIP\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = HIP_TABLE[[\"MERGE_ID\",\"SEQ_ID\",\"GROUPcv\",\"subdir\",'PD']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "### medial frontal gyrus samples\n",
    "MFG_meta = pd.concat([MFG_cases, MFG_control], axis=0, ignore_index=True)\n",
    "MFG_meta[\"GROUPcv\"]= MFG_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "# make a MERGE_ID column because the formatting is inconsistent\n",
    "MFG_meta['MERGE_ID'] = \"MFG_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "MFG_covar['MERGE_ID'] = MFG_covar['SAMPLE']\n",
    "# the fastqs are in SEQ_ID \n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "MFG_TABLE = pd.merge(MFG_covar, MFG_meta, on='MERGE_ID', how='inner')\n",
    "MFG_TABLE['subdir']=\"MFG\"\n",
    "\n",
    "\n",
    "\n",
    "# Substantia Nigra\n",
    "SN_meta = pd.concat([SN_cases, SN_control], axis=0, ignore_index=True)\n",
    "SN_meta[\"GROUPcv\"] = SN_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "SN_meta['MERGE_ID'] = \"SN_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "SN_covar['MERGE_ID'] = SN_covar['SAMPLE']\n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "SN_TABLE = pd.merge(SN_covar, SN_meta, on='MERGE_ID', how='outer')\n",
    "SN_TABLE['subdir']=\"SN\"\n",
    "\n",
    "\n",
    "### concatenate SN, MSG, and HIP tables into one 'all_samples' table\n",
    "all_samples = pd.concat([HIP_TABLE, MFG_TABLE, SN_TABLE], axis=0, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL = SAMPLE.merge(all_samples, left_on='sample_id', right_on='MERGE_ID', how='left')\n",
    "SAMPLE_ALL.to_csv(\"alternate_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['batch'] = SAMPLE_ALL['BATCH']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# SAMPLE = force_enum_string(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in SAMPLE: file_MD5**\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 75/75 empty rows\n",
      "\n",
      "\t- RIN: 75/75 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- pm_PH: 75/75 empty rows\n",
      "## Enum fields have valid values in SAMPLE. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sample_report = ReportCollector(destination=\"print\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE, sample_report)\n",
    "print(sample_report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>19-21</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>20-58</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0  MFG_HC_1225            12-25      12-25      rep1                1  \\\n",
       "1  MFG_HC_0602            06-02      06-02      rep1                1   \n",
       "2  MFG_PD_0009            00-09      00-09      rep1                1   \n",
       "3  MFG_PD_1921            19-21      19-21      rep1                1   \n",
       "4  MFG_PD_2058            20-58      20-58      rep1                1   \n",
       "\n",
       "   repeated_sample tissue          brain_region  source_RIN  RIN  ...   \n",
       "0                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...  \\\n",
       "1                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "2                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "3                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "4                0  brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id  disease_ontology_term_id   \n",
       "0                                  unknown              PATO:0000461  \\\n",
       "1                                  unknown              PATO:0000461   \n",
       "2                                  unknown             MONDO:0005180   \n",
       "3                                  unknown             MONDO:0005180   \n",
       "4                                  unknown             MONDO:0005180   \n",
       "\n",
       "  tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id   \n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004  \\\n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "2          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "3          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "4          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "\n",
       "  suspension_type DV2000 pm_PH donor_id    batch  \n",
       "0         nucleus    NaN   NaN      NaN  BATCH_4  \n",
       "1         nucleus    NaN   NaN      NaN  BATCH_4  \n",
       "2         nucleus    NaN   NaN      NaN  BATCH_4  \n",
       "3         nucleus    NaN   NaN      NaN  BATCH_4  \n",
       "4         nucleus    NaN   NaN      NaN  BATCH_4  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make the colunn order of SAMPLE match the CDE.Field\n",
    "# SAMPLE = SAMPLE[CDE.Field.tolist()]\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'source_sample_id', 'subject_id', 'replicate',\n",
       "       'replicate_count', 'repeated_sample', 'tissue', 'brain_region',\n",
       "       'source_RIN', 'RIN', 'molecular_source', 'input_cell_count', 'assay',\n",
       "       'sequencing_end', 'sequencing_length', 'sequencing_instrument',\n",
       "       'file_type', 'file_name', 'file_description', 'file_MD5', 'technology',\n",
       "       'omic', 'adjustment', 'content', 'time', 'header', 'annotation',\n",
       "       'preprocessing_references', 'configuration_file',\n",
       "       'organism_ontology_term_id', 'development_stage_ontology_term_id',\n",
       "       'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id',\n",
       "       'disease_ontology_term_id', 'tissue_ontology_term_id',\n",
       "       'cell_type_ontology_term_id', 'assay_ontology_term_id',\n",
       "       'suspension_type', 'DV2000', 'pm_PH', 'donor_id', 'batch'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix file_name and file_MD5 which need to be exploded (do this last for simplicity. i.e. to keep one sample per row rather than one file per row)\n",
    "\n",
    "# Step 1: Split the values in the columns based on commas\n",
    "SAMPLE['file_name'] = SAMPLE['file_name'].str.split(',')\n",
    "SAMPLE['file_MD5(R1,R2)'] = SAMPLE['file_MD5(R1,R2)'].str.split(',')\n",
    "\n",
    "# Step 2: Explode both 'file_name' and 'file_MD5(R1,R2)' columns together\n",
    "SAMPLE = SAMPLE.explode(['file_name', 'file_MD5(R1,R2)'])\n",
    "\n",
    "# Step 3: Rename the \"file_MD5(R1,R2)\" column to \"file_MD5\"\n",
    "SAMPLE = SAMPLE.rename(columns={\"file_MD5(R1,R2)\": \"file_MD5\"})\n",
    "\n",
    "SAMPLE.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['sample_id', 'source_sample_id', 'subject_id', 'replicate',\n",
       "       'replicate_count', 'repeated_sample', 'batch', 'tissue', 'brain_region',\n",
       "       'source_RIN', 'RIN', 'molecular_source', 'input_cell_count', 'assay',\n",
       "       'sequencing_end', 'sequencing_length', 'sequencing_instrument',\n",
       "       'file_type', 'file_name', 'file_description', 'file_MD5', 'technology',\n",
       "       'omic', 'adjustment', 'content', 'time', 'header', 'annotation',\n",
       "       'preprocessing_references', 'configuration_file',\n",
       "       'organism_ontology_term_id', 'development_stage_ontology_term_id',\n",
       "       'sex_ontology_term_id', 'self_reported_ethnicity_ontology_term_id',\n",
       "       'disease_ontology_term_id', 'tissue_ontology_term_id',\n",
       "       'cell_type_ontology_term_id', 'assay_ontology_term_id',\n",
       "       'suspension_type', 'DV200', 'pm_PH', 'donor_id'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Extract the fields with DataType as \"Enum\" or \"String\" for the \"sample\" table from CDE.csv\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "SAMPLE.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>BATCH_4</td>\n",
       "      <td>brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0  MFG_HC_1225            12-25      12-25      rep1                1  \\\n",
       "0  MFG_HC_1225            12-25      12-25      rep1                1   \n",
       "1  MFG_HC_0602            06-02      06-02      rep1                1   \n",
       "1  MFG_HC_0602            06-02      06-02      rep1                1   \n",
       "2  MFG_PD_0009            00-09      00-09      rep1                1   \n",
       "\n",
       "   repeated_sample    batch tissue          brain_region  source_RIN  ...   \n",
       "0                0  BATCH_4  brain  Middle_Frontal_Gyrus         NaN  ...  \\\n",
       "0                0  BATCH_4  brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "1                0  BATCH_4  brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "1                0  BATCH_4  brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "2                0  BATCH_4  brain  Middle_Frontal_Gyrus         NaN  ...   \n",
       "\n",
       "   sex_ontology_term_id self_reported_ethnicity_ontology_term_id   \n",
       "0   PATO:0000384 (male)                                  unknown  \\\n",
       "0   PATO:0000384 (male)                                  unknown   \n",
       "1   PATO:0000384 (male)                                  unknown   \n",
       "1   PATO:0000384 (male)                                  unknown   \n",
       "2   PATO:0000384 (male)                                  unknown   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id   \n",
       "0              PATO:0000461          UBERON:0002702  \\\n",
       "0              PATO:0000461          UBERON:0002702   \n",
       "1              PATO:0000461          UBERON:0002702   \n",
       "1              PATO:0000461          UBERON:0002702   \n",
       "2             MONDO:0005180          UBERON:0002702   \n",
       "\n",
       "  cell_type_ontology_term_id assay_ontology_term_id suspension_type DV200   \n",
       "0               NA(multiple)            EFO:0030004         nucleus        \\\n",
       "0               NA(multiple)            EFO:0030004         nucleus         \n",
       "1               NA(multiple)            EFO:0030004         nucleus         \n",
       "1               NA(multiple)            EFO:0030004         nucleus         \n",
       "2               NA(multiple)            EFO:0030004         nucleus         \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN      NaN  \n",
       "0   NaN      NaN  \n",
       "1   NaN      NaN  \n",
       "1   NaN      NaN  \n",
       "2   NaN      NaN  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SUBJECT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows\n",
      "## Enum fields have valid values in SUBJECT. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\", delimiter=\"\\t\", dtype=dtypes_dict)\n",
    "\n",
    "# SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "\n",
    "subject_report = ReportCollector(destination=\"print\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE, subject_report)\n",
    "print(subject_report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>duration_pmi</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>80</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>3.50</td>\n",
       "      <td>No PD nor other neurological disorder</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>84</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>2.66</td>\n",
       "      <td>Other neurological disorder</td>\n",
       "      <td>Mild Cognitive Impairment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>64</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>4.00</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>82</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>3.93</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>Banner Sun Health Research Institute</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>87</td>\n",
       "      <td>White</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>3.17</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id                          biobank_name organism   \n",
       "0    HC_1225             12-25  Banner Sun Health Research Institute    Human  \\\n",
       "1    HC_0602             06-02  Banner Sun Health Research Institute    Human   \n",
       "2    PD_0009             00-09  Banner Sun Health Research Institute    Human   \n",
       "3    PD_1921             19-21  Banner Sun Health Research Institute    Human   \n",
       "4    PD_2058             20-58  Banner Sun Health Research Institute    Human   \n",
       "\n",
       "    sex  age_at_collection   race     ethnicity  duration_pmi   \n",
       "0  Male                 80  White  Not Reported          3.50  \\\n",
       "1  Male                 84  White  Not Reported          2.66   \n",
       "2  Male                 64  White  Not Reported          4.00   \n",
       "3  Male                 82  White  Not Reported          3.93   \n",
       "4  Male                 87  White  Not Reported          3.17   \n",
       "\n",
       "                       primary_diagnosis     primary_diagnosis_text  \n",
       "0  No PD nor other neurological disorder                        NaN  \n",
       "1            Other neurological disorder  Mild Cognitive Impairment  \n",
       "2                          Idiopathic PD                        NaN  \n",
       "3                          Idiopathic PD                        NaN  \n",
       "4                          Idiopathic PD                        NaN  "
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "SUBJECT.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-lee/metadata')"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### CLINPATH\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 75/75 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 75/75 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 75/75 empty rows\n",
      "\n",
      "\t- path_year_death: 75/75 empty rows\n",
      "\n",
      "\t- brain_weight: 75/75 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 75/75 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "\t- region_level_2:Hippocampus\n",
      "\t>\t change to: Superior frontal gyrus, Middle frontal gyrus, Inferior frontal gyrus, Superior temporal gyrus, Middle temporal gyrus, Inferior temporal gyrus, Fusiform gyrus, Transentorhinal region, Entorinal region, Subiculum, CA1-CA4, Amygdala, Periamygdala cortex, Anterior cingulate gyrus, Posterior cingulate gyrus, Superior parietal lobule, Inferior parietal lobule, Parastriate cortex, Peristriate cortex, Striate cortex, Insular cortex, Caudate nucleus, Putamen, Globus pallidus, Thalamus, Subthalamic nucleus, Substantia nigra, Pontine tegmentum, Pontine base, Medulla tegmentum, Medulla base, Cerebellar vermis, Cerebellar hemisphere, Dentate nucleus, Right colon, Left colon, Unknown\n",
      "\t- path_braak_asyn:L. Olfactory Bulb-Only, Lla. Brainstem Predominant, Llb. Limbic Predominant, LV. Neocortical, Lll. Brainstem/Limbic, 0. No Lewy bodies\n",
      "\t>\t change to: 0, 1, 2, 3, 4, 5, 6, 1/2, 3/4, 4/5, 5/6\n",
      "\t- path_mckeith:L. Olfactory Bulb-Only, Lla. Brainstem Predominant, Llb. Limbic Predominant, LV. Neocortical, Lll. Brainstem/Limbic, 0. No Lewy bodies\n",
      "\t>\t change to: Neocortical, Limbic (transitional), Brainstem, Amygdala Predominant, Olfactory bulb only, Limbic, transitional (brainstem and limbic involvement), Diffuse, neocortical (brainstem, limbic and neocortical involvement), Olfactory Bulb-Only, Limbic (amygdala) predominant, Absent, Present, but extent unknown\n",
      "\t- path_nia_ri:Criteria not met, Not AD\n",
      "\t>\t change to: Low, Intermediate, High, None\n",
      "\t- TDP43:Na\n",
      "\t>\t change to: None in medial temporal lobe, Present in amygdala, only, Present in hippocampus, only, Present in amygdala and hippocampus, only, Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern), Unknown\n",
      "\t- amyloid_angiopathy_severity_scale:Cerebral amyloid angiopathy, temporal and occipital lobe, Cerebral amyloid angiopathy, frontal lobe\n",
      "\t>\t change to: None, Mild, Moderate, Severe, Not assessed, Unknown\n",
      "\t- path_ad_level:Microscopic changes of Alzheimer's disease, insufficient for diagnosis, Microscopic lesions of Alzheimer's disease, insufficient for diagnosis\n",
      "\t>\t change to: No evidence of Alzheimer's disease neuropathological change, Low level Alzheimer's disease neuropathological change, At least low level Alzheimer's disease neuropathological change, Intermediate level Alzheimer's disease neuropathological change, At least intermediate level Alzheimer's disease neuropathological change, High level Alzheimer's disease neuropathological change, Unknown\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (nan):**\n",
      "- hx_melanomaeducation_level \n",
      "- cognitive_status \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CLINPATH = pd.read_csv(f\"{metadata_path}/ogmetadata/CLINPATH.csv\",delimiter=\",\", dtype=dtypes_dict)\n",
    "\n",
    "clinpath_report = ReportCollector(destination=\"print\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE, clinpath_report)\n",
    "\n",
    "print(clinpath_report.get_log())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         l. Olfactory Bulb-Only\n",
       "1     lla. Brainstem Predominant\n",
       "2        llb. Limbic Predominant\n",
       "3                lV. Neocortical\n",
       "4                lV. Neocortical\n",
       "                 ...            \n",
       "70         lll. Brainstem/Limbic\n",
       "71               lV. Neocortical\n",
       "72               lV. Neocortical\n",
       "73         lll. Brainstem/Limbic\n",
       "74         lll. Brainstem/Limbic\n",
       "Name: path_mckeith, Length: 75, dtype: object"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# change \"Hippocampus\" to \"CA1-CA4\"\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace('Hippocampus', 'CA1-CA4')\n",
    "\n",
    "# skip hx_melanoma and education level for now as there is not a \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# leave cognitive status as is, since there is no \"Unknown\" or \"Not Reported\" option in the CDE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# potential \"path_braak_asyn\" coding . Note that \"I\" are actually \"l\"\n",
    "braak_map = {'l. Olfactory Bulb-Only':\"1/2\", 'la. Brainstem Predominant':\"3\",\n",
    "       'llb. Limbic Predominant':\"3/4\", 'lV. Neocortical':\"5\",\n",
    "       'lll. Brainstem/Limbic':\"3/4\", '0. No Lewy bodies':\"0\"}\n",
    "# set to NaN for now since this is actualy path_mckeith coding\n",
    "\n",
    "CLINPATH['path_braak_asyn'] = \"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mckeith_map = {'l. Olfactory Bulb-Only':\"Olfactory bulb only\", 'lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'llb. Limbic Predominant':\"Limbic (transitional)\", 'lV. Neocortical':\"Neocortical\",\n",
    "       'lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\"}\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path_ad_level</th>\n",
       "      <th>path_mckeith</th>\n",
       "      <th>path_nia_ri</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Olfactory bulb only</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Brainstem</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Limbic (transitional)</td>\n",
       "      <td>Not AD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Neocortical</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Neocortical</td>\n",
       "      <td>Low</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Amygdala Predominant</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Neocortical</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Neocortical</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td>Amygdala Predominant</td>\n",
       "      <td>Criteria not met</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td>Amygdala Predominant</td>\n",
       "      <td>Low</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows √ó 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        path_ad_level           path_mckeith   \n",
       "0   No evidence of Alzheimer's disease neuropathol...    Olfactory bulb only  \\\n",
       "1   No evidence of Alzheimer's disease neuropathol...              Brainstem   \n",
       "2   No evidence of Alzheimer's disease neuropathol...  Limbic (transitional)   \n",
       "3   No evidence of Alzheimer's disease neuropathol...            Neocortical   \n",
       "4   Microscopic changes of Alzheimer's disease, in...            Neocortical   \n",
       "..                                                ...                    ...   \n",
       "70  Microscopic changes of Alzheimer's disease, in...   Amygdala Predominant   \n",
       "71  No evidence of Alzheimer's disease neuropathol...            Neocortical   \n",
       "72  No evidence of Alzheimer's disease neuropathol...            Neocortical   \n",
       "73  No evidence of Alzheimer's disease neuropathol...   Amygdala Predominant   \n",
       "74  Microscopic changes of Alzheimer's disease, in...   Amygdala Predominant   \n",
       "\n",
       "         path_nia_ri  \n",
       "0   Criteria not met  \n",
       "1   Criteria not met  \n",
       "2             Not AD  \n",
       "3   Criteria not met  \n",
       "4                Low  \n",
       "..               ...  \n",
       "70  Criteria not met  \n",
       "71  Criteria not met  \n",
       "72  Criteria not met  \n",
       "73  Criteria not met  \n",
       "74               Low  \n",
       "\n",
       "[75 rows x 3 columns]"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH[['path_ad_level', 'path_mckeith', 'path_nia_ri']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 75/75 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 75/75 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 75/75 empty rows\n",
      "\n",
      "\t- path_year_death: 75/75 empty rows\n",
      "\n",
      "\t- brain_weight: 75/75 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 75/75 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "\t- path_nia_ri:Criteria not met, Not AD\n",
      "\t>\t change to: Low, Intermediate, High, None\n",
      "\t- TDP43:Na\n",
      "\t>\t change to: None in medial temporal lobe, Present in amygdala, only, Present in hippocampus, only, Present in amygdala and hippocampus, only, Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern), Unknown\n",
      "\t- amyloid_angiopathy_severity_scale:Cerebral amyloid angiopathy, temporal and occipital lobe, Cerebral amyloid angiopathy, frontal lobe\n",
      "\t>\t change to: None, Mild, Moderate, Severe, Not assessed, Unknown\n",
      "\t- path_ad_level:Microscopic changes of Alzheimer's disease, insufficient for diagnosis, Microscopic lesions of Alzheimer's disease, insufficient for diagnosis\n",
      "\t>\t change to: No evidence of Alzheimer's disease neuropathological change, Low level Alzheimer's disease neuropathological change, At least low level Alzheimer's disease neuropathological change, Intermediate level Alzheimer's disease neuropathological change, At least intermediate level Alzheimer's disease neuropathological change, High level Alzheimer's disease neuropathological change, Unknown\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (nan):**\n",
      "- hx_melanomaeducation_level \n",
      "- cognitive_status \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "clinpath_report = ReportCollector(destination=\"print\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE, clinpath_report)\n",
    "\n",
    "print(clinpath_report.get_log())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>sn_neuronal_loss</th>\n",
       "      <th>path_infarcs</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not AD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Low</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id time_from_baseline GP2_id hemisphere   \n",
       "0  MFG_HC_1225            12-25                  0    NaN    Unknown  \\\n",
       "1  MFG_HC_0602            06-02                  0    NaN    Unknown   \n",
       "2  MFG_PD_0009            00-09                  0    NaN    Unknown   \n",
       "3  MFG_PD_1921            19-21                  0    NaN    Unknown   \n",
       "4  MFG_PD_2058            20-58                  0    NaN    Unknown   \n",
       "\n",
       "  region_level_1        region_level_2 region_level_3 AMPPD_id family_history   \n",
       "0   Frontal lobe  Middle frontal gyrus        unknown      NaN   Not Reported  \\\n",
       "1   Frontal lobe  Middle frontal gyrus        unknown      NaN   Not Reported   \n",
       "2   Frontal lobe  Middle frontal gyrus        unknown      NaN   Not Reported   \n",
       "3   Frontal lobe  Middle frontal gyrus        unknown      NaN   Not Reported   \n",
       "4   Frontal lobe  Middle frontal gyrus        unknown      NaN   Not Reported   \n",
       "\n",
       "   ... sn_neuronal_loss  path_infarcs       path_nia_ri  path_nia_aa_a   \n",
       "0  ...              NaN           NaN  Criteria not met            NaN  \\\n",
       "1  ...              NaN           NaN  Criteria not met            NaN   \n",
       "2  ...              NaN           NaN            Not AD            NaN   \n",
       "3  ...              NaN           NaN  Criteria not met            NaN   \n",
       "4  ...              NaN           NaN               Low            NaN   \n",
       "\n",
       "  path_nia_aa_b path_nia_aa_c                         TDP43   \n",
       "0           NaN           NaN                           NaN  \\\n",
       "1           NaN           NaN                           NaN   \n",
       "2           NaN           NaN                           NaN   \n",
       "3           NaN           NaN  None in medial temporal lobe   \n",
       "4           NaN           NaN  None in medial temporal lobe   \n",
       "\n",
       "  arteriolosclerosis_severity_scale  amyloid_angiopathy_severity_scale   \n",
       "0                               NaN                                NaN  \\\n",
       "1                               NaN                                NaN   \n",
       "2                               NaN                                NaN   \n",
       "3                               NaN                                NaN   \n",
       "4                               NaN                                NaN   \n",
       "\n",
       "                                       path_ad_level  \n",
       "0  No evidence of Alzheimer's disease neuropathol...  \n",
       "1  No evidence of Alzheimer's disease neuropathol...  \n",
       "2  No evidence of Alzheimer's disease neuropathol...  \n",
       "3  No evidence of Alzheimer's disease neuropathol...  \n",
       "4  Microscopic changes of Alzheimer's disease, in...  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL_CP = SAMPLE_ALL.merge(CLINPATH, on='sample_id', how='outer')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_ALL_CP.to_csv(\"./clean/team-Lee/auxiluary_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-lee')"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the column order\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE)\n",
    "\n",
    "# write the clean metadata\n",
    "STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Lee\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make sure cleaned files are correct\n",
    "\n",
    "SUBJECT = read_meta_table(f\"{export_root}/SUBJECT.csv\", dtypes_dict)\n",
    "CLINPATH = read_meta_table(f\"{export_root}/CLINPATH.csv\", dtypes_dict)\n",
    "STUDY = read_meta_table(f\"{export_root}/STUDY.csv\", dtypes_dict)\n",
    "PROTOCOL = read_meta_table(f\"{export_root}/PROTOCOL.csv\", dtypes_dict)\n",
    "SAMPLE = read_meta_table(f\"{export_root}/SAMPLE.csv\", dtypes_dict)\n",
    "\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{export_root}/SUBJECT.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# CLINPATH = pd.read_csv(f\"{export_root}/CLINPATH.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# STUDY = pd.read_csv(f\"{export_root}/STUDY.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# PROTOCOL = pd.read_csv(f\"{export_root}/PROTOCOL.csv\",header=0,index_col=0, dtype=dtypes_dict)\n",
    "# SAMPLE = pd.read_csv(f\"{export_root}/SAMPLE.csv\",header=0,index_col=0, dtype=dtypes_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows\n",
      "## Enum fields have valid values in SUBJECT. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table, table_name = SUBJECT, \"SUBJECT\"\n",
    "\n",
    "report = ReportCollector(destination=\"print\")\n",
    "validate_table(table, table_name, CDE, report)\n",
    "print(report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 150/150 empty rows\n",
      "\n",
      "\t- RIN: 150/150 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- DV200: 150/150 empty rows\n",
      "\n",
      "\t- pm_PH: 150/150 empty rows\n",
      "## Enum fields have valid values in SAMPLE. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table, table_name = SAMPLE, \"SAMPLE\"\n",
    "\n",
    "report = ReportCollector(destination=\"print\")\n",
    "validate_table(table, table_name, CDE, report)\n",
    "print(report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 75/75 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 75/75 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 75/75 empty rows\n",
      "\n",
      "\t- path_year_death: 75/75 empty rows\n",
      "\n",
      "\t- brain_weight: 75/75 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 75/75 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "\t- path_nia_ri:Criteria not met, Not AD\n",
      "\t>\t change to: Low, Intermediate, High, None\n",
      "\t- TDP43:Na\n",
      "\t>\t change to: None in medial temporal lobe, Present in amygdala, only, Present in hippocampus, only, Present in amygdala and hippocampus, only, Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern), Unknown\n",
      "\t- amyloid_angiopathy_severity_scale:Cerebral amyloid angiopathy, temporal and occipital lobe, Cerebral amyloid angiopathy, frontal lobe\n",
      "\t>\t change to: None, Mild, Moderate, Severe, Not assessed, Unknown\n",
      "\t- path_ad_level:Microscopic changes of Alzheimer's disease, insufficient for diagnosis, Microscopic lesions of Alzheimer's disease, insufficient for diagnosis\n",
      "\t>\t change to: No evidence of Alzheimer's disease neuropathological change, Low level Alzheimer's disease neuropathological change, At least low level Alzheimer's disease neuropathological change, Intermediate level Alzheimer's disease neuropathological change, At least intermediate level Alzheimer's disease neuropathological change, High level Alzheimer's disease neuropathological change, Unknown\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (nan):**\n",
      "- hx_melanomaeducation_level \n",
      "- cognitive_status \n",
      "- path_braak_asyn \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- dig_slide_avail \n",
      "- quant_path_avail\n",
      "\n"
     ]
    }
   ],
   "source": [
    "table, table_name = CLINPATH, \"CLINPATH\"\n",
    "\n",
    "report = ReportCollector(destination=\"print\")\n",
    "validate_table(table, table_name, CDE, report)\n",
    "print(report.get_log())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scverse10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
