{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "# ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "15 September 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n",
    "\n",
    "## STEPS\n",
    "\n",
    "### imports\n",
    "- pandas\n",
    "- pathlib\n",
    "\n",
    "### Load CDE for validation\n",
    "- check all columns\n",
    "\n",
    "### Team Lee\n",
    "- load .tsv, csv tables\n",
    "- fix format\n",
    "- load additional metadata\n",
    "\n",
    "- add batch columns\n",
    "- add missing columns\n",
    "\n",
    "\n",
    "### Team Hafler\n",
    "- load excel file with tables\n",
    "- add batch info\n",
    "- add missing columns\n",
    "\n",
    "\n",
    "\n",
    "### Team Hardy\n",
    "- load excel file with tables\n",
    "- add batch info\n",
    "- add missing columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pathlib import Path\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load CDE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>ClinPath field</th>\n",
       "      <th>team_Hafler type</th>\n",
       "      <th>ClinPath description</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bard</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Project Description:   Brief description of th...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table                Field   \n",
       "0  STUDY         project_name  \\\n",
       "1  STUDY      project_dataset   \n",
       "2  STUDY  project_description   \n",
       "3  STUDY       ASAP_team_name   \n",
       "4  STUDY        ASAP_lab_name   \n",
       "\n",
       "                                         Description DataType  Required   \n",
       "0  Project Name:   A Title of the overall project...   String  Required  \\\n",
       "1  Dataset Name:   A unique name is required for ...   String  Required   \n",
       "2  Project Description:   Brief description of th...   String  Required   \n",
       "3  ASAP Team Name:   Name of the ASAP CRN Team. i...     Enum  Required   \n",
       "4  Lab Name. :   Lab name that is submitting data...   String  Required   \n",
       "\n",
       "                                          Validation  Unnamed: 6   \n",
       "0                                                NaN         NaN  \\\n",
       "1                                                NaN         NaN   \n",
       "2                                                NaN         NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...         NaN   \n",
       "4                                                NaN         NaN   \n",
       "\n",
       "  ClinPath field team_Hafler type ClinPath description Unnamed: 10  \n",
       "0            NaN              NaN                  NaN        bard  \n",
       "1            NaN              NaN                  NaN         NaN  \n",
       "2            NaN              NaN                  NaN         NaN  \n",
       "3            NaN              NaN                  NaN         NaN  \n",
       "4            NaN              NaN                  NaN         NaN  "
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDE_path = Path.cwd() / \"CDE.csv\" \n",
    "CDE = pd.read_csv(CDE_path )\n",
    "\n",
    "CDE.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# create helpers - validate_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_table(table: pd.DataFrame, table_name: str, CDE: pd.DataFrame):\n",
    "\n",
    "    retval = 1\n",
    "\n",
    "    # Filter out rows specific to the given table_name from the CDE\n",
    "    specific_cde_df = CDE[CDE['Table'] == table_name]\n",
    "    \n",
    "    # Extract fields that have a data type of \"Enum\" and retrieve their validation entries\n",
    "    enum_fields_dict = dict(zip(specific_cde_df[specific_cde_df['DataType'] == \"Enum\"]['Field'], \n",
    "                               specific_cde_df[specific_cde_df['DataType'] == \"Enum\"]['Validation']))\n",
    "    \n",
    "    # Extract fields that are marked as \"Required\"\n",
    "    required_fields = specific_cde_df[specific_cde_df['Required'] == \"Required\"]['Field'].tolist()\n",
    "    optional_fields = specific_cde_df[specific_cde_df['Required'] == \"Optional\"]['Field'].tolist()\n",
    "\n",
    "    table = force_enum_string(table, table_name, CDE)\n",
    "\n",
    "    # Check for missing \"Required\" fields\n",
    "    missing_required_fields = [field for field in required_fields if field not in table.columns]\n",
    "    \n",
    "    if missing_required_fields:\n",
    "        print(f\"\\tMissing Required Fields in {table_name}: {', '.join(missing_required_fields)}\")\n",
    "    else:\n",
    "        print(f\"\\tAll required fields are present in {table_name}.\")\n",
    "\n",
    "    # Check for empty or NaN values\n",
    "    for test_field,test_name in zip([required_fields, optional_fields], [\"Required\", \"Optional\"]):\n",
    "        empty_or_nan_fields = {}\n",
    "        for field in test_field:\n",
    "            if field in table.columns:\n",
    "                invalid_count = table[field].isna().sum()\n",
    "                if invalid_count > 0:\n",
    "                    empty_or_nan_fields[field] = invalid_count\n",
    "                    \n",
    "        if empty_or_nan_fields:\n",
    "            print(f\"\\t\\t{test_name} Fields with Empty or NaN values:\")\n",
    "            for field, count in empty_or_nan_fields.items():\n",
    "                print(f\"\\t\\t\\t- {field}: {count} rows\")\n",
    "            retval = 0\n",
    "        else:\n",
    "            print(f\"No empty or NaN values found in {test_name} fields.\")\n",
    "    \n",
    "\n",
    "\n",
    "    # Check for invalid Enum field values\n",
    "    invalid_field_values = {}\n",
    "    for field, validation_str in enum_fields_dict.items():\n",
    "        valid_values = eval(validation_str)\n",
    "        if field in table.columns:\n",
    "            invalid_values = table[~table[field].isin(valid_values)][field].unique()\n",
    "            if invalid_values.any():\n",
    "                invalid_field_values[field] = invalid_values\n",
    "    \n",
    "    if invalid_field_values:\n",
    "        print(\"\\tInvalid Field/Value pairs:\")\n",
    "        for field, values in invalid_field_values.items():\n",
    "            print(f\"\\t\\t\\t- {field}: {', '.join(map(str, values))}\")\n",
    "        retval = 0\n",
    "    else:\n",
    "        print(f\"\\tAll Enum fields have valid values in {table_name}.\")\n",
    "\n",
    "    return retval\n",
    "\n",
    "######## HELPERS ########\n",
    "# Define a function to only capitalize the first letter of a string\n",
    "def capitalize_first_letter(s):\n",
    "    if not isinstance(s, str) or len(s) == 0:  # Check if the value is a string and non-empty\n",
    "        return s\n",
    "    return s[0].upper() + s[1:]\n",
    "\n",
    "def force_enum_string(df, df_name, CDE):\n",
    "\n",
    "    string_enum_fields = CDE[(CDE[\"Table\"] == df_name) & \n",
    "                                (CDE[\"DataType\"].isin([\"Enum\", \"String\"]))][\"Field\"].tolist()\n",
    "    # Convert the specified columns to string data type using astype() without a loop\n",
    "    columns_to_convert = {col: 'str' for col in string_enum_fields if col in df.columns}\n",
    "    df = df.astype(columns_to_convert)\n",
    "\n",
    "    # enum_fields = CDE[ (CDE[\"Table\"] == df_name) & \n",
    "    #                             (CDE[\"DataType\"]==\"Enum\") ][\"Field\"].tolist()\n",
    "    \n",
    "    for col in string_enum_fields:\n",
    "        if col in df.columns and col not in [\"assay\", \"file_type\"]:\n",
    "            df[col] = df[col].apply(capitalize_first_letter)\n",
    "\n",
    "    return df\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write clean metadata tables according to CDE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SAMPLE = SAMPLE[CDE[CDE[\"Table\"]==\"SAMPLE\"].Field.tolist()]\n",
    "\n",
    "\n",
    "def reorder_table_to_CDE(df, df_name, CDE):\n",
    "    col_order = CDE[CDE[\"Table\"]==df_name].Field.tolist()\n",
    "    \n",
    "    df_out = pd.DataFrame()\n",
    "    for col in col_order:\n",
    "        if col in df.columns:   \n",
    "            df_out[col] = df[col]\n",
    "        else:\n",
    "            df_out[col] = \"\"\n",
    "\n",
    "    return df_out\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "## write clean metadata tables according to CDE \n",
    "# SAMPLE = SAMPLE[CDE[CDE[\"Table\"]==\"SAMPLE\"].Field.tolist()]\n",
    "\n",
    "\n",
    "def reorder_table_to_CDE(df, df_name, CDE):\n",
    "    col_order = CDE[CDE[\"Table\"]==df_name].Field.tolist()\n",
    "    \n",
    "    df_out = pd.DataFrame()\n",
    "    for col in col_order:\n",
    "        if col in df.columns:   \n",
    "            df_out[col] = df[col]\n",
    "        else:\n",
    "            df_out[col] = \"\"\n",
    "\n",
    "    return df_out\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clean each Team Table\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Lee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-lee\")\n",
    "metadata_path = data_path / \"metadata/ogmetadata\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\", delimiter=\"\\t\")\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\",delimiter=\",\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.tsv\",delimiter=\"\\t\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.tsv\",delimiter=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Team-Lee-Bras-Lab-Info</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>Data type</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Note</th>\n",
       "      <th>Required/Optional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name/Title</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unique and clear title.</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset name</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A Dataset name is required for each submission...</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Brief description of the goals and objectives ...</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team e.g. \"Scherzer\"</td>\n",
       "      <td>Enum</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bras</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>ASAP Lab under the above team e.g. \"Dong\"</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Unnamed: 0           Unnamed: 1   \n",
       "0  Is senescence a component of human PD and does...         project_name  \\\n",
       "1    Human snRNA-seq PD Senesence Jose Bras Team Lee      project_dataset   \n",
       "2  Characterize the neuropathological progression...  project_description   \n",
       "3                                           TEAM-LEE       ASAP_team_name   \n",
       "4                                               Bras        ASAP_lab_name   \n",
       "\n",
       "                                          Unnamed: 2 Team-Lee-Bras-Lab-Info   \n",
       "0                                 Project Name/Title                 String  \\\n",
       "1                                       Dataset name                 String   \n",
       "2  Brief description of the goals and objectives ...                 String   \n",
       "3                          ASAP Team e.g. \"Scherzer\"                   Enum   \n",
       "4          ASAP Lab under the above team e.g. \"Dong\"                 String   \n",
       "\n",
       "                                          Field   \n",
       "0                                           NaN  \\\n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]   \n",
       "4                                           NaN   \n",
       "\n",
       "                                         Description Data type  Validation   \n",
       "0                            Unique and clear title.  Required         NaN  \\\n",
       "1  A Dataset name is required for each submission...  Required         NaN   \n",
       "2                                                NaN  Required         NaN   \n",
       "3                                                NaN  Required         NaN   \n",
       "4                                                NaN  Required         NaN   \n",
       "\n",
       "   Note  Required/Optional  \n",
       "0   NaN                NaN  \n",
       "1   NaN                NaN  \n",
       "2   NaN                NaN  \n",
       "3   NaN                NaN  \n",
       "4   NaN                NaN  "
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = pd.read_csv(metadata_path / \"STUDY.tsv\",delimiter=\"\\t\")\n",
    "STUDY.to_csv(data_path / \"STUDY_.csv\")\n",
    "STUDY = pd.read_csv(data_path / \"STUDY_.csv\")\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name   \n",
       "1  Is senescence a component of human PD and does...  \\\n",
       "\n",
       "0                                  project_dataset   \n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee  \\\n",
       "\n",
       "0                                project_description ASAP_team_name   \n",
       "1  Characterize the neuropathological progression...       TEAM-LEE  \\\n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email   \n",
       "1          Bras   Jose, Bras  jose.bras@vai.org  \\\n",
       "\n",
       "0                                       submitter_id    submitter_name   \n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra  \\\n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI   \n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN  \\\n",
       "\n",
       "0 publication_PMID number_of_brain_samples   \n",
       "1              NaN                      75  \\\n",
       "\n",
       "0                                      brain_regions   \n",
       "1  hippocampus; middle frontal gyrus; substantia ...  \\\n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id   \n",
       "1  human PD and control postmortem brains       NaN                  NaN  \\\n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"Unnamed: 1\",\"Unnamed: 0\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "# STUDY[[\"Unnamed: 1\"]].transpose().reset_index().drop(columns=[\"index\"]), tmp\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect additional metadata from covar.csv .. i.e. batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "metadata_path = Path.home() / (\"Projects/ASAP/team-lee/metadata\")\n",
    "HIP_covar = pd.read_csv(f\"{metadata_path}/HIP/covar.csv\")\n",
    "HIP_cases = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "HIP_control = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{metadata_path}/MFG/covar.csv\") # includes 'PMI' ?\n",
    "MFG_cases = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "MFG_control = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "\n",
    "SN_covar = pd.read_csv(f\"{metadata_path}/SN/covar.csv\")\n",
    "SN_cases = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "SN_control = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_controls.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hippocampus samples\n",
    "# HIP_cases[\"GROUPcv\"]=\"PD\"\n",
    "# HIP_control[\"GROUPcv\"]=\"HC\"\n",
    "\n",
    "HIP_meta = pd.concat([HIP_cases, HIP_control], axis=0, ignore_index=True)\n",
    "HIP_meta[\"GROUPcv\"]= HIP_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "HIP_meta['MERGE_ID'] = \"HIP_\" + HIP_meta['GROUPcv'] +\"_\" + HIP_meta['CaseID'].str.replace('-','')\n",
    "HIP_covar['MERGE_ID'] = HIP_covar['COUNT_ID']\n",
    "# the fastqs follow COUNT_ID insteald of SEQ_ID naming convention\n",
    "HIP_covar['SEQ_ID'] = HIP_covar['COUNT_ID']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [],
   "source": [
    "# there's a bug in the meta table... skip for now\n",
    "HIP_TABLE = pd.merge(HIP_covar, HIP_meta, on='MERGE_ID', how='outer')\n",
    "\n",
    "# HIP_TABLE = HIP_covar\n",
    "HIP_TABLE['subdir']=\"HIP\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = HIP_TABLE[[\"MERGE_ID\",\"SEQ_ID\",\"GROUPcv\",\"subdir\",'PD']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "### medial frontal gyrus samples\n",
    "MFG_meta = pd.concat([MFG_cases, MFG_control], axis=0, ignore_index=True)\n",
    "MFG_meta[\"GROUPcv\"]= MFG_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "# make a MERGE_ID column because the formatting is inconsistent\n",
    "MFG_meta['MERGE_ID'] = \"MFG_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "MFG_covar['MERGE_ID'] = MFG_covar['SAMPLE']\n",
    "# the fastqs are in SEQ_ID \n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "MFG_TABLE = pd.merge(MFG_covar, MFG_meta, on='MERGE_ID', how='inner')\n",
    "MFG_TABLE['subdir']=\"MFG\"\n",
    "\n",
    "\n",
    "\n",
    "# Substantia Nigra\n",
    "SN_meta = pd.concat([SN_cases, SN_control], axis=0, ignore_index=True)\n",
    "SN_meta[\"GROUPcv\"] = SN_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "SN_meta['MERGE_ID'] = \"SN_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "SN_covar['MERGE_ID'] = SN_covar['SAMPLE']\n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "SN_TABLE = pd.merge(SN_covar, SN_meta, on='MERGE_ID', how='outer')\n",
    "SN_TABLE['subdir']=\"SN\"\n",
    "\n",
    "\n",
    "### concatenate SN, MSG, and HIP tables into one 'all_samples' table\n",
    "all_samples = pd.concat([HIP_TABLE, MFG_TABLE, SN_TABLE], axis=0, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL = SAMPLE.merge(all_samples, left_on='sample_id', right_on='MERGE_ID', how='left')\n",
    "SAMPLE_ALL.to_csv(\"alternate_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['batch'] = SAMPLE_ALL['BATCH']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the fields with DataType as \"Enum\" or \"String\" for the \"sample\" table from CDE.csv\n",
    "\n",
    "SAMPLE = force_enum_string(SAMPLE, \"SAMPLE\", CDE)\n",
    "# for field in string_enum_fields:\n",
    "#     if field in SAMPLE.columns:\n",
    "#         SAMPLE[field] = SAMPLE[field].astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name   \n",
       "1  Is senescence a component of human PD and does...  \\\n",
       "\n",
       "0                                  project_dataset   \n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee  \\\n",
       "\n",
       "0                                project_description ASAP_team_name   \n",
       "1  Characterize the neuropathological progression...       TEAM-LEE  \\\n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email   \n",
       "1          Bras   Jose, Bras  jose.bras@vai.org  \\\n",
       "\n",
       "0                                       submitter_id    submitter_name   \n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra  \\\n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI   \n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN  \\\n",
       "\n",
       "0 publication_PMID number_of_brain_samples   \n",
       "1              NaN                      75  \\\n",
       "\n",
       "0                                      brain_regions   \n",
       "1  hippocampus; middle frontal gyrus; substantia ...  \\\n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id   \n",
       "1  human PD and control postmortem brains       NaN                  NaN  \\\n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tMissing Required Fields in STUDY: contributor_names\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in STUDY.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 204,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing the function with STUDY.csv and CDE.csv\n",
    "STUDY.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "validate_table(STUDY, \"STUDY\", CDE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in STUDY.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in STUDY.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 205,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Need to rename submitter_id to contributor_names\n",
    "STUDY = STUDY.rename(columns={\"submitter_id\": \"contributor_names\"})\n",
    "validate_table(STUDY, \"STUDY\", CDE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in PROTOCOL.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in PROTOCOL.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing the function with PROTOCOL.csv and CDE.csv\n",
    "\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SUBJECT.\n",
      "No empty or NaN values found in Required fields.\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- primary_diagnosis_text: 23 rows\n",
      "\tAll Enum fields have valid values in SUBJECT.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract the fields with DataType as \"Enum\" or \"String\" for the \"sample\" table from CDE.csv\n",
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "\n",
    "SUBJECT = force_enum_string(SUBJECT, \"SUBJECT\", CDE)\n",
    "\n",
    "# Testing the function with SUBJECT.csv and CDE.csv\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tMissing Required Fields in SAMPLE: file_MD5\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- source_RIN: 75 rows\n",
      "\t\t\t- RIN: 75 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- pm_PH: 75 rows\n",
      "\tAll Enum fields have valid values in SAMPLE.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "SAMPLE = force_enum_string(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "\n",
    "# Testing the function with SAMPLE.csv and CDE.csv\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "# sequence length will need to be converted to a string\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE['file_type'] = SAMPLE['file_type'].replace({\"Fastq\":\"fastq\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>19-21</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>20-58</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0  MFG_HC_1225            12-25      12-25      Rep1                1  \\\n",
       "1  MFG_HC_0602            06-02      06-02      Rep1                1   \n",
       "2  MFG_PD_0009            00-09      00-09      Rep1                1   \n",
       "3  MFG_PD_1921            19-21      19-21      Rep1                1   \n",
       "4  MFG_PD_2058            20-58      20-58      Rep1                1   \n",
       "\n",
       "   repeated_sample tissue          brain_region  source_RIN  RIN  ...   \n",
       "0                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...  \\\n",
       "1                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "2                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "3                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "4                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id  disease_ontology_term_id   \n",
       "0                                  Unknown              PATO:0000461  \\\n",
       "1                                  Unknown              PATO:0000461   \n",
       "2                                  Unknown             MONDO:0005180   \n",
       "3                                  Unknown             MONDO:0005180   \n",
       "4                                  Unknown             MONDO:0005180   \n",
       "\n",
       "  tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id   \n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004  \\\n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "2          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "3          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "4          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "\n",
       "  suspension_type DV2000 pm_PH donor_id    batch  \n",
       "0         nucleus    NaN   NaN           BATCH_4  \n",
       "1         nucleus    NaN   NaN           BATCH_4  \n",
       "2         nucleus    NaN   NaN           BATCH_4  \n",
       "3         nucleus    NaN   NaN           BATCH_4  \n",
       "4         nucleus    NaN   NaN           BATCH_4  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make the colunn order of SAMPLE match the CDE.Field\n",
    "# SAMPLE = SAMPLE[CDE.Field.tolist()]\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix file_name and file_MD5 which need to be exploded (do this last for simplicity. i.e. to keep one sample per row rather than one file per row)\n",
    "\n",
    "# Step 1: Split the values in the columns based on commas\n",
    "SAMPLE['file_name'] = SAMPLE['file_name'].str.split(',')\n",
    "SAMPLE['file_MD5(R1,R2)'] = SAMPLE['file_MD5(R1,R2)'].str.split(',')\n",
    "\n",
    "# Step 2: Explode both 'file_name' and 'file_MD5(R1,R2)' columns together\n",
    "SAMPLE = SAMPLE.explode(['file_name', 'file_MD5(R1,R2)'])\n",
    "\n",
    "# Step 3: Rename the \"file_MD5(R1,R2)\" column to \"file_MD5\"\n",
    "SAMPLE = SAMPLE.rename(columns={\"file_MD5(R1,R2)\": \"file_MD5\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0  MFG_HC_1225            12-25      12-25      Rep1                1  \\\n",
       "0  MFG_HC_1225            12-25      12-25      Rep1                1   \n",
       "1  MFG_HC_0602            06-02      06-02      Rep1                1   \n",
       "1  MFG_HC_0602            06-02      06-02      Rep1                1   \n",
       "2  MFG_PD_0009            00-09      00-09      Rep1                1   \n",
       "\n",
       "   repeated_sample tissue          brain_region  source_RIN  RIN  ...   \n",
       "0                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...  \\\n",
       "0                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "1                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "1                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "2                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id  disease_ontology_term_id   \n",
       "0                                  Unknown              PATO:0000461  \\\n",
       "0                                  Unknown              PATO:0000461   \n",
       "1                                  Unknown              PATO:0000461   \n",
       "1                                  Unknown              PATO:0000461   \n",
       "2                                  Unknown             MONDO:0005180   \n",
       "\n",
       "  tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id   \n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004  \\\n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "2          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "\n",
       "  suspension_type DV2000 pm_PH donor_id    batch  \n",
       "0         nucleus    NaN   NaN           BATCH_4  \n",
       "0         nucleus    NaN   NaN           BATCH_4  \n",
       "1         nucleus    NaN   NaN           BATCH_4  \n",
       "1         nucleus    NaN   NaN           BATCH_4  \n",
       "2         nucleus    NaN   NaN           BATCH_4  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in CLINPATH.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- age_at_onset: 75 rows\n",
      "\t\t\t- age_at_diagnosis: 75 rows\n",
      "\t\t\t- first_motor_symptom: 75 rows\n",
      "\t\t\t- path_year_death: 75 rows\n",
      "\t\t\t- brain_weight: 75 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- smoking_years: 75 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- region_level_2: Hippocampus\n",
      "\t\t\t- path_braak_asyn: L. Olfactory Bulb-Only, Lla. Brainstem Predominant, Llb. Limbic Predominant, LV. Neocortical, Lll. Brainstem/Limbic, 0. No Lewy bodies\n",
      "\t\t\t- path_mckeith: L. Olfactory Bulb-Only, Lla. Brainstem Predominant, Llb. Limbic Predominant, LV. Neocortical, Lll. Brainstem/Limbic, 0. No Lewy bodies\n",
      "\t\t\t- path_nia_ri: Criteria not met, Not AD\n",
      "\t\t\t- TDP43: , Na\n",
      "\t\t\t- amyloid_angiopathy_severity_scale: , Cerebral amyloid angiopathy, temporal and occipital lobe, Cerebral amyloid angiopathy, frontal lobe\n",
      "\t\t\t- path_ad_level: Microscopic changes of Alzheimer's disease, insufficient for diagnosis, Microscopic lesions of Alzheimer's disease, insufficient for diagnosis\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH = force_enum_string(CLINPATH, \"CLINPATH\", CDE)\n",
    "CLINPATH.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "# Testing the function with CLINPATH.csv and CDE.csv\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['region_level_2'].unique()\n",
    "\n",
    "# change \"Hippocampus\" to \"CA1-CA4\"\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace('Hippocampus', 'CA1-CA4')\n",
    "\n",
    "# skip hx_melanoma and education level for now as there is not a \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# leave te APOE_e4_status as is for now . multiple are coded as \"2,3\" \n",
    "# leave cognitive status as is, since there is no \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# potential \"path_braak_asyn\" coding \n",
    "braak_map = {'L. Olfactory Bulb-Only':\"1/2\", 'Lla. Brainstem Predominant':\"3\",\n",
    "       'Llb. Limbic Predominant':\"3/4\", 'LV. Neocortical':\"5\",\n",
    "       'Lll. Brainstem/Limbic':\"3/4\", '0. No Lewy bodies':\"0\"}\n",
    "# set to NaN for now since this is actualy path_mckeith coding\n",
    "\n",
    "CLINPATH['path_braak_asyn'] = \"\"\n",
    "\n",
    "mckeith_map = {'L. Olfactory Bulb-Only':\"Olfactory bulb only\", 'Lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'Llb. Limbic Predominant':\"Limbic (transitional)\", 'LV. Neocortical':\"Neocortical\",\n",
    "       'Lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\"}\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in CLINPATH.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- age_at_onset: 75 rows\n",
      "\t\t\t- age_at_diagnosis: 75 rows\n",
      "\t\t\t- first_motor_symptom: 75 rows\n",
      "\t\t\t- path_year_death: 75 rows\n",
      "\t\t\t- brain_weight: 75 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- smoking_years: 75 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- path_nia_ri: Criteria not met, Not AD\n",
      "\t\t\t- TDP43: , Na\n",
      "\t\t\t- amyloid_angiopathy_severity_scale: , Cerebral amyloid angiopathy, temporal and occipital lobe, Cerebral amyloid angiopathy, frontal lobe\n",
      "\t\t\t- path_ad_level: Microscopic changes of Alzheimer's disease, insufficient for diagnosis, Microscopic lesions of Alzheimer's disease, insufficient for diagnosis\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validate_table(CLINPATH, \"CLINPATH\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>sn_neuronal_loss</th>\n",
       "      <th>path_infarcs</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Not AD</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Low</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>SN_PD_1858</td>\n",
       "      <td>18-58</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>SN_PD_1902</td>\n",
       "      <td>19-02</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>SN_PD_1973</td>\n",
       "      <td>19-73</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>SN_PD_2005</td>\n",
       "      <td>20-05</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>Yes</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>SN_PD_2038</td>\n",
       "      <td>20-38</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Low</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Present in amygdala, only</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id source_sample_id time_from_baseline GP2_id hemisphere   \n",
       "0   MFG_HC_1225            12-25                  0           Unknown  \\\n",
       "1   MFG_HC_0602            06-02                  0           Unknown   \n",
       "2   MFG_PD_0009            00-09                  0           Unknown   \n",
       "3   MFG_PD_1921            19-21                  0           Unknown   \n",
       "4   MFG_PD_2058            20-58                  0           Unknown   \n",
       "..          ...              ...                ...    ...        ...   \n",
       "70   SN_PD_1858            18-58                  0           Unknown   \n",
       "71   SN_PD_1902            19-02                  0           Unknown   \n",
       "72   SN_PD_1973            19-73                  0           Unknown   \n",
       "73   SN_PD_2005            20-05                  0           Unknown   \n",
       "74   SN_PD_2038            20-38                  0           Unknown   \n",
       "\n",
       "   region_level_1        region_level_2 region_level_3 AMPPD_id   \n",
       "0    Frontal lobe  Middle frontal gyrus        Unknown           \\\n",
       "1    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "2    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "3    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "4    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "..            ...                   ...            ...      ...   \n",
       "70       Midbrain      Substantia nigra        Unknown            \n",
       "71       Midbrain      Substantia nigra        Unknown            \n",
       "72       Midbrain      Substantia nigra        Unknown            \n",
       "73       Midbrain      Substantia nigra        Unknown            \n",
       "74       Midbrain      Substantia nigra        Unknown            \n",
       "\n",
       "   family_history  ... sn_neuronal_loss  path_infarcs       path_nia_ri   \n",
       "0    Not Reported  ...                                 Criteria not met  \\\n",
       "1    Not Reported  ...                                 Criteria not met   \n",
       "2    Not Reported  ...                                           Not AD   \n",
       "3    Not Reported  ...                                 Criteria not met   \n",
       "4    Not Reported  ...                                              Low   \n",
       "..            ...  ...              ...           ...               ...   \n",
       "70   Not Reported  ...                                 Criteria not met   \n",
       "71   Not Reported  ...                                 Criteria not met   \n",
       "72   Not Reported  ...                            Yes  Criteria not met   \n",
       "73   Not Reported  ...                            Yes  Criteria not met   \n",
       "74   Not Reported  ...                                              Low   \n",
       "\n",
       "    path_nia_aa_a path_nia_aa_b path_nia_aa_c                         TDP43   \n",
       "0                                                                            \\\n",
       "1                                                                             \n",
       "2                                                                             \n",
       "3                                              None in medial temporal lobe   \n",
       "4                                              None in medial temporal lobe   \n",
       "..            ...           ...           ...                           ...   \n",
       "70                                                                  Unknown   \n",
       "71                                             None in medial temporal lobe   \n",
       "72                                             None in medial temporal lobe   \n",
       "73                                             None in medial temporal lobe   \n",
       "74                                                Present in amygdala, only   \n",
       "\n",
       "   arteriolosclerosis_severity_scale  amyloid_angiopathy_severity_scale   \n",
       "0                                                                        \\\n",
       "1                                                                         \n",
       "2                                                                         \n",
       "3                                                                         \n",
       "4                                                                         \n",
       "..                               ...                                ...   \n",
       "70                                                                        \n",
       "71                                                                        \n",
       "72                                                                        \n",
       "73                                                                        \n",
       "74                                                                        \n",
       "\n",
       "                                        path_ad_level  \n",
       "0   No evidence of Alzheimer's disease neuropathol...  \n",
       "1   No evidence of Alzheimer's disease neuropathol...  \n",
       "2   No evidence of Alzheimer's disease neuropathol...  \n",
       "3   No evidence of Alzheimer's disease neuropathol...  \n",
       "4   Microscopic changes of Alzheimer's disease, in...  \n",
       "..                                                ...  \n",
       "70  Microscopic changes of Alzheimer's disease, in...  \n",
       "71  No evidence of Alzheimer's disease neuropathol...  \n",
       "72  No evidence of Alzheimer's disease neuropathol...  \n",
       "73  No evidence of Alzheimer's disease neuropathol...  \n",
       "74  Microscopic changes of Alzheimer's disease, in...  \n",
       "\n",
       "[75 rows x 52 columns]"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL_CP = SAMPLE_ALL.merge(CLINPATH, on='sample_id', how='outer')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_ALL_CP.to_csv(\"./clean/team-Lee/auxiluary_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the column order\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE)\n",
    "\n",
    "# write the clean metadata\n",
    "STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Lee\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "      <th>dig_slide_avail</th>\n",
       "      <th>quant_path_avail</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Not AD</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MFG_PD_1921</td>\n",
       "      <td>19-21</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MFG_PD_2058</td>\n",
       "      <td>20-58</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Middle frontal gyrus</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Low</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>SN_PD_1858</td>\n",
       "      <td>18-58</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>SN_PD_1902</td>\n",
       "      <td>19-02</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>SN_PD_1973</td>\n",
       "      <td>19-73</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>SN_PD_2005</td>\n",
       "      <td>20-05</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Criteria not met</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>None in medial temporal lobe</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>No evidence of Alzheimer's disease neuropathol...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>SN_PD_2038</td>\n",
       "      <td>20-38</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td></td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>...</td>\n",
       "      <td>Low</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Present in amygdala, only</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Microscopic changes of Alzheimer's disease, in...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id source_sample_id time_from_baseline GP2_id hemisphere   \n",
       "0   MFG_HC_1225            12-25                  0           Unknown  \\\n",
       "1   MFG_HC_0602            06-02                  0           Unknown   \n",
       "2   MFG_PD_0009            00-09                  0           Unknown   \n",
       "3   MFG_PD_1921            19-21                  0           Unknown   \n",
       "4   MFG_PD_2058            20-58                  0           Unknown   \n",
       "..          ...              ...                ...    ...        ...   \n",
       "70   SN_PD_1858            18-58                  0           Unknown   \n",
       "71   SN_PD_1902            19-02                  0           Unknown   \n",
       "72   SN_PD_1973            19-73                  0           Unknown   \n",
       "73   SN_PD_2005            20-05                  0           Unknown   \n",
       "74   SN_PD_2038            20-38                  0           Unknown   \n",
       "\n",
       "   region_level_1        region_level_2 region_level_3 AMPPD_id   \n",
       "0    Frontal lobe  Middle frontal gyrus        Unknown           \\\n",
       "1    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "2    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "3    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "4    Frontal lobe  Middle frontal gyrus        Unknown            \n",
       "..            ...                   ...            ...      ...   \n",
       "70       Midbrain      Substantia nigra        Unknown            \n",
       "71       Midbrain      Substantia nigra        Unknown            \n",
       "72       Midbrain      Substantia nigra        Unknown            \n",
       "73       Midbrain      Substantia nigra        Unknown            \n",
       "74       Midbrain      Substantia nigra        Unknown            \n",
       "\n",
       "   family_history  ...       path_nia_ri  path_nia_aa_a  path_nia_aa_b   \n",
       "0    Not Reported  ...  Criteria not met                                \\\n",
       "1    Not Reported  ...  Criteria not met                                 \n",
       "2    Not Reported  ...            Not AD                                 \n",
       "3    Not Reported  ...  Criteria not met                                 \n",
       "4    Not Reported  ...               Low                                 \n",
       "..            ...  ...               ...            ...            ...   \n",
       "70   Not Reported  ...  Criteria not met                                 \n",
       "71   Not Reported  ...  Criteria not met                                 \n",
       "72   Not Reported  ...  Criteria not met                                 \n",
       "73   Not Reported  ...  Criteria not met                                 \n",
       "74   Not Reported  ...               Low                                 \n",
       "\n",
       "    path_nia_aa_c                         TDP43   \n",
       "0                                                \\\n",
       "1                                                 \n",
       "2                                                 \n",
       "3                  None in medial temporal lobe   \n",
       "4                  None in medial temporal lobe   \n",
       "..            ...                           ...   \n",
       "70                                      Unknown   \n",
       "71                 None in medial temporal lobe   \n",
       "72                 None in medial temporal lobe   \n",
       "73                 None in medial temporal lobe   \n",
       "74                    Present in amygdala, only   \n",
       "\n",
       "   arteriolosclerosis_severity_scale amyloid_angiopathy_severity_scale   \n",
       "0                                                                       \\\n",
       "1                                                                        \n",
       "2                                                                        \n",
       "3                                                                        \n",
       "4                                                                        \n",
       "..                               ...                               ...   \n",
       "70                                                                       \n",
       "71                                                                       \n",
       "72                                                                       \n",
       "73                                                                       \n",
       "74                                                                       \n",
       "\n",
       "                                        path_ad_level  dig_slide_avail   \n",
       "0   No evidence of Alzheimer's disease neuropathol...                   \\\n",
       "1   No evidence of Alzheimer's disease neuropathol...                    \n",
       "2   No evidence of Alzheimer's disease neuropathol...                    \n",
       "3   No evidence of Alzheimer's disease neuropathol...                    \n",
       "4   Microscopic changes of Alzheimer's disease, in...                    \n",
       "..                                                ...              ...   \n",
       "70  Microscopic changes of Alzheimer's disease, in...                    \n",
       "71  No evidence of Alzheimer's disease neuropathol...                    \n",
       "72  No evidence of Alzheimer's disease neuropathol...                    \n",
       "73  No evidence of Alzheimer's disease neuropathol...                    \n",
       "74  Microscopic changes of Alzheimer's disease, in...                    \n",
       "\n",
       "   quant_path_avail  \n",
       "0                    \n",
       "1                    \n",
       "2                    \n",
       "3                    \n",
       "4                    \n",
       "..              ...  \n",
       "70                   \n",
       "71                   \n",
       "72                   \n",
       "73                   \n",
       "74                   \n",
       "\n",
       "[75 rows x 54 columns]"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Hafler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-hafler/metadata')"
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## convert to seurat Object\n",
    "data_path = Path.home() / (\"Projects/ASAP\")\n",
    "metadata_path = data_path / \"team-hafler/metadata\"\n",
    "\n",
    "sheets = [\"SAMPLE\",\"SUBJECT\",\"CLINPATH\",\"STUDY\",\"PROTOCOL\"]\n",
    "excel_path = data_path / \"ASAP_CDE_ALL_Team_Hafler_v1.xlsx\"\n",
    "STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\",header=1).drop(columns=\"Field\")\n",
    "CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\",header=1).drop(columns=\"Field\")\n",
    "SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\",header=1).drop(columns=\"Field\")\n",
    "SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\",header=1).drop(columns=\"Field\")\n",
    "PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\",header=1).drop(columns=\"Field\")\n",
    "metadata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_hafler_batch(sample_df):\n",
    "\n",
    "    # First batch: HSDG07HC HSDG10HC HSDG148PD HSDG199PD\n",
    "    # batch[batch.sample_id in ['hSDG07HC', 'hSDG10HC', 'hSDG148PD', 'hSDG199PD']]=1\n",
    "    Batch_1 = ['hSDG07', 'hSDG10', 'hSDG148', 'hSDG199'] \n",
    "    # Second batch: hsDG101HC hsDG13HC hsDG151PD hsDG197PD hsDG30HC hsDG99HC\n",
    "    Batch_2 = ['hSDG101', 'hSDG13', 'hSDG151', 'hSDG197', 'hSDG30', 'hSDG99']\n",
    "    # Third batch: hsDG142PD hsDG208PD\n",
    "    Batch_3 = ['hSDG142', 'hSDG208'] \n",
    "\n",
    "\n",
    "    batch_col = []\n",
    "    for row in sample_df.sample_id:\n",
    "        if row in Batch_1:\n",
    "            batch_col.append(\"Batch_1\")\n",
    "        elif row in Batch_2:\n",
    "            batch_col.append(\"Batch_2\")\n",
    "        elif row in Batch_3:\n",
    "            batch_col.append(\"Batch_3\")\n",
    "        else:\n",
    "            print(\"ERROR >>>>>>>> not no batch info\")\n",
    "            batch_col.append(\"\")\n",
    "\n",
    "\n",
    "    sample_df['batch'] = batch_col\n",
    "    return sample_df\n",
    "\n",
    "SAMPLE = add_hafler_batch(SAMPLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>hSDG10</td>\n",
       "      <td>hSDG10</td>\n",
       "      <td>HC04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>hSDG13</td>\n",
       "      <td>hSDG13</td>\n",
       "      <td>HC02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>hSDG142</td>\n",
       "      <td>hSDG142</td>\n",
       "      <td>PD04</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>hSDG148</td>\n",
       "      <td>hSDG148</td>\n",
       "      <td>PD05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>hSDG151</td>\n",
       "      <td>hSDG151</td>\n",
       "      <td>PD01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>hSDG197</td>\n",
       "      <td>hSDG197</td>\n",
       "      <td>PD02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>hSDG199</td>\n",
       "      <td>hSDG199</td>\n",
       "      <td>PD06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>hSDG208</td>\n",
       "      <td>hSDG208</td>\n",
       "      <td>PD03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_3</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>hSDG30</td>\n",
       "      <td>hSDG30</td>\n",
       "      <td>HC05</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>hSDG99</td>\n",
       "      <td>hSDG99</td>\n",
       "      <td>HC06</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   sample_id source_sample_id subject_id replicate  replicate_count   \n",
       "0     hSDG07           hSDG07       HC01      Rep1                1  \\\n",
       "1     hSDG07           hSDG07       HC01      Rep1                1   \n",
       "2     hSDG07           hSDG07       HC01      Rep1                1   \n",
       "3    hSDG101          hSDG101       HC03      Rep1                1   \n",
       "4    hSDG101          hSDG101       HC03      Rep1                1   \n",
       "5    hSDG101          hSDG101       HC03      Rep1                1   \n",
       "6     hSDG10           hSDG10       HC04      Rep1                1   \n",
       "7     hSDG10           hSDG10       HC04      Rep1                1   \n",
       "8     hSDG10           hSDG10       HC04      Rep1                1   \n",
       "9     hSDG13           hSDG13       HC02      Rep1                1   \n",
       "10    hSDG13           hSDG13       HC02      Rep1                1   \n",
       "11    hSDG13           hSDG13       HC02      Rep1                1   \n",
       "12   hSDG142          hSDG142       PD04      Rep1                1   \n",
       "13   hSDG142          hSDG142       PD04      Rep1                1   \n",
       "14   hSDG142          hSDG142       PD04      Rep1                1   \n",
       "15   hSDG148          hSDG148       PD05      Rep1                1   \n",
       "16   hSDG148          hSDG148       PD05      Rep1                1   \n",
       "17   hSDG148          hSDG148       PD05      Rep1                1   \n",
       "18   hSDG151          hSDG151       PD01      Rep1                1   \n",
       "19   hSDG151          hSDG151       PD01      Rep1                1   \n",
       "20   hSDG151          hSDG151       PD01      Rep1                1   \n",
       "21   hSDG197          hSDG197       PD02      Rep1                1   \n",
       "22   hSDG197          hSDG197       PD02      Rep1                1   \n",
       "23   hSDG197          hSDG197       PD02      Rep1                1   \n",
       "24   hSDG199          hSDG199       PD06      Rep1                1   \n",
       "25   hSDG199          hSDG199       PD06      Rep1                1   \n",
       "26   hSDG199          hSDG199       PD06      Rep1                1   \n",
       "27   hSDG208          hSDG208       PD03      Rep1                1   \n",
       "28   hSDG208          hSDG208       PD03      Rep1                1   \n",
       "29   hSDG208          hSDG208       PD03      Rep1                1   \n",
       "30    hSDG30           hSDG30       HC05      Rep1                1   \n",
       "31    hSDG30           hSDG30       HC05      Rep1                1   \n",
       "32    hSDG30           hSDG30       HC05      Rep1                1   \n",
       "33    hSDG99           hSDG99       HC06      Rep1                1   \n",
       "34    hSDG99           hSDG99       HC06      Rep1                1   \n",
       "35    hSDG99           hSDG99       HC06      Rep1                1   \n",
       "\n",
       "    repeated_sample    batch tissue       brain_region  source_RIN  ...   \n",
       "0                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...  \\\n",
       "1                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "2                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "3                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "4                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "5                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "6                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "7                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "8                 0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "9                 0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "10                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "11                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "12                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "13                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "14                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "15                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "16                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "17                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "18                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "19                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "20                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "21                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "22                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "23                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "24                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "25                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "26                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "27                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "28                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "29                0  Batch_3  Brain  Prefrontal Cortex         NaN  ...   \n",
       "30                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "31                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "32                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "33                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "34                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "35                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "\n",
       "    sex_ontology_term_id self_reported_ethnicity_ontology_term_id   \n",
       "0                    NaN                                      NaN  \\\n",
       "1                    NaN                                      NaN   \n",
       "2                    NaN                                      NaN   \n",
       "3                    NaN                                      NaN   \n",
       "4                    NaN                                      NaN   \n",
       "5                    NaN                                      NaN   \n",
       "6                    NaN                                      NaN   \n",
       "7                    NaN                                      NaN   \n",
       "8                    NaN                                      NaN   \n",
       "9                    NaN                                      NaN   \n",
       "10                   NaN                                      NaN   \n",
       "11                   NaN                                      NaN   \n",
       "12                   NaN                                      NaN   \n",
       "13                   NaN                                      NaN   \n",
       "14                   NaN                                      NaN   \n",
       "15                   NaN                                      NaN   \n",
       "16                   NaN                                      NaN   \n",
       "17                   NaN                                      NaN   \n",
       "18                   NaN                                      NaN   \n",
       "19                   NaN                                      NaN   \n",
       "20                   NaN                                      NaN   \n",
       "21                   NaN                                      NaN   \n",
       "22                   NaN                                      NaN   \n",
       "23                   NaN                                      NaN   \n",
       "24                   NaN                                      NaN   \n",
       "25                   NaN                                      NaN   \n",
       "26                   NaN                                      NaN   \n",
       "27                   NaN                                      NaN   \n",
       "28                   NaN                                      NaN   \n",
       "29                   NaN                                      NaN   \n",
       "30                   NaN                                      NaN   \n",
       "31                   NaN                                      NaN   \n",
       "32                   NaN                                      NaN   \n",
       "33                   NaN                                      NaN   \n",
       "34                   NaN                                      NaN   \n",
       "35                   NaN                                      NaN   \n",
       "\n",
       "   disease_ontology_term_id tissue_ontology_term_id   \n",
       "0                       NaN                     NaN  \\\n",
       "1                       NaN                     NaN   \n",
       "2                       NaN                     NaN   \n",
       "3                       NaN                     NaN   \n",
       "4                       NaN                     NaN   \n",
       "5                       NaN                     NaN   \n",
       "6                       NaN                     NaN   \n",
       "7                       NaN                     NaN   \n",
       "8                       NaN                     NaN   \n",
       "9                       NaN                     NaN   \n",
       "10                      NaN                     NaN   \n",
       "11                      NaN                     NaN   \n",
       "12                      NaN                     NaN   \n",
       "13                      NaN                     NaN   \n",
       "14                      NaN                     NaN   \n",
       "15                      NaN                     NaN   \n",
       "16                      NaN                     NaN   \n",
       "17                      NaN                     NaN   \n",
       "18                      NaN                     NaN   \n",
       "19                      NaN                     NaN   \n",
       "20                      NaN                     NaN   \n",
       "21                      NaN                     NaN   \n",
       "22                      NaN                     NaN   \n",
       "23                      NaN                     NaN   \n",
       "24                      NaN                     NaN   \n",
       "25                      NaN                     NaN   \n",
       "26                      NaN                     NaN   \n",
       "27                      NaN                     NaN   \n",
       "28                      NaN                     NaN   \n",
       "29                      NaN                     NaN   \n",
       "30                      NaN                     NaN   \n",
       "31                      NaN                     NaN   \n",
       "32                      NaN                     NaN   \n",
       "33                      NaN                     NaN   \n",
       "34                      NaN                     NaN   \n",
       "35                      NaN                     NaN   \n",
       "\n",
       "   cell_type_ontology_term_id assay_ontology_term_id suspension_type DV2000   \n",
       "0                         NaN                    NaN             NaN    NaN  \\\n",
       "1                         NaN                    NaN             NaN    NaN   \n",
       "2                         NaN                    NaN             NaN    NaN   \n",
       "3                         NaN                    NaN             NaN    NaN   \n",
       "4                         NaN                    NaN             NaN    NaN   \n",
       "5                         NaN                    NaN             NaN    NaN   \n",
       "6                         NaN                    NaN             NaN    NaN   \n",
       "7                         NaN                    NaN             NaN    NaN   \n",
       "8                         NaN                    NaN             NaN    NaN   \n",
       "9                         NaN                    NaN             NaN    NaN   \n",
       "10                        NaN                    NaN             NaN    NaN   \n",
       "11                        NaN                    NaN             NaN    NaN   \n",
       "12                        NaN                    NaN             NaN    NaN   \n",
       "13                        NaN                    NaN             NaN    NaN   \n",
       "14                        NaN                    NaN             NaN    NaN   \n",
       "15                        NaN                    NaN             NaN    NaN   \n",
       "16                        NaN                    NaN             NaN    NaN   \n",
       "17                        NaN                    NaN             NaN    NaN   \n",
       "18                        NaN                    NaN             NaN    NaN   \n",
       "19                        NaN                    NaN             NaN    NaN   \n",
       "20                        NaN                    NaN             NaN    NaN   \n",
       "21                        NaN                    NaN             NaN    NaN   \n",
       "22                        NaN                    NaN             NaN    NaN   \n",
       "23                        NaN                    NaN             NaN    NaN   \n",
       "24                        NaN                    NaN             NaN    NaN   \n",
       "25                        NaN                    NaN             NaN    NaN   \n",
       "26                        NaN                    NaN             NaN    NaN   \n",
       "27                        NaN                    NaN             NaN    NaN   \n",
       "28                        NaN                    NaN             NaN    NaN   \n",
       "29                        NaN                    NaN             NaN    NaN   \n",
       "30                        NaN                    NaN             NaN    NaN   \n",
       "31                        NaN                    NaN             NaN    NaN   \n",
       "32                        NaN                    NaN             NaN    NaN   \n",
       "33                        NaN                    NaN             NaN    NaN   \n",
       "34                        NaN                    NaN             NaN    NaN   \n",
       "35                        NaN                    NaN             NaN    NaN   \n",
       "\n",
       "   pm_PH  donor_id  \n",
       "0    NaN       NaN  \n",
       "1    NaN       NaN  \n",
       "2    NaN       NaN  \n",
       "3    NaN       NaN  \n",
       "4    NaN       NaN  \n",
       "5    NaN       NaN  \n",
       "6    NaN       NaN  \n",
       "7    NaN       NaN  \n",
       "8    NaN       NaN  \n",
       "9    NaN       NaN  \n",
       "10   NaN       NaN  \n",
       "11   NaN       NaN  \n",
       "12   NaN       NaN  \n",
       "13   NaN       NaN  \n",
       "14   NaN       NaN  \n",
       "15   NaN       NaN  \n",
       "16   NaN       NaN  \n",
       "17   NaN       NaN  \n",
       "18   NaN       NaN  \n",
       "19   NaN       NaN  \n",
       "20   NaN       NaN  \n",
       "21   NaN       NaN  \n",
       "22   NaN       NaN  \n",
       "23   NaN       NaN  \n",
       "24   NaN       NaN  \n",
       "25   NaN       NaN  \n",
       "26   NaN       NaN  \n",
       "27   NaN       NaN  \n",
       "28   NaN       NaN  \n",
       "29   NaN       NaN  \n",
       "30   NaN       NaN  \n",
       "31   NaN       NaN  \n",
       "32   NaN       NaN  \n",
       "33   NaN       NaN  \n",
       "34   NaN       NaN  \n",
       "35   NaN       NaN  \n",
       "\n",
       "[36 rows x 42 columns]"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix replicate & replicate_count\n",
    "SAMPLE.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "SAMPLE['replicate'] = \"Rep1\"\n",
    "SAMPLE['replicate_count'] = 1\n",
    "\n",
    "SAMPLE.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "STUDY = force_enum_string(STUDY, \"STUDY\", CDE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in STUDY.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in STUDY.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing the function with STUDY.csv and CDE.csv\n",
    "validate_table(STUDY, \"STUDY\", CDE)\n",
    "\n",
    "\n",
    "# Need to add contributor_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in PROTOCOL.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in PROTOCOL.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing the function with PROTOCOL.csv and CDE.csv\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SUBJECT.\n",
      "No empty or NaN values found in Required fields.\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- primary_diagnosis_text: 12 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- sex: F, M\n",
      "\t\t\t- race: B, W\n",
      "\t\t\t- primary_diagnosis: Normal control, Idiopathic Parkinson's disease\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "# Testing the function with SUBJECT.csv and CDE.csv\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SUBJECT = force_enum_string(SUBJECT, \"SUBJECT\", CDE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SUBJECT.\n",
      "No empty or NaN values found in Required fields.\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- primary_diagnosis_text: 12 rows\n",
      "\tAll Enum fields have valid values in SUBJECT.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT['sex'] = SUBJECT['sex'].replace({'F':\"Female\", 'M':\"Male\"})\n",
    "SUBJECT['race'] = SUBJECT['race'].replace({'W':\"White\", 'B':\"Black or African American\"})\n",
    "\n",
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace({'Normal control':\"Healthy Control\", \"Idiopathic Parkinson's disease\":\"Idiopathic PD\"})\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SAMPLE.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- source_RIN: 36 rows\n",
      "\t\t\t- RIN: 36 rows\n",
      "\t\t\t- time: 36 rows\n",
      "\t\t\t- suspension_type: 36 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- pm_PH: 36 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- assay: v3.1 - Single Index, 10x Genomics \n",
      "\t\t\t- sequencing_length: 150bp x2\n",
      "\t\t\t- organism_ontology_term_id: Nan\n",
      "\t\t\t- sex_ontology_term_id: Nan\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.replace(\"Nan\", \"\", inplace=True)\n",
    "SAMPLE.replace(\"nan\", \"\", inplace=True)\n",
    "\n",
    "# Testing the function with SAMPLE.csv and CDE.csv\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "# sequence length will need to be converted to a string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = force_enum_string(SAMPLE, \"SAMPLE\", CDE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "# set time == 0 for all samples\n",
    "SAMPLE['time'] = 0\n",
    "\n",
    "SAMPLE['file_type'] = SAMPLE['file_type'].replace({\"Fastq\":\"fastq\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# need to join with subject to get \"sex\" and convert to ontology term\n",
    "SAMPLE_SUBJECT = SAMPLE.merge(SUBJECT, on='subject_id',  how='left')\n",
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['sex_ontology_term_id'] = SAMPLE_SUBJECT['sex'].replace({\"Male\":\"PATO:0000384 (male)\", \"Female\":\"PATO:0000383 (female)\" })\n",
    "\n",
    "# ignore development_stage_ontology_term_id, self_reported_ethnicity_ontology_term_id, assay_ontology_term_id, etc for now. (Check wiht Le)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix assay\n",
    "SAMPLE['assay'] = SAMPLE['assay'].replace({'v3.1 - Single Index, 10x Genomics ':\"v3.1 - Single Index\"})\n",
    "# fix assay\n",
    "SAMPLE['sequencing_length'] = SAMPLE['sequencing_length'].replace({'150bp x2':\"150\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SAMPLE.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- source_RIN: 36 rows\n",
      "\t\t\t- RIN: 36 rows\n",
      "\t\t\t- suspension_type: 36 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- pm_PH: 36 rows\n",
      "\tAll Enum fields have valid values in SAMPLE.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validate_table(SAMPLE, \"SAMPLE\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tMissing Required Fields in CLINPATH: path_thal\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- age_at_onset: 12 rows\n",
      "\t\t\t- age_at_diagnosis: 12 rows\n",
      "\t\t\t- first_motor_symptom: 12 rows\n",
      "\t\t\t- path_year_death: 12 rows\n",
      "\t\t\t- brain_weight: 12 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- smoking_years: 12 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- region_level_2: Prefrontal cortex\n",
      "\t\t\t- hx_dementia_mci: Nan\n",
      "\t\t\t- hx_melanoma: Nan\n",
      "\t\t\t- education_level: Nan\n",
      "\t\t\t- smoking_status: Nan\n",
      "\t\t\t- APOE_e4_status: 3,3, 2,3\n",
      "\t\t\t- path_autopsy_dx_main: Nan\n",
      "\t\t\t- path_braak_nft: Nan\n",
      "\t\t\t- path_braak_asyn: Nan\n",
      "\t\t\t- path_cerad: Nan\n",
      "\t\t\t- known_pathogenic_mutation: Nan\n",
      "\t\t\t- path_mckeith: Nan\n",
      "\t\t\t- sn_neuronal_loss: Nan\n",
      "\t\t\t- path_infarcs: Nan\n",
      "\t\t\t- path_nia_ri: Nan\n",
      "\t\t\t- path_nia_aa_a: Nan\n",
      "\t\t\t- path_nia_aa_b: Nan\n",
      "\t\t\t- path_nia_aa_c: Nan\n",
      "\t\t\t- TDP43: Nan\n",
      "\t\t\t- arteriolosclerosis_severity_scale: Nan\n",
      "\t\t\t- amyloid_angiopathy_severity_scale: Nan\n",
      "\t\t\t- path_ad_level: Nan\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH.replace(\"Nan\", \"\", inplace=True)\n",
    "CLINPATH.replace(\"nan\", \"\", inplace=True)\n",
    "\n",
    "# Testing the function with CLINPATH.csv and CDE.csv\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH = force_enum_string(CLINPATH, \"CLINPATH\", CDE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [],
   "source": [
    "# redact \"Prefrontal Cortex\" from region_level_2 for now\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace({'Prefrontal Cortex':\"\"})\n",
    "\n",
    "# leave te APOE_e4_status as is for now . multiple are coded as \"2,3\" \n",
    "# but remove commas\n",
    "CLINPATH[\"APOE_e4_status\"] = CLINPATH[\"APOE_e4_status\"].str.replace(\",\",\"\")\n",
    "\n",
    "# need to fix the path_autopsy_dx_main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tMissing Required Fields in CLINPATH: path_thal\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- age_at_onset: 12 rows\n",
      "\t\t\t- age_at_diagnosis: 12 rows\n",
      "\t\t\t- first_motor_symptom: 12 rows\n",
      "\t\t\t- path_year_death: 12 rows\n",
      "\t\t\t- brain_weight: 12 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- smoking_years: 12 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- region_level_2: Prefrontal cortex\n",
      "\t\t\t- hx_dementia_mci: Nan\n",
      "\t\t\t- hx_melanoma: Nan\n",
      "\t\t\t- education_level: Nan\n",
      "\t\t\t- smoking_status: Nan\n",
      "\t\t\t- path_autopsy_dx_main: Nan\n",
      "\t\t\t- path_braak_nft: Nan\n",
      "\t\t\t- path_braak_asyn: Nan\n",
      "\t\t\t- path_cerad: Nan\n",
      "\t\t\t- known_pathogenic_mutation: Nan\n",
      "\t\t\t- path_mckeith: Nan\n",
      "\t\t\t- sn_neuronal_loss: Nan\n",
      "\t\t\t- path_infarcs: Nan\n",
      "\t\t\t- path_nia_ri: Nan\n",
      "\t\t\t- path_nia_aa_a: Nan\n",
      "\t\t\t- path_nia_aa_b: Nan\n",
      "\t\t\t- path_nia_aa_c: Nan\n",
      "\t\t\t- TDP43: Nan\n",
      "\t\t\t- arteriolosclerosis_severity_scale: Nan\n",
      "\t\t\t- amyloid_angiopathy_severity_scale: Nan\n",
      "\t\t\t- path_ad_level: Nan\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>sn_neuronal_loss</th>\n",
       "      <th>path_infarcs</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HC01</td>\n",
       "      <td>HSDG07</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HC02</td>\n",
       "      <td>HSDG13</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC03</td>\n",
       "      <td>HSDG101</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC04</td>\n",
       "      <td>HSDG10</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC05</td>\n",
       "      <td>HSDG30</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Frontal lobe</td>\n",
       "      <td>Prefrontal cortex</td>\n",
       "      <td>Grey matter</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id source_sample_id time_from_baseline GP2_id hemisphere   \n",
       "0      HC01           HSDG07                Nan    Nan    Unknown  \\\n",
       "1      HC02           HSDG13                Nan    Nan    Unknown   \n",
       "2      HC03          HSDG101                Nan    Nan    Unknown   \n",
       "3      HC04           HSDG10                Nan    Nan    Unknown   \n",
       "4      HC05           HSDG30                Nan    Nan    Unknown   \n",
       "\n",
       "  region_level_1     region_level_2 region_level_3 AMPPD_id family_history   \n",
       "0   Frontal lobe  Prefrontal cortex    Grey matter      Nan        Unknown  \\\n",
       "1   Frontal lobe  Prefrontal cortex    Grey matter      Nan        Unknown   \n",
       "2   Frontal lobe  Prefrontal cortex    Grey matter      Nan        Unknown   \n",
       "3   Frontal lobe  Prefrontal cortex    Grey matter      Nan        Unknown   \n",
       "4   Frontal lobe  Prefrontal cortex    Grey matter      Nan        Unknown   \n",
       "\n",
       "   ... sn_neuronal_loss  path_infarcs  path_nia_ri  path_nia_aa_a   \n",
       "0  ...              Nan           Nan          Nan            Nan  \\\n",
       "1  ...              Nan           Nan          Nan            Nan   \n",
       "2  ...              Nan           Nan          Nan            Nan   \n",
       "3  ...              Nan           Nan          Nan            Nan   \n",
       "4  ...              Nan           Nan          Nan            Nan   \n",
       "\n",
       "  path_nia_aa_b path_nia_aa_c TDP43 arteriolosclerosis_severity_scale   \n",
       "0           Nan           Nan   Nan                               Nan  \\\n",
       "1           Nan           Nan   Nan                               Nan   \n",
       "2           Nan           Nan   Nan                               Nan   \n",
       "3           Nan           Nan   Nan                               Nan   \n",
       "4           Nan           Nan   Nan                               Nan   \n",
       "\n",
       "   amyloid_angiopathy_severity_scale path_ad_level  \n",
       "0                                Nan           Nan  \n",
       "1                                Nan           Nan  \n",
       "2                                Nan           Nan  \n",
       "3                                Nan           Nan  \n",
       "4                                Nan           Nan  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validate_table(CLINPATH, \"CLINPATH\", CDE)\n",
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = data_path / \"team-hafler\"\n",
    "data_path = Path.home() / (\"Projects/ASAP/team-hafler\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the column order\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE)\n",
    "\n",
    "# write the clean metadata\n",
    "STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Hafler\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY.to_csv( metadata_path / \"STUDY.csv\")\n",
    "CLINPATH.to_csv( metadata_path / \"CLINPATH.csv\")\n",
    "SUBJECT.to_csv( metadata_path / \"SUBJECT.csv\")\n",
    "SAMPLE.to_csv( metadata_path / \"SAMPLE.csv\")\n",
    "PROTOCOL.to_csv( metadata_path / \"PROTOCOL.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[\"22\", \"23\",\"24\",\"33\", \"34\", \"44\", \"Unknown\"]'], dtype=object)"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH[\"APOE_e4_status\"].unique()\n",
    "\n",
    "CDE[CDE[\"Field\"]== \"APOE_e4_status\"].Validation.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "export_root = Path.cwd() / \"clean/team-Lee\"\n",
    "SAMPLE_lee = pd.read_csv(export_root / \"SAMPLE.csv\")\n",
    "n_samples_lee = SAMPLE_lee[\"sample_id\"].nunique()\n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Hafler\"\n",
    "SAMPLE_hafler = pd.read_csv(export_root / \"SAMPLE.csv\")\n",
    "n_samples_hafler = SAMPLE_hafler[\"sample_id\"].nunique()\n",
    "\n",
    "\n",
    "total_samples = n_samples_lee + n_samples_hafler\n",
    "total_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-hardy\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\")\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\")\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tMissing Required Fields in STUDY: project_name, project_dataset, project_description, ASAP_team_name, ASAP_lab_name, PI_full_name, PI_email, contributor_names, submitter_name, submittor_email, ASAP_grant_id, other_funding_source, publication_DOI, publication_PMID, number_of_brain_samples, brain_regions, types_of_samples, DUA_version\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in STUDY.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>project_description</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Ryten Lab</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>Mina Ryten</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PI_email</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>contributor_names</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>submitter_name</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>submitter_email</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ASAP_grant_id</td>\n",
       "      <td>ASAP-000478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>brain_regions</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DUA_version</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>metadata_version_date</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       name                                              value\n",
       "0              project_name  Understanding mechanisms of Parkinson's diseas...\n",
       "1           project_dataset                                    Hardy snRNA-seq\n",
       "2       project_description  Genetic analysis has identified many risk gene...\n",
       "3            ASAP_team_name                                         TEAM-HARDY\n",
       "4             ASAP_lab_name                                          Ryten Lab\n",
       "5              PI_full_name                                         Mina Ryten\n",
       "6                  PI_email                               mina.ryten@ucl.ac.uk\n",
       "7         contributor_names  Aine Fairbrother-Browne, Jonathan Brenton, Mel...\n",
       "8            submitter_name                            Aine Fairbrother-Browne\n",
       "9           submitter_email               aine.fairbrother-browne.18@ucl.ac.uk\n",
       "10            ASAP_grant_id                                        ASAP-000478\n",
       "11     other_funding_source                                                NaN\n",
       "12          publication_DOI                                                NaN\n",
       "13         publication_PMID                                                NaN\n",
       "14  number_of_brain_samples                                                128\n",
       "15            brain_regions  Inferior Parietal Lobule (IPL), Anterior Cingu...\n",
       "16         types_of_samples  Late stage (Braak 5-6) PD and control post-mor...\n",
       "17                PI_ORCHID                                0000-0001-9520-6957\n",
       "18     PI_google_scholar_id  https://scholar.google.co.uk/citations?user=lt...\n",
       "19              DUA_version                                                NaN\n",
       "20    metadata_version_date                                 Version 1, 09/2023"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "validate_table(STUDY, \"STUDY\", CDE)\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>babom_ACG</td>\n",
       "      <td>P2/14</td>\n",
       "      <td>babom</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>B1</td>\n",
       "      <td>brain</td>\n",
       "      <td>ACG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0009835</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3611</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3612</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3613</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3614</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3615</th>\n",
       "      <td>zupam_IPL</td>\n",
       "      <td>P78/11</td>\n",
       "      <td>zupam</td>\n",
       "      <td>rep2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>B2</td>\n",
       "      <td>brain</td>\n",
       "      <td>IPL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0006088</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO:0008913</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3616 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample_id source_subject_id subject_id replicate  replicate_count   \n",
       "0     babom_ACG             P2/14      babom       NaN                1  \\\n",
       "1     babom_ACG             P2/14      babom       NaN                1   \n",
       "2     babom_ACG             P2/14      babom       NaN                1   \n",
       "3     babom_ACG             P2/14      babom       NaN                1   \n",
       "4     babom_ACG             P2/14      babom       NaN                1   \n",
       "...         ...               ...        ...       ...              ...   \n",
       "3611  zupam_IPL            P78/11      zupam      rep2                2   \n",
       "3612  zupam_IPL            P78/11      zupam      rep2                2   \n",
       "3613  zupam_IPL            P78/11      zupam      rep2                2   \n",
       "3614  zupam_IPL            P78/11      zupam      rep2                2   \n",
       "3615  zupam_IPL            P78/11      zupam      rep2                2   \n",
       "\n",
       "      repeated_sample batch tissue brain_region  source_RIN  ...   \n",
       "0                   0    B1  brain          ACG         NaN  ...  \\\n",
       "1                   0    B1  brain          ACG         NaN  ...   \n",
       "2                   0    B1  brain          ACG         NaN  ...   \n",
       "3                   0    B1  brain          ACG         NaN  ...   \n",
       "4                   0    B1  brain          ACG         NaN  ...   \n",
       "...               ...   ...    ...          ...         ...  ...   \n",
       "3611                1    B2  brain          IPL         NaN  ...   \n",
       "3612                1    B2  brain          IPL         NaN  ...   \n",
       "3613                1    B2  brain          IPL         NaN  ...   \n",
       "3614                1    B2  brain          IPL         NaN  ...   \n",
       "3615                1    B2  brain          IPL         NaN  ...   \n",
       "\n",
       "       sex_ontology_term_id self_reported_ethnicity_ontology_term_id   \n",
       "0     PATO:0000383 (female)                                      NaN  \\\n",
       "1     PATO:0000383 (female)                                      NaN   \n",
       "2     PATO:0000383 (female)                                      NaN   \n",
       "3     PATO:0000383 (female)                                      NaN   \n",
       "4     PATO:0000383 (female)                                      NaN   \n",
       "...                     ...                                      ...   \n",
       "3611  PATO:0000383 (female)                                      NaN   \n",
       "3612  PATO:0000383 (female)                                      NaN   \n",
       "3613  PATO:0000383 (female)                                      NaN   \n",
       "3614  PATO:0000383 (female)                                      NaN   \n",
       "3615  PATO:0000383 (female)                                      NaN   \n",
       "\n",
       "      disease_ontology_term_id tissue_ontology_term_id   \n",
       "0                MONDO:0005180          UBERON:0009835  \\\n",
       "1                MONDO:0005180          UBERON:0009835   \n",
       "2                MONDO:0005180          UBERON:0009835   \n",
       "3                MONDO:0005180          UBERON:0009835   \n",
       "4                MONDO:0005180          UBERON:0009835   \n",
       "...                        ...                     ...   \n",
       "3611             MONDO:0005180          UBERON:0006088   \n",
       "3612             MONDO:0005180          UBERON:0006088   \n",
       "3613             MONDO:0005180          UBERON:0006088   \n",
       "3614             MONDO:0005180          UBERON:0006088   \n",
       "3615             MONDO:0005180          UBERON:0006088   \n",
       "\n",
       "     cell_type_ontology_term_id  assay_ontology_term_id suspension_type   \n",
       "0                           NaN             EFO:0008913         nucleus  \\\n",
       "1                           NaN             EFO:0008913         nucleus   \n",
       "2                           NaN             EFO:0008913         nucleus   \n",
       "3                           NaN             EFO:0008913         nucleus   \n",
       "4                           NaN             EFO:0008913         nucleus   \n",
       "...                         ...                     ...             ...   \n",
       "3611                        NaN             EFO:0008913         nucleus   \n",
       "3612                        NaN             EFO:0008913         nucleus   \n",
       "3613                        NaN             EFO:0008913         nucleus   \n",
       "3614                        NaN             EFO:0008913         nucleus   \n",
       "3615                        NaN             EFO:0008913         nucleus   \n",
       "\n",
       "     DV2000 pm_PH donor_id  \n",
       "0       NaN   NaN      NaN  \n",
       "1       NaN   NaN      NaN  \n",
       "2       NaN   NaN      NaN  \n",
       "3       NaN   NaN      NaN  \n",
       "4       NaN   NaN      NaN  \n",
       "...     ...   ...      ...  \n",
       "3611    NaN   NaN      NaN  \n",
       "3612    NaN   NaN      NaN  \n",
       "3613    NaN   NaN      NaN  \n",
       "3614    NaN   NaN      NaN  \n",
       "3615    NaN   NaN      NaN  \n",
       "\n",
       "[3616 rows x 42 columns]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there seems to be something funky with SAMPLE\n",
    "# SAMPLE = SAMPLE[SAMPLE[\"batch\"]==\"B1\"]\n",
    "# SAMPLE.drop_duplicates(inplace=True) #, subset=[ \"file_name\"])\n",
    "SAMPLE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>project_description</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Ryten Lab</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name                                              value\n",
       "0         project_name  Understanding mechanisms of Parkinson's diseas...\n",
       "1      project_dataset                                    Hardy snRNA-seq\n",
       "2  project_description  Genetic analysis has identified many risk gene...\n",
       "3       ASAP_team_name                                         TEAM-HARDY\n",
       "4        ASAP_lab_name                                          Ryten Lab"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>project_description</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>PI_email</td>\n",
       "      <td>contributor_names</td>\n",
       "      <td>submitter_name</td>\n",
       "      <td>submitter_email</td>\n",
       "      <td>...</td>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>brain_regions</td>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>DUA_version</td>\n",
       "      <td>metadata_version_date</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  0                1    \n",
       "0                                       project_name  project_dataset  \\\n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "                                                  2               3    \n",
       "0                                project_description  ASAP_team_name  \\\n",
       "1  Genetic analysis has identified many risk gene...      TEAM-HARDY   \n",
       "\n",
       "              4             5                     6    \n",
       "0  ASAP_lab_name  PI_full_name              PI_email  \\\n",
       "1      Ryten Lab    Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "                                                  7                        8    \n",
       "0                                  contributor_names           submitter_name  \\\n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "                                     9   ...                    11   \n",
       "0                       submitter_email  ...  other_funding_source  \\\n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                   NaN   \n",
       "\n",
       "                12                13                       14   \n",
       "0  publication_DOI  publication_PMID  number_of_brain_samples  \\\n",
       "1              NaN               NaN                      128   \n",
       "\n",
       "                                                  15   \n",
       "0                                      brain_regions  \\\n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "                                                  16                   17   \n",
       "0                                   types_of_samples            PI_ORCHID  \\\n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "                                                  18           19   \n",
       "0                               PI_google_scholar_id  DUA_version  \\\n",
       "1  https://scholar.google.co.uk/citations?user=lt...          NaN   \n",
       "\n",
       "                      20  \n",
       "0  metadata_version_date  \n",
       "1     Version 1, 09/2023  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submitter_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  project_dataset   \n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq  \\\n",
       "\n",
       "0                                project_description ASAP_team_name   \n",
       "1  Genetic analysis has identified many risk gene...     TEAM-HARDY  \\\n",
       "\n",
       "0 ASAP_lab_name PI_full_name              PI_email   \n",
       "1     Ryten Lab   Mina Ryten  mina.ryten@ucl.ac.uk  \\\n",
       "\n",
       "0                                  contributor_names           submitter_name   \n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne  \\\n",
       "\n",
       "0                       submitter_email  ... other_funding_source   \n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                  NaN  \\\n",
       "\n",
       "0 publication_DOI publication_PMID number_of_brain_samples   \n",
       "1             NaN              NaN                     128  \\\n",
       "\n",
       "0                                      brain_regions   \n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...  \\\n",
       "\n",
       "0                                   types_of_samples            PI_ORCHID   \n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957  \\\n",
       "\n",
       "0                               PI_google_scholar_id DUA_version   \n",
       "1  https://scholar.google.co.uk/citations?user=lt...         NaN  \\\n",
       "\n",
       "0 metadata_version_date  \n",
       "1    Version 1, 09/2023  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in STUDY.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in STUDY.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 251,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "# fix the column order\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "validate_table(STUDY, \"STUDY\", CDE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sample_collection_summary</td>\n",
       "      <td>This dataset contains cortical regions only, p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cell_extraction_summary</td>\n",
       "      <td>From protocols.io: This protocol is used to is...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lib_prep_summary</td>\n",
       "      <td>'Nuclei were extracted from homogenised post-m...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>data_processing_summary</td>\n",
       "      <td>Cell ranger was used to convert raw sequencing...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>github_url</td>\n",
       "      <td>Raw to fastq to mapped: https://github.com/RHR...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        name   \n",
       "0  sample_collection_summary  \\\n",
       "1    cell_extraction_summary   \n",
       "2           lib_prep_summary   \n",
       "3    data_processing_summary   \n",
       "4                 github_url   \n",
       "\n",
       "                                               value  \n",
       "0  This dataset contains cortical regions only, p...  \n",
       "1  From protocols.io: This protocol is used to is...  \n",
       "2  'Nuclei were extracted from homogenised post-m...  \n",
       "3  Cell ranger was used to convert raw sequencing...  \n",
       "4  Raw to fastq to mapped: https://github.com/RHR...  "
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "PROTOCOL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_collection_summary</th>\n",
       "      <th>cell_extraction_summary</th>\n",
       "      <th>lib_prep_summary</th>\n",
       "      <th>data_processing_summary</th>\n",
       "      <th>github_url</th>\n",
       "      <th>protocols_io_DOI</th>\n",
       "      <th>other_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>This dataset contains cortical regions only, p...</td>\n",
       "      <td>From protocols.io: This protocol is used to is...</td>\n",
       "      <td>'Nuclei were extracted from homogenised post-m...</td>\n",
       "      <td>Cell ranger was used to convert raw sequencing...</td>\n",
       "      <td>Raw to fastq to mapped: https://github.com/RHR...</td>\n",
       "      <td>Nuclear extraction protocol: 10.17504/protocol...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0                          sample_collection_summary   \n",
       "1  This dataset contains cortical regions only, p...  \\\n",
       "\n",
       "0                            cell_extraction_summary   \n",
       "1  From protocols.io: This protocol is used to is...  \\\n",
       "\n",
       "0                                   lib_prep_summary   \n",
       "1  'Nuclei were extracted from homogenised post-m...  \\\n",
       "\n",
       "0                            data_processing_summary   \n",
       "1  Cell ranger was used to convert raw sequencing...  \\\n",
       "\n",
       "0                                         github_url   \n",
       "1  Raw to fastq to mapped: https://github.com/RHR...  \\\n",
       "\n",
       "0                                   protocols_io_DOI other_reference  \n",
       "1  Nuclear extraction protocol: 10.17504/protocol...             NaN  "
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = PROTOCOL[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "PROTOCOL = tmp.drop([0])\n",
    "PROTOCOL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in PROTOCOL.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in PROTOCOL.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix the column order\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SUBJECT.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- ethnicity: 64 rows\n",
      "\t\t\t- duration_pmi: 1 rows\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- race: Nan\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 255,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)\n",
    "\n",
    "# Testing the function with SUBJECT.csv and CDE.csv\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "SUBJECT.replace(\"nan\", \"\", inplace=True)\n",
    "SUBJECT.fillna(\"\", inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 257,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in SAMPLE.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- source_RIN: 3616 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- pm_PH: 3616 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- sequencing_length: 190\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 257,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "# allos sequence_length == 190 for now\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE)\n",
    "\n",
    "\n",
    "# add 'replicate' coding (nans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 258,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE.replace(\"nan\", \"\", inplace=True)\n",
    "# sequence length will need to be converted to a string\n",
    "SAMPLE.fillna(\"\", inplace=True)\n",
    "# Testing the function with SAMPLE.csv and CDE.csv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in CLINPATH.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- age_at_onset: 138 rows\n",
      "\t\t\t- age_at_diagnosis: 10 rows\n",
      "\t\t\t- first_motor_symptom: 138 rows\n",
      "\t\t\t- path_year_death: 138 rows\n",
      "\t\t\t- brain_weight: 138 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- smoking_years: 138 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- family_history: Nan\n",
      "\t\t\t- hx_dementia_mci: Nan\n",
      "\t\t\t- hx_melanoma: Nan\n",
      "\t\t\t- education_level: Nan\n",
      "\t\t\t- smoking_status: Nan\n",
      "\t\t\t- APOE_e4_status: Nan\n",
      "\t\t\t- cognitive_status: Nan\n",
      "\t\t\t- path_autopsy_dx_main: Control brain, Pathological ageing, Control brain / Path ageing, Argyrophilic grain disease, Control brain, Cerebrovascular disease (small vessel), Cerebrovascular disease (small vessel), Control brain, Alzheimer`s disease (intermediate level AD pathological change), Control brain / Path ageing, CAA, Nan\n",
      "\t\t\t- path_braak_nft: 2.0, 1.0, 3.0, 0.0, 4.0, 6.0, Nan\n",
      "\t\t\t- path_braak_asyn: 6.0, 0.0, 5.0, Nan\n",
      "\t\t\t- path_cerad: Nan\n",
      "\t\t\t- path_thal: At least 4, Nan\n",
      "\t\t\t- known_pathogenic_mutation: Nan\n",
      "\t\t\t- path_mckeith: Diffuse neocortical, Limbic transitional, Diffuse Neocortical, Nan\n",
      "\t\t\t- sn_neuronal_loss: Nan\n",
      "\t\t\t- path_infarcs: Nan\n",
      "\t\t\t- path_nia_ri: Nan\n",
      "\t\t\t- path_nia_aa_a: 1.0, 2.0, 0.0, 3.0, Nan\n",
      "\t\t\t- path_nia_aa_b: 1.0, 2.0, 0.0, 3.0, Nan\n",
      "\t\t\t- path_nia_aa_c: 0.0, 1.0, 2.0, 3.0, Nan\n",
      "\t\t\t- TDP43: Nan\n",
      "\t\t\t- arteriolosclerosis_severity_scale: Nan\n",
      "\t\t\t- amyloid_angiopathy_severity_scale: Nan\n",
      "\t\t\t- path_ad_level: Nan, No evidence\n",
      "\t\t\t- dig_slide_avail: Yes\n",
      "\t\t\t- quant_path_avail: Yes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 259,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE)\n",
    "\n",
    "\n",
    "# Testing the function with CLINPATH.csv and CDE.csv\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hack to make sure the replace below works.\n",
    "CLINPATH.replace(\"\", pd.NA, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace 'path_braak_asyn' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_braak_asyn'] = CLINPATH['path_braak_asyn'].apply(lambda val: str(int(val)) if pd.notna(val) else \"\")\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_braak_nft'] = CLINPATH['path_braak_nft'].apply(lambda val: str(int(val)) if pd.notna(val) else \"\").replace({\"0\":\"0\", \n",
    "                                                                                                                           \"1\":\"I\", \n",
    "                                                                                                                           \"2\": \"II\", \n",
    "                                                                                                                           \"3\":\"III\", \n",
    "                                                                                                                           \"4\":\"IV\", \n",
    "                                                                                                                           \"5\":\"V\", \n",
    "                                                                                                                           \"6\":\"VI\"})\n",
    "\n",
    "# code family_history as \"Not Reported\" (currently empty)\n",
    "CLINPATH['family_history'] = \"Not Reported\"\n",
    "\n",
    "\n",
    "\n",
    "# check APOE_e4_status ? currently empty\n",
    "\n",
    "# `path_autopsy_dx_main`  actually seems good parser might be wrong\n",
    "\n",
    "# code \"at least 4\" as \"4/5\" \n",
    "\n",
    "CLINPATH['path_thal'] = CLINPATH['path_thal'].replace({'At least 4':\"4/5\"})\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace({'Diffuse neocortical': \"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\", \n",
    "                                                        'Limbic transitional': \"Limbic (transitional)\" ,\n",
    "                                                        'Diffuse Neocortical':\"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\"})\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_a'] = CLINPATH['path_nia_aa_a'].apply(lambda val: str(int(val)) if pd.notna(val) else \"\")\n",
    "\n",
    "CLINPATH['path_nia_aa_a'] = CLINPATH['path_nia_aa_a'].replace({\"0\":\"A0\", \n",
    "                                                                                                                           \"1\":\"A1\", \n",
    "                                                                                                                           \"2\": \"A2\", \n",
    "                                                                                                                           \"3\":\"A3\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_b'] = CLINPATH['path_nia_aa_b'].apply(lambda val: str(int(val)) if pd.notna(val) else \"\").replace({\"0\":\"B0\", \n",
    "                                                                                                                           \"1\":\"B1\", \n",
    "                                                                                                                           \"2\": \"B2\", \n",
    "                                                                                                                           \"3\":\"B3\"})\n",
    "\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_c'] = CLINPATH['path_nia_aa_c'].apply(lambda val: str(int(val)) if pd.notna(val) else \"\").replace({\"0\":\"C0\", \n",
    "                                                                                                                           \"1\":\"C1\", \n",
    "                                                                                                                           \"2\": \"C2\", \n",
    "                                                                                                                           \"3\":\"C3\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "CLINPATH['path_ad_level'] = CLINPATH['path_ad_level'].replace({\"No evidence\": \"No evidence of Alzheimer\\'s disease neuropathological change\"})\n",
    "\n",
    "\n",
    "# empty 'hx_dementia_mci', 'hx_melanoma', 'education_level', 'cognitive_status'\n",
    "#  coded as nan will be fixed with .fillna(\"\") below\n",
    "CLINPATH.replace(\"nan\", \"\", inplace=True)\n",
    "CLINPATH.fillna(\"\", inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array(['[\"Yes, No\"]'], dtype=object), array(['Yes'], dtype=object))"
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDE[CDE[\"Field\"]==\"dig_slide_avail\"].Validation.unique(), CLINPATH['dig_slide_avail'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in CLINPATH.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- path_autopsy_dx_main: Control brain, Pathological ageing, Control brain / Path ageing, Argyrophilic grain disease, Control brain, Cerebrovascular disease (small vessel), Cerebrovascular disease (small vessel), Control brain, Alzheimer`s disease (intermediate level AD pathological change), Control brain / Path ageing, CAA, \n",
      "\t\t\t- path_nia_aa_a: A0, \n",
      "\t\t\t- path_nia_aa_b: B0, \n",
      "\t\t\t- path_nia_aa_c: C0, \n",
      "\t\t\t- dig_slide_avail: Yes\n",
      "\t\t\t- quant_path_avail: Yes\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Testing the function with CLINPATH.csv and CDE.csv\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# # write the clean metadata\n",
    "# STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "# PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "# CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "# SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "# SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# # also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Hardy\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "basically hold the list of the GP2ID and the original clinical ID pairs + how many samples are in the GP2 (s1 only or s1, s2,...) for all GP2 submitted individuals. It takes the sample manifest, scan the clinical ID to check if this is the additional submission of those already in the GP2 or not and then if its new, give new GP2ID and GP2sampleID. If the clinical_id is already existing in the GP2 then only provide GP2sampleID (GP2ID_sX+1). Also it errors if the original sample ID submitted is equal to the one in the list. (No duplication of sample ID from the same cohort)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "team_names = [\"team-lee\", \"team-hafler\", \"team-hardy\", \"team-jakobsson\", \"team-sherzer\",\"team-sulzer\", \"tam-voet\",\"team-wood\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "team_code = [\"LEE\", \"HAF\", \"HAR\", \"JAK\", \"SHE\", \"SUL\", \"VOE\", \"WOO\"]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TEAM-LEE',\n",
       " 'TEAM-HAFLER',\n",
       " 'TEAM-HARDY',\n",
       " 'TEAM-JAKOBSSON',\n",
       " 'TEAM-SHERZER',\n",
       " 'TEAM-SULZER',\n",
       " 'TAM-VOET',\n",
       " 'TEAM-WOOD']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.upper() for x in team_names]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    uids = [str(id) for id in df_nodups['sample_id'].unique()]\n",
    "    mapid = {}\n",
    "    for uid in uids:\n",
    "        mapid[uid]= n\n",
    "        n += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def getgp2idsv2(dfproc, n, study_code):\n",
    "    df_dups = dfproc[dfproc.duplicated(keep=False, subset=['clinical_id'])].sort_values('clinical_id').reset_index(drop = True).copy()\n",
    "    if df_dups.shape[0]>0:\n",
    "        dupids_mapper = dict(zip(df_dups.clinical_id.unique(),\n",
    "                            [num+n for num in range(len(df_dups.clinical_id.unique()))]))\n",
    "        \n",
    "        df_dup_chunks = []\n",
    "        for clin_id, gp2id in dupids_mapper.items():\n",
    "            df_dups_subset = df_dups[df_dups.clinical_id==clin_id].copy()\n",
    "            df_dups_subset['GP2ID'] = [f'{study_code}_{gp2id:06}' for i in range(df_dups_subset.shape[0])]\n",
    "            df_dups_subset['SampleRepNo'] = ['s'+str(i+1) for i in range(df_dups_subset.shape[0])]\n",
    "            df_dups_subset['GP2sampleID'] = df_dups_subset['GP2ID'] + '_' + df_dups_subset['SampleRepNo']\n",
    "            df_dup_chunks.append(df_dups_subset)\n",
    "        df_dups_wids = pd.concat(df_dup_chunks)\n",
    "\n",
    "    df_nodups = dfproc[~dfproc.duplicated(keep=False, subset=['clinical_id'])].sort_values('clinical_id').reset_index(drop = True).copy()\n",
    "\n",
    "    if df_dups.shape[0]>0:\n",
    "        n =  len(list(dupids_mapper.values())) + n\n",
    "    else:\n",
    "        n = n\n",
    "\n",
    "    uids = [str(id) for id in df_nodups['sample_id'].unique()]\n",
    "    mapid = {}\n",
    "    for uid in uids:\n",
    "        mapid[uid]= n\n",
    "        n += 1\n",
    "    df_nodups_wids = df_nodups.copy()\n",
    "    df_nodups_wids['uid_idx'] = df_nodups_wids['sample_id'].map(mapid)\n",
    "    df_nodups_wids['GP2ID'] = [f'{study_code}_{i:06}' for i in df_nodups_wids.uid_idx]\n",
    "    df_nodups_wids['uid_idx_cumcount'] = df_nodups_wids.groupby('GP2ID').cumcount() + 1\n",
    "    df_nodups_wids['GP2sampleID'] = df_nodups_wids.GP2ID + '_s' + df_nodups_wids.uid_idx_cumcount.astype('str')\n",
    "    df_nodups_wids['SampleRepNo'] = 's' + df_nodups_wids.uid_idx_cumcount.astype('str')\n",
    "    df_nodups_wids.drop(['uid_idx','uid_idx_cumcount'], axis = 1, inplace = True)\n",
    "\n",
    "    if df_dups.shape[0]>0:\n",
    "        df_newids = pd.concat([df_dups_wids, df_nodups_wids])\n",
    "    else:\n",
    "        df_newids = df_nodups_wids\n",
    "    \n",
    "    return(df_newids)\n",
    "\n",
    "def assign_unique_gp2clinicalids(df, clinicalid_subset):\n",
    "\n",
    "    if isinstance(clinicalid_subset, pd.Series):\n",
    "        clinicalid_subset = clinicalid_subset.to_frame().T\n",
    "\n",
    "    sampleid = clinicalid_subset.sort_values(by=['master_GP2sampleID'])\\\n",
    "                                .reset_index(drop = True)\\\n",
    "                                .dropna(subset=['master_GP2sampleID'], axis = 0)\n",
    "    sampleid = sampleid.loc[sampleid.index[-1], 'master_GP2sampleID'].split(\"_\")\n",
    "    getuniqueid = sampleid[0] + \"_\" + sampleid[1]\n",
    "    get_sidrepno = int(sampleid[2].replace(\"s\",\"\")) + 1\n",
    "\n",
    "    index_modify = clinicalid_subset['index'].unique() #clinicalid_subset[clinicalid_subset['GP2sampleID'].isnull()] #.index\n",
    "    assign_gp2sampleid = [getuniqueid + \"_s\" + str(get_sidrepno + i) for i in range(len(index_modify))]\n",
    "    df.loc[index_modify, 'GP2sampleID'] = assign_gp2sampleid\n",
    "    getnewidrows = df.loc[index_modify].copy()\n",
    "    return (getnewidrows)\n",
    "\n",
    "def master_keyv2(studies):\n",
    "    # ACCESS MASTERGP2IDS_JSON IN GP2 BUCKET\n",
    "    client = storage.Client()\n",
    "    bucket = client.get_bucket('eu-samplemanifest')\n",
    "    blob = bucket.blob('IDSTRACKER/GP2IDSMAPPER.json')\n",
    "    \n",
    "    ids_tracker = {}\n",
    "    with blob.open(\"r\") as f:\n",
    "        for k, v in ijson.kvitems(f, ''):\n",
    "            if k in studies:\n",
    "                ids_tracker.update({k:v})\n",
    "    \n",
    "    return(ids_tracker)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'----------------------------------------'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "40*\"-\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "        # GENERATE GP2 IDs #\n",
    "        jumptwice()\n",
    "        st.subheader('GP2 IDs assignment...')\n",
    "        studynames = list(df['study'].unique())\n",
    "\n",
    "        if st.session_state['master_get'] == None: # TO ONLY RUN ONCE\n",
    "            #ids_tracker = generategp2ids.master_key(studies = studynames)\n",
    "            ids_tracker = generategp2ids.master_keyv2(studies = studynames)\n",
    "            study_subsets = []\n",
    "            log_new = []\n",
    "            df['GP2sampleID'] = None\n",
    "            # GP2 ID ASSIGNMENT CODE BLOCK\n",
    "            for study in studynames:\n",
    "                st.write(f\"Getting GP2IDs for {study} samples\")\n",
    "                df_subset = df[df.study==study].copy()\n",
    "                try:\n",
    "                    #study_tracker = st.session_state['store_tracker'][study]\n",
    "                    study_tracker = ids_tracker[study]\n",
    "                    study_tracker_df = pd.DataFrame.from_dict(study_tracker,\n",
    "                                                            orient='index',\n",
    "                                                            columns = ['master_GP2sampleID','clinical_id'])\\\n",
    "                                                    .rename_axis('master_sample_id').reset_index()\\\n",
    "                                                    .astype(str)\n",
    "\n",
    "                    # Check if any sample ID exists in df_subset.\n",
    "                    sample_id_unique = pd.merge(study_tracker_df, df_subset,\n",
    "                                                left_on=['master_sample_id'], right_on=['sample_id'], how='inner')\n",
    "                    if not sample_id_unique.empty:\n",
    "                        st.error('We have detected sample ids submitted on previous versions')\n",
    "                        st.error('Please, correct these sample IDs so that they are unique and resubmit the sample manifest.')\n",
    "                        sample_id_unique = sample_id_unique.rename(columns={\"clinical_id_y\": \"clinical_id\"})\n",
    "                        st.dataframe(\n",
    "                        sample_id_unique[['study','sample_id','clinical_id']].style.set_properties(**{\"background-color\": \"brown\", \"color\": \"lawngreen\"})\n",
    "                        )\n",
    "                        stopapp=True\n",
    "                    else:\n",
    "                        stopapp=False\n",
    "                except:\n",
    "                    study_tracker = None\n",
    "                    stopapp = False\n",
    "                if stopapp:\n",
    "                    st.stop()\n",
    "\n",
    "                if bool(study_tracker):\n",
    "                    # WORK ON DUPLICATED IDS\n",
    "                    df_subset = df_subset.reset_index()\n",
    "                    data_duplicated = pd.merge(df_subset, study_tracker_df, on=['clinical_id'], how='inner')\n",
    "                    df_subset = df_subset.set_index('index')\n",
    "                    df_subset.index.name = None\n",
    "\n",
    "                    if data_duplicated.shape[0]>0:\n",
    "                        new_clinicaldups = True\n",
    "                        newids_clinicaldups = data_duplicated.groupby('clinical_id')\\\n",
    "                                                        .apply(lambda x: generategp2ids.assign_unique_gp2clinicalids(df_subset,x))\n",
    "\n",
    "                        if newids_clinicaldups.shape[0]>0:\n",
    "                            newids_clinicaldups = newids_clinicaldups.reset_index(drop=True)[['study','clinical_id','sample_id','GP2sampleID']]\n",
    "                            log_new.append(newids_clinicaldups)\n",
    "                    else:\n",
    "                        new_clinicaldups = False\n",
    "                        newids_clinicaldups = pd.DataFrame()\n",
    "\n",
    "                    # GET GP2 IDs METADATA for new CLINICAL-SAMPLE ID pairs\n",
    "                    df_newids = df_subset[df_subset['GP2sampleID'].isnull()].reset_index(drop = True).copy()\n",
    "                    if not df_newids.empty: # Get new GP2 IDs\n",
    "                        df_wids = df_subset[~df_subset['GP2sampleID'].isnull()].reset_index(drop = True).copy()\n",
    "                        df_wids['GP2ID'] = df_wids['GP2sampleID'].apply(lambda x: (\"_\").join(x.split(\"_\")[:-1]))\n",
    "                        df_wids['SampleRepNo'] = df_wids['GP2sampleID'].apply(lambda x: x.split(\"_\")[-1])#.replace(\"s\",\"\"))\n",
    "\n",
    "                        n=int(max(study_tracker_df['master_GP2sampleID'].to_list()).split(\"_\")[1])+1\n",
    "                        df_newids = generategp2ids.getgp2idsv2(df_newids, n, study)\n",
    "                        df_subset = pd.concat([df_newids, df_wids], axis = 0)\n",
    "                        study_subsets.append(df_subset)\n",
    "                        log_new.append(df_newids[['study','clinical_id','sample_id','GP2sampleID']])\n",
    "                        \n",
    "                    else: # TO CONSIDER THE CASE IN WHICH WE ONLY HAD DUPLICATE IDS MAPPED ON THE MASTER FILE\n",
    "                        df_subset['GP2ID'] = df_subset['GP2sampleID'].apply(lambda x: (\"_\").join(x.split(\"_\")[:-1]))\n",
    "                        df_subset['SampleRepNo'] = df_subset['GP2sampleID'].apply(lambda x: x.split(\"_\")[-1])#.replace(\"s\",\"\"))\n",
    "                        study_subsets.append(df_subset)\n",
    "\n",
    "                # Brand new data - NO STUDY TRACKER FOR THIS COHORT\n",
    "                else:\n",
    "                    study = study\n",
    "                    new_clinicaldups = False # Duplicates from master key json are treated differently to brand new data\n",
    "                    n = 1\n",
    "                    df_newids = generategp2ids.getgp2idsv2(df_subset, n, study)\n",
    "                    study_subsets.append(df_newids)\n",
    "\n",
    "\n",
    "                # CODE TO UPDATE THE GET FILE WE WILL USE TO UPDATE MASTER JSON\n",
    "                if (new_clinicaldups) and (newids_clinicaldups.shape[0]>0):\n",
    "                    tmp = pd.concat([df_newids[['study','clinical_id','sample_id','GP2sampleID']], newids_clinicaldups])\n",
    "                    tmp['master_value'] = list(zip(tmp['GP2sampleID'],\n",
    "                                                    tmp['clinical_id']))\n",
    "                    ids_log = tmp.groupby('study').apply(lambda x: dict(zip(x['sample_id'],\n",
    "                                                                            x['master_value']))).to_dict()\n",
    "                else:\n",
    "                    df_update_master = df_newids.copy()\n",
    "                    df_update_master['master_value'] = list(zip(df_update_master['GP2sampleID'],\n",
    "                                                            df_update_master['clinical_id']))\n",
    "                    ids_log = df_update_master.groupby('study').apply(lambda x: dict(zip(x['sample_id'],\n",
    "                                                                                    x['master_value']))).to_dict()\n",
    "\n",
    "                #generategp2ids.update_masterids(ids_log, study_tracker) # THIS WILL BE UPDATED ONCE THE USET CONFIRMS THE QC ( AT THE END)\n",
    "                \n",
    "                #if st.session_state['master_get'] == None:\n",
    "                if (isinstance(st.session_state['all_ids'], list)):\n",
    "                    st.session_state['all_ids'].append( [ids_log, study_tracker] )\n",
    "                if st.session_state['all_ids'] == None:\n",
    "                    st.session_state['all_ids'] = [ [ids_log, study_tracker] ]\n",
    "            \n",
    "\n",
    "            # OUT OF FOR LOOP // END OF GP2 IDS ASSIGNMENT. LET'S RESUME df.\n",
    "            df = pd.concat(study_subsets, axis = 0)\n",
    "            df = df[list(df)[-3:] + list(df)[:-3]]\n",
    "            st.write(\"GPS IDs assignment... OK\")\n",
    "\n",
    "            #if st.session_state['master_get'] == None:\n",
    "            st.session_state['df_copy'] = df\n",
    "            if len(log_new) > 0:\n",
    "                allnew = pd.concat(log_new, axis = 0).reset_index(drop=True)\n",
    "                st.write(\"Thanks for uploading a new version of the sample manifest\")\n",
    "                st.write(f'We have detected a total of {allnew.shape[0]} new samples')\n",
    "                st.write(\"We have assigned new GP2IDs to those. Showing them below...\")\n",
    "                st.dataframe(\n",
    "                allnew.style.set_properties(**{\"background-color\": \"brown\", \"color\": \"lawngreen\"})\n",
    "                #allnew.style.set_properties(**{\"background-color\": \"brown\", \"color\": \"lawngreen\"})\n",
    "                )\n",
    "            else:\n",
    "                aggridPlotter(df)\n",
    "\n",
    "            st.session_state['df_finalids'] = df\n",
    "            st.session_state['master_get'] = 'DONE'\n",
    "\n",
    "        else:\n",
    "            df = st.session_state['df_finalids']\n",
    "            aggridPlotter(df)\n",
    "            # df_builder = GridOptionsBuilder.from_dataframe(st.session_state['df_copy'])\n",
    "            # df_builder.configure_grid_options(alwaysShowHorizontalScroll = True,\n",
    "            #                                     enableRangeSelection=True,\n",
    "            #                                     pagination=True,\n",
    "            #                                     paginationPageSize=10000,\n",
    "            #                                     domLayout='normal')\n",
    "            # godf = df_builder.build()\n",
    "            # AgGrid(st.session_state['df_copy'],gridOptions=godf, theme='streamlit', height=300)\n",
    "            #df = st.session_state['df_finalids']\n",
    "        #st.session_state['master_get'] = 'DONE'\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scverse10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
