{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "# ASAP CRN Metadata validation - wave 1\n",
    "\n",
    "Teams Hardy, Hafler, and Lee need some minor rationalization.\n",
    "\n",
    "\n",
    "05 Nov 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from utils.io import read_meta_table, get_dtypes_dict, ReportCollector\n",
    "from utils.qcutils import prep_table, validate_table, validate_file, read_ASAP_CDE, reorder_table_to_CDE\n",
    "\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tests\n",
    "\n",
    "Test reading tables with non-unicode text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Activation of transposable elements as a trigg...</td>\n",
       "      <td>Single nuclei sequencing of brain regions from...</td>\n",
       "      <td>A range of genetic, clinical and pathological ...</td>\n",
       "      <td>TEAM-JAKOBSSON</td>\n",
       "      <td>Gale-Hammell Lab</td>\n",
       "      <td>Molly, Gale Hammell</td>\n",
       "      <td>Molly.GaleHammell@nyulangone.org</td>\n",
       "      <td>Anita, Adami; Talitha, Forcier; Raquel, Garza;...</td>\n",
       "      <td>Oliver, H, Tam</td>\n",
       "      <td>oliver.tam@nyulangone.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71</td>\n",
       "      <td>Substantia nigra, prefrontal cortex, amygdala,...</td>\n",
       "      <td>Late stage PD &amp; control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Creative Commons Attribution license CC BY 4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        project_name  \\\n",
       "0  Activation of transposable elements as a trigg...   \n",
       "\n",
       "                                     project_dataset  \\\n",
       "0  Single nuclei sequencing of brain regions from...   \n",
       "\n",
       "                                 project_description  ASAP_team_name  \\\n",
       "0  A range of genetic, clinical and pathological ...  TEAM-JAKOBSSON   \n",
       "\n",
       "      ASAP_lab_name         PI_full_name                          PI_email  \\\n",
       "0  Gale-Hammell Lab  Molly, Gale Hammell  Molly.GaleHammell@nyulangone.org   \n",
       "\n",
       "                                   contributor_names  submitter_name  \\\n",
       "0  Anita, Adami; Talitha, Forcier; Raquel, Garza;...  Oliver, H, Tam   \n",
       "\n",
       "             submittor_email  ... other_funding_source  publication_DOI  \\\n",
       "0  oliver.tam@nyulangone.org  ...                  NaN              NaN   \n",
       "\n",
       "   publication_PMID  number_of_brain_samples  \\\n",
       "0               NaN                       71   \n",
       "\n",
       "                                       brain_regions         types_of_samples  \\\n",
       "0  Substantia nigra, prefrontal cortex, amygdala,...  Late stage PD & control   \n",
       "\n",
       "  PI_ORCHID  PI_google_scholar_id  \\\n",
       "0       NaN                   NaN   \n",
       "\n",
       "                                      DUA_version metadata_version_date  \n",
       "0  Creative Commons Attribution license CC BY 4.0                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fnm = Path.home() / 'Downloads' / 'STUDY_symbols.csv'\n",
    "\n",
    "# STUDY = pd.read_csv(fnm, index_col=0)\n",
    "try:\n",
    "    STUDY = pd.read_csv(fnm)\n",
    "except UnicodeDecodeError:\n",
    "    STUDY = pd.read_csv(fnm, encoding='latin1')\n",
    "    # these work too\n",
    "    # STUDY = pd.read_csv(fnm, encoding='ISO-8859-1')\n",
    "    # STUDY = pd.read_csv(fnm, encoding='Windows-1252')\n",
    "\n",
    "\n",
    "STUDY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Test reading the CDE from google sheets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>team_dataset_id</td>\n",
       "      <td>The \"project_name\" is often too verbose for pr...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table            Field                                        Description  \\\n",
       "0  STUDY   ASAP_team_name  ASAP Team Name:   Name of the ASAP CRN Team. i...   \n",
       "1  STUDY    ASAP_lab_name  Lab Name. :   Lab name that is submitting data...   \n",
       "2  STUDY     project_name  Project Name:   A Title of the overall project...   \n",
       "3  STUDY  team_dataset_id  The \"project_name\" is often too verbose for pr...   \n",
       "4  STUDY  project_dataset  Dataset Name:   A unique name is required for ...   \n",
       "\n",
       "  DataType  Required                                         Validation  \n",
       "0     Enum  Required  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...  \n",
       "1   String  Required                                                NaN  \n",
       "2   String  Required                                                NaN  \n",
       "3   String  Required                                                NaN  \n",
       "4   String  Required                                                NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# # ASAP CDE ALL sheet\n",
    "# sheet_id = \"1SYyOw0zppqdRN-Jwlx6m8ZcBkqZgokjcO39bw6UwF8I\"\n",
    "# url_1a = \"https://docs.google.com/spreadsheets/d/1SYyOw0zppqdRN-Jwlx6m8ZcBkqZgokjcO39bw6UwF8I/edit?usp=sharing\"\n",
    "# url_1b = \"https://docs.google.com/spreadsheets/d/1SYyOw0zppqdRN-Jwlx6m8ZcBkqZgokjcO39bw6UwF8I/edit#gid=43504703\"\n",
    "# sheet_name = \"CDE_final\"\n",
    "# url = f\"https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}\"\n",
    "\n",
    "\n",
    "# ASAP_CDE sheet\n",
    "url_2a = \"https://docs.google.com/spreadsheets/d/1xjxLftAyD0B8mPuOKUp5cKMKjkcsrp_zr9yuVULBLG8/edit#gid=1120395413\"\n",
    "url_2b = \"https://docs.google.com/spreadsheets/d/1xjxLftAyD0B8mPuOKUp5cKMKjkcsrp_zr9yuVULBLG8/edit?usp=sharing\"\n",
    "sheet_id = \"1xjxLftAyD0B8mPuOKUp5cKMKjkcsrp_zr9yuVULBLG8\"\n",
    "sheet_name = \"ASAP_CDE_v2\"  # sheet_name = \"ASAP_CDE_v1\"\n",
    "url = f\"https://docs.google.com/spreadsheets/d/{sheet_id}/gviz/tq?tqx=out:csv&sheet={sheet_name}\"\n",
    "pd.read_csv(url).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Table</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>DataType</th>\n",
       "      <th>Required</th>\n",
       "      <th>Validation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team Name:   Name of the ASAP CRN Team. i...</td>\n",
       "      <td>Enum</td>\n",
       "      <td>Required</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>Lab Name. :   Lab name that is submitting data...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name:   A Title of the overall project...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>team_dataset_id</td>\n",
       "      <td>The \"project_name\" is often too verbose for pr...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>STUDY</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset Name:   A unique name is required for ...</td>\n",
       "      <td>String</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Table            Field                                        Description  \\\n",
       "0  STUDY   ASAP_team_name  ASAP Team Name:   Name of the ASAP CRN Team. i...   \n",
       "1  STUDY    ASAP_lab_name  Lab Name. :   Lab name that is submitting data...   \n",
       "2  STUDY     project_name  Project Name:   A Title of the overall project...   \n",
       "3  STUDY  team_dataset_id  The \"project_name\" is often too verbose for pr...   \n",
       "4  STUDY  project_dataset  Dataset Name:   A unique name is required for ...   \n",
       "\n",
       "  DataType  Required                                         Validation  \n",
       "0     Enum  Required  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\", \"TEAM-...  \n",
       "1   String  Required                                                NaN  \n",
       "2   String  Required                                                NaN  \n",
       "3   String  Required                                                NaN  \n",
       "4   String  Required                                                NaN  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "url_1 = url_2a.replace(\"/edit#gid=\", \"/export?format=csv&gid=\")\n",
    "pd.read_csv(url).head()                          "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean each Team Table\n",
    "write clean metadata tables according to CDE "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Team Lee"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read url\n"
     ]
    }
   ],
   "source": [
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-lee\")\n",
    "metadata_path = data_path / \"metadata/ogmetadata\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.tsv\", delimiter=\"\\t\")\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\",delimiter=\",\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.tsv\",delimiter=\"\\t\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.tsv\",delimiter=\"\\t\")\n",
    "\n",
    "metadata_version = \"v1\"\n",
    "\n",
    "CDE_df,dtypes_dict = read_ASAP_CDE(metadata_version=metadata_version)\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 1</th>\n",
       "      <th>Unnamed: 2</th>\n",
       "      <th>Team-Lee-Bras-Lab-Info</th>\n",
       "      <th>Field</th>\n",
       "      <th>Description</th>\n",
       "      <th>Data type</th>\n",
       "      <th>Validation</th>\n",
       "      <th>Note</th>\n",
       "      <th>Required/Optional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>project_name</td>\n",
       "      <td>Project Name/Title</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Unique and clear title.</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>Dataset name</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A Dataset name is required for each submission...</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>project_description</td>\n",
       "      <td>Brief description of the goals and objectives ...</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP Team e.g. \"Scherzer\"</td>\n",
       "      <td>Enum</td>\n",
       "      <td>[\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bras</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>ASAP Lab under the above team e.g. \"Dong\"</td>\n",
       "      <td>String</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Required</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Unnamed: 0           Unnamed: 1  \\\n",
       "0  Is senescence a component of human PD and does...         project_name   \n",
       "1    Human snRNA-seq PD Senesence Jose Bras Team Lee      project_dataset   \n",
       "2  Characterize the neuropathological progression...  project_description   \n",
       "3                                           TEAM-LEE       ASAP_team_name   \n",
       "4                                               Bras        ASAP_lab_name   \n",
       "\n",
       "                                          Unnamed: 2 Team-Lee-Bras-Lab-Info  \\\n",
       "0                                 Project Name/Title                 String   \n",
       "1                                       Dataset name                 String   \n",
       "2  Brief description of the goals and objectives ...                 String   \n",
       "3                          ASAP Team e.g. \"Scherzer\"                   Enum   \n",
       "4          ASAP Lab under the above team e.g. \"Dong\"                 String   \n",
       "\n",
       "                                          Field  \\\n",
       "0                                           NaN   \n",
       "1                                           NaN   \n",
       "2                                           NaN   \n",
       "3  [\"TEAM-LEE\",\"TEAM-HAFLER\",\"TEAM-HARDY\",....]   \n",
       "4                                           NaN   \n",
       "\n",
       "                                         Description Data type  Validation  \\\n",
       "0                            Unique and clear title.  Required         NaN   \n",
       "1  A Dataset name is required for each submission...  Required         NaN   \n",
       "2                                                NaN  Required         NaN   \n",
       "3                                                NaN  Required         NaN   \n",
       "4                                                NaN  Required         NaN   \n",
       "\n",
       "   Note  Required/Optional  \n",
       "0   NaN                NaN  \n",
       "1   NaN                NaN  \n",
       "2   NaN                NaN  \n",
       "3   NaN                NaN  \n",
       "4   NaN                NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY = pd.read_csv(metadata_path / \"STUDY.tsv\",delimiter=\"\\t\")\n",
    "STUDY.to_csv(data_path / \"STUDY_.csv\")\n",
    "STUDY = pd.read_csv(data_path / \"STUDY_.csv\")\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Is senescence a component of human PD and does...</td>\n",
       "      <td>Human snRNA-seq PD Senesence Jose Bras Team Lee</td>\n",
       "      <td>Characterize the neuropathological progression...</td>\n",
       "      <td>TEAM-LEE</td>\n",
       "      <td>Bras</td>\n",
       "      <td>Jose, Bras</td>\n",
       "      <td>jose.bras@vai.org</td>\n",
       "      <td>Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...</td>\n",
       "      <td>Kaitlyn E Westra</td>\n",
       "      <td>kaitlyn.westra@vai.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75</td>\n",
       "      <td>hippocampus; middle frontal gyrus; substantia ...</td>\n",
       "      <td>human PD and control postmortem brains</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>unsure</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  \\\n",
       "1  Is senescence a component of human PD and does...   \n",
       "\n",
       "0                                  project_dataset  \\\n",
       "1  Human snRNA-seq PD Senesence Jose Bras Team Lee   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "1  Characterize the neuropathological progression...       TEAM-LEE   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name           PI_email  \\\n",
       "1          Bras   Jose, Bras  jose.bras@vai.org   \n",
       "\n",
       "0                                       submitter_id    submitter_name  \\\n",
       "1  Lee, L, Marshall ; Kimberly, E, Paquette ; Kai...  Kaitlyn E Westra   \n",
       "\n",
       "0         submittor_email  ... other_funding_source publication_DOI  \\\n",
       "1  kaitlyn.westra@vai.org  ...                  NaN             NaN   \n",
       "\n",
       "0 publication_PMID number_of_brain_samples  \\\n",
       "1              NaN                      75   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "1  hippocampus; middle frontal gyrus; substantia ...   \n",
       "\n",
       "0                        types_of_samples PI_ORCHID PI_google_scholar_id  \\\n",
       "1  human PD and control postmortem brains       NaN                  NaN   \n",
       "\n",
       "0 DUA_version metadata_version_date  \n",
       "1      unsure                   NaN  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"Unnamed: 1\",\"Unnamed: 0\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "# STUDY[[\"Unnamed: 1\"]].transpose().reset_index().drop(columns=[\"index\"]), tmp\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *STUDY* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in STUDY. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "metadata_version_date = \"v1, SEPT2023\"\n",
    "STUDY[\"metadata_version_date\"] = METADATA_VERSION_DATE\n",
    "# Testing the function with STUDY.csv and CDE.csv\n",
    "STUDY.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "# Need to rename submitter_id to contributor_names\n",
    "STUDY = STUDY.rename(columns={\"submitter_id\": \"contributor_names\"})\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "\n",
    "validate_table(STUDY, \"STUDY\", CDE_df, report)\n",
    "\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect additional metadata from covar.csv .. i.e. batch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "metadata_path = Path.home() / (\"Projects/ASAP/team-lee/metadata\")\n",
    "HIP_covar = pd.read_csv(f\"{metadata_path}/HIP/covar.csv\")\n",
    "HIP_cases = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "HIP_control = pd.read_csv(f\"{metadata_path}/HIP/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "MFG_covar = pd.read_csv(f\"{metadata_path}/MFG/covar.csv\") # includes 'PMI' ?\n",
    "MFG_cases = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "MFG_control = pd.read_csv(f\"{metadata_path}/MFG/PD_ASAP_Sample_batch_information_banner_controls.csv\")\n",
    "\n",
    "\n",
    "SN_covar = pd.read_csv(f\"{metadata_path}/SN/covar.csv\")\n",
    "SN_cases = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_cases.csv\").dropna(axis=0,how='all')\n",
    "SN_control = pd.read_csv(f\"{metadata_path}/SN/PD_ASAP_Sample_batch_information_banner_controls.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hippocampus samples\n",
    "# HIP_cases[\"GROUPcv\"]=\"PD\"\n",
    "# HIP_control[\"GROUPcv\"]=\"HC\"\n",
    "\n",
    "HIP_meta = pd.concat([HIP_cases, HIP_control], axis=0, ignore_index=True)\n",
    "HIP_meta[\"GROUPcv\"]= HIP_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "HIP_meta['MERGE_ID'] = \"HIP_\" + HIP_meta['GROUPcv'] +\"_\" + HIP_meta['CaseID'].str.replace('-','')\n",
    "HIP_covar['MERGE_ID'] = HIP_covar['COUNT_ID']\n",
    "# the fastqs follow COUNT_ID insteald of SEQ_ID naming convention\n",
    "HIP_covar['SEQ_ID'] = HIP_covar['COUNT_ID']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# there's a bug in the meta table... skip for now\n",
    "HIP_TABLE = pd.merge(HIP_covar, HIP_meta, on='MERGE_ID', how='outer')\n",
    "\n",
    "# HIP_TABLE = HIP_covar\n",
    "HIP_TABLE['subdir']=\"HIP\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = HIP_TABLE[[\"MERGE_ID\",\"SEQ_ID\",\"GROUPcv\",\"subdir\",'PD']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "### medial frontal gyrus samples\n",
    "MFG_meta = pd.concat([MFG_cases, MFG_control], axis=0, ignore_index=True)\n",
    "MFG_meta[\"GROUPcv\"]= MFG_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "# make a MERGE_ID column because the formatting is inconsistent\n",
    "MFG_meta['MERGE_ID'] = \"MFG_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "MFG_covar['MERGE_ID'] = MFG_covar['SAMPLE']\n",
    "# the fastqs are in SEQ_ID \n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "MFG_TABLE = pd.merge(MFG_covar, MFG_meta, on='MERGE_ID', how='inner')\n",
    "MFG_TABLE['subdir']=\"MFG\"\n",
    "\n",
    "\n",
    "\n",
    "# Substantia Nigra\n",
    "SN_meta = pd.concat([SN_cases, SN_control], axis=0, ignore_index=True)\n",
    "SN_meta[\"GROUPcv\"] = SN_meta[\"PD\"].apply(lambda x: \"PD\" if (x==\"yes\") else \"HC\")\n",
    "\n",
    "SN_meta['MERGE_ID'] = \"SN_\" + MFG_meta['GROUPcv'] +\"_\" + MFG_meta['CaseID'].str.replace('-','')\n",
    "SN_covar['MERGE_ID'] = SN_covar['SAMPLE']\n",
    "\n",
    "# there's a bug in the meta table... skip for now\n",
    "SN_TABLE = pd.merge(SN_covar, SN_meta, on='MERGE_ID', how='outer')\n",
    "SN_TABLE['subdir']=\"SN\"\n",
    "\n",
    "\n",
    "### concatenate SN, MSG, and HIP tables into one 'all_samples' table\n",
    "all_samples = pd.concat([HIP_TABLE, MFG_TABLE, SN_TABLE], axis=0, ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL = SAMPLE.merge(all_samples, left_on='sample_id', right_on='MERGE_ID', how='left')\n",
    "SAMPLE_ALL.to_csv(\"alternate_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['batch'] = SAMPLE_ALL['BATCH']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the fields with DataType as \"Enum\" or \"String\" for the \"sample\" table from CDE.csv\n",
    "SAMPLE = prep_table(SAMPLE, CDE_df)\n",
    "# for field in string_enum_fields:\n",
    "#     if field in SAMPLE.columns:\n",
    "#         SAMPLE[field] = SAMPLE[field].astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Testing the function with PROTOCOL.csv and CDE.csv\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 23/25 empty rows\n",
      "## Enum fields have valid values in SUBJECT. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Extract the fields with DataType as \"Enum\" or \"String\" for the \"sample\" table from CDE.csv\n",
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "\n",
    "SUBJECT = prep_table(SUBJECT, CDE_df)\n",
    "\n",
    "\n",
    "report = ReportCollector(destination=\"logs\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "SAMPLE = prep_table(SAMPLE, CDE_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE['file_type'] = SAMPLE['file_type'].replace({\"Fastq\":\"fastq\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix file_name and file_MD5 which need to be exploded (do this last for simplicity. i.e. to keep one sample per row rather than one file per row)\n",
    "\n",
    "# Step 1: Split the values in the columns based on commas\n",
    "SAMPLE['file_name'] = SAMPLE['file_name'].str.split(',')\n",
    "SAMPLE['file_MD5(R1,R2)'] = SAMPLE['file_MD5(R1,R2)'].str.split(',')\n",
    "\n",
    "# Step 2: Explode both 'file_name' and 'file_MD5(R1,R2)' columns together\n",
    "SAMPLE = SAMPLE.explode(['file_name', 'file_MD5(R1,R2)'])\n",
    "\n",
    "# Step 3: Rename the \"file_MD5(R1,R2)\" column to \"file_MD5\"\n",
    "SAMPLE = SAMPLE.rename(columns={\"file_MD5(R1,R2)\": \"file_MD5\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 150/150 empty rows\n",
      "\n",
      "\t- RIN: 150/150 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- pm_PH: 150/150 empty rows\n",
      "## Enum fields have valid values in SAMPLE. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MFG_HC_1225</td>\n",
       "      <td>12-25</td>\n",
       "      <td>12-25</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MFG_HC_0602</td>\n",
       "      <td>06-02</td>\n",
       "      <td>06-02</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>PATO:0000461</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MFG_PD_0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>00-09</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle_Frontal_Gyrus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>MONDO:0005180</td>\n",
       "      <td>UBERON:0002702</td>\n",
       "      <td>NA(multiple)</td>\n",
       "      <td>EFO:0030004</td>\n",
       "      <td>nucleus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>BATCH_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     sample_id source_sample_id subject_id replicate  replicate_count  \\\n",
       "0  MFG_HC_1225            12-25      12-25      Rep1                1   \n",
       "0  MFG_HC_1225            12-25      12-25      Rep1                1   \n",
       "1  MFG_HC_0602            06-02      06-02      Rep1                1   \n",
       "1  MFG_HC_0602            06-02      06-02      Rep1                1   \n",
       "2  MFG_PD_0009            00-09      00-09      Rep1                1   \n",
       "\n",
       "   repeated_sample tissue          brain_region  source_RIN  RIN  ...  \\\n",
       "0                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "0                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "1                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "1                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "2                0  Brain  Middle_Frontal_Gyrus         NaN  NaN  ...   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id  disease_ontology_term_id  \\\n",
       "0                                  Unknown              PATO:0000461   \n",
       "0                                  Unknown              PATO:0000461   \n",
       "1                                  Unknown              PATO:0000461   \n",
       "1                                  Unknown              PATO:0000461   \n",
       "2                                  Unknown             MONDO:0005180   \n",
       "\n",
       "  tissue_ontology_term_id cell_type_ontology_term_id assay_ontology_term_id  \\\n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "0          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "1          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "2          UBERON:0002702               NA(multiple)            EFO:0030004   \n",
       "\n",
       "  suspension_type DV2000 pm_PH donor_id    batch  \n",
       "0         nucleus    NaN   NaN           BATCH_4  \n",
       "0         nucleus    NaN   NaN           BATCH_4  \n",
       "1         nucleus    NaN   NaN           BATCH_4  \n",
       "1         nucleus    NaN   NaN           BATCH_4  \n",
       "2         nucleus    NaN   NaN           BATCH_4  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 75/75 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 75/75 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 75/75 empty rows\n",
      "\n",
      "\t- path_year_death: 75/75 empty rows\n",
      "\n",
      "\t- brain_weight: 75/75 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 75/75 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*region_level_2*_:  invalid values üí©'Hippocampus'\n",
      "    - valid ‚û°Ô∏è 'Superior frontal gyrus', 'Middle frontal gyrus', 'Inferior frontal gyrus', 'Superior temporal gyrus', 'Middle temporal gyrus', 'Inferior temporal gyrus', 'Fusiform gyrus', 'Transentorhinal region', 'Entorinal region', 'Subiculum', 'CA1-CA4', 'Amygdala', 'Periamygdala cortex', 'Anterior cingulate gyrus', 'Posterior cingulate gyrus', 'Superior parietal lobule', 'Inferior parietal lobule', 'Parastriate cortex', 'Peristriate cortex', 'Striate cortex', 'Insular cortex', 'Caudate nucleus', 'Putamen', 'Globus pallidus', 'Thalamus', 'Subthalamic nucleus', 'Substantia nigra', 'Pontine tegmentum', 'Pontine base', 'Medulla tegmentum', 'Medulla base', 'Cerebellar vermis', 'Cerebellar hemisphere', 'Dentate nucleus', 'Right colon', 'Left colon', 'Unknown'\n",
      "- _*path_braak_asyn*_:  invalid values üí©'L. Olfactory Bulb-Only', 'Lla. Brainstem Predominant', 'Llb. Limbic Predominant', 'LV. Neocortical', 'Lll. Brainstem/Limbic', '0. No Lewy bodies'\n",
      "    - valid ‚û°Ô∏è '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_mckeith*_:  invalid values üí©'L. Olfactory Bulb-Only', 'Lla. Brainstem Predominant', 'Llb. Limbic Predominant', 'LV. Neocortical', 'Lll. Brainstem/Limbic', '0. No Lewy bodies'\n",
      "    - valid ‚û°Ô∏è 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown'\n",
      "- _*path_nia_ri*_:  invalid values üí©'Criteria not met', 'Not AD'\n",
      "    - valid ‚û°Ô∏è 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*TDP43*_:  invalid values üí©'', 'Na'\n",
      "    - valid ‚û°Ô∏è 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values üí©'', 'Cerebral amyloid angiopathy, temporal and occipital lobe', 'Cerebral amyloid angiopathy, frontal lobe'\n",
      "    - valid ‚û°Ô∏è 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_ad_level*_:  invalid values üí©'Microscopic changes of Alzheimer's disease, insufficient for diagnosis', 'Microscopic lesions of Alzheimer's disease, insufficient for diagnosis'\n",
      "    - valid ‚û°Ô∏è 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CLINPATH = prep_table(CLINPATH, CDE_df)\n",
    "\n",
    "\n",
    "CLINPATH.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['region_level_2'].unique()\n",
    "\n",
    "# change \"Hippocampus\" to \"CA1-CA4\"\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace('Hippocampus', 'CA1-CA4')\n",
    "\n",
    "# skip hx_melanoma and education level for now as there is not a \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# leave te APOE_e4_status as is for now . multiple are coded as \"2,3\" \n",
    "# leave cognitive status as is, since there is no \"Unknown\" or \"Not Reported\" option in the CDE\n",
    "\n",
    "# potential \"path_braak_asyn\" coding \n",
    "braak_map = {'L. Olfactory Bulb-Only':\"1/2\", 'Lla. Brainstem Predominant':\"3\",\n",
    "       'Llb. Limbic Predominant':\"3/4\", 'LV. Neocortical':\"5\",\n",
    "       'Lll. Brainstem/Limbic':\"3/4\", '0. No Lewy bodies':\"0\"}\n",
    "# set to NaN for now since this is actualy path_mckeith coding\n",
    "\n",
    "CLINPATH['path_braak_asyn'] = \"\"\n",
    "\n",
    "mckeith_map = {'L. Olfactory Bulb-Only':\"Olfactory bulb only\", 'Lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'Llb. Limbic Predominant':\"Limbic (transitional)\", 'LV. Neocortical':\"Neocortical\",\n",
    "       'Lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\"}\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in CLINPATH.\n",
      "\t\tRequired Fields with Empty or NaN values:\n",
      "\t\t\t- age_at_onset: 75 rows\n",
      "\t\t\t- age_at_diagnosis: 75 rows\n",
      "\t\t\t- first_motor_symptom: 75 rows\n",
      "\t\t\t- path_year_death: 75 rows\n",
      "\t\t\t- brain_weight: 75 rows\n",
      "\t\tOptional Fields with Empty or NaN values:\n",
      "\t\t\t- smoking_years: 75 rows\n",
      "\tInvalid Field/Value pairs:\n",
      "\t\t\t- path_nia_ri: Criteria not met, Not AD\n",
      "\t\t\t- TDP43: , Na\n",
      "\t\t\t- amyloid_angiopathy_severity_scale: , Cerebral amyloid angiopathy, temporal and occipital lobe, Cerebral amyloid angiopathy, frontal lobe\n",
      "\t\t\t- path_ad_level: Microscopic changes of Alzheimer's disease, insufficient for diagnosis, Microscopic lesions of Alzheimer's disease, insufficient for diagnosis\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "validate_table(CLINPATH, \"CLINPATH\", CDE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 75/75 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 75/75 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 75/75 empty rows\n",
      "\n",
      "\t- path_year_death: 75/75 empty rows\n",
      "\n",
      "\t- brain_weight: 75/75 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 75/75 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*path_nia_ri*_:  invalid values üí©'Criteria not met', 'Not AD'\n",
      "    - valid ‚û°Ô∏è 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*TDP43*_:  invalid values üí©'', 'Na'\n",
      "    - valid ‚û°Ô∏è 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values üí©'', 'Cerebral amyloid angiopathy, temporal and occipital lobe', 'Cerebral amyloid angiopathy, frontal lobe'\n",
      "    - valid ‚û°Ô∏è 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_ad_level*_:  invalid values üí©'Microscopic changes of Alzheimer's disease, insufficient for diagnosis', 'Microscopic lesions of Alzheimer's disease, insufficient for diagnosis'\n",
      "    - valid ‚û°Ô∏è 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SAMPLE_ALL_CP = SAMPLE_ALL.merge(CLINPATH, on='sample_id', how='outer')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE_ALL_CP.to_csv(\"./clean/team-Lee/auxiluary_metadata.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the column order\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE)\n",
    "\n",
    "# write the clean metadata\n",
    "STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Lee\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Hafler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/team-hafler/metadata')"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## convert to seurat Object\n",
    "data_path = Path.home() / (\"Projects/ASAP\")\n",
    "metadata_path = data_path / \"team-hafler/metadata\"\n",
    "\n",
    "sheets = [\"SAMPLE\",\"SUBJECT\",\"CLINPATH\",\"STUDY\",\"PROTOCOL\"]\n",
    "excel_path = metadata_path / \"ASAP_CDE_ALL_Team_Hafler_v1.xlsx\"\n",
    "STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\",header=1).drop(columns=\"Field\")\n",
    "CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\",header=1).drop(columns=\"Field\")\n",
    "SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\",header=1).drop(columns=\"Field\")\n",
    "SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\",header=1).drop(columns=\"Field\")\n",
    "PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\",header=1).drop(columns=\"Field\")\n",
    "metadata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def add_hafler_batch(sample_df):\n",
    "\n",
    "    # First batch: HSDG07HC HSDG10HC HSDG148PD HSDG199PD\n",
    "    # batch[batch.sample_id in ['hSDG07HC', 'hSDG10HC', 'hSDG148PD', 'hSDG199PD']]=1\n",
    "    Batch_1 = ['hSDG07', 'hSDG10', 'hSDG148', 'hSDG199'] \n",
    "    # Second batch: hsDG101HC hsDG13HC hsDG151PD hsDG197PD hsDG30HC hsDG99HC\n",
    "    Batch_2 = ['hSDG101', 'hSDG13', 'hSDG151', 'hSDG197', 'hSDG30', 'hSDG99']\n",
    "    # Third batch: hsDG142PD hsDG208PD\n",
    "    Batch_3 = ['hSDG142', 'hSDG208'] \n",
    "\n",
    "\n",
    "    batch_col = []\n",
    "    for row in sample_df.sample_id:\n",
    "        if row in Batch_1:\n",
    "            batch_col.append(\"Batch_1\")\n",
    "        elif row in Batch_2:\n",
    "            batch_col.append(\"Batch_2\")\n",
    "        elif row in Batch_3:\n",
    "            batch_col.append(\"Batch_3\")\n",
    "        else:\n",
    "            print(\"ERROR >>>>>>>> not no batch info\")\n",
    "            batch_col.append(\"\")\n",
    "\n",
    "\n",
    "    sample_df['batch'] = batch_col\n",
    "    return sample_df\n",
    "\n",
    "SAMPLE = add_hafler_batch(SAMPLE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>source_RIN</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV2000</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hSDG07</td>\n",
       "      <td>hSDG07</td>\n",
       "      <td>HC01</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hSDG101</td>\n",
       "      <td>hSDG101</td>\n",
       "      <td>HC03</td>\n",
       "      <td>Rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>Batch_2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Prefrontal Cortex</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_id source_sample_id subject_id replicate  replicate_count  \\\n",
       "0    hSDG07           hSDG07       HC01      Rep1                1   \n",
       "1    hSDG07           hSDG07       HC01      Rep1                1   \n",
       "2    hSDG07           hSDG07       HC01      Rep1                1   \n",
       "3   hSDG101          hSDG101       HC03      Rep1                1   \n",
       "4   hSDG101          hSDG101       HC03      Rep1                1   \n",
       "\n",
       "   repeated_sample    batch tissue       brain_region  source_RIN  ...  \\\n",
       "0                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "1                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "2                0  Batch_1  Brain  Prefrontal Cortex         NaN  ...   \n",
       "3                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "4                0  Batch_2  Brain  Prefrontal Cortex         NaN  ...   \n",
       "\n",
       "   sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0                   NaN                                      NaN   \n",
       "1                   NaN                                      NaN   \n",
       "2                   NaN                                      NaN   \n",
       "3                   NaN                                      NaN   \n",
       "4                   NaN                                      NaN   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id cell_type_ontology_term_id  \\\n",
       "0                      NaN                     NaN                        NaN   \n",
       "1                      NaN                     NaN                        NaN   \n",
       "2                      NaN                     NaN                        NaN   \n",
       "3                      NaN                     NaN                        NaN   \n",
       "4                      NaN                     NaN                        NaN   \n",
       "\n",
       "  assay_ontology_term_id suspension_type DV2000 pm_PH  donor_id  \n",
       "0                    NaN             NaN    NaN   NaN       NaN  \n",
       "1                    NaN             NaN    NaN   NaN       NaN  \n",
       "2                    NaN             NaN    NaN   NaN       NaN  \n",
       "3                    NaN             NaN    NaN   NaN       NaN  \n",
       "4                    NaN             NaN    NaN   NaN       NaN  \n",
       "\n",
       "[5 rows x 42 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix replicate & replicate_count\n",
    "SAMPLE.replace(\"Nan\", \"\", inplace=True)\n",
    "\n",
    "SAMPLE['replicate'] = \"Rep1\"\n",
    "SAMPLE['replicate_count'] = 1\n",
    "\n",
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    V1_08/27/2023\n",
       "Name: metadata_version_date, dtype: object"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "STUDY = prep_table(STUDY, CDE_df)\n",
    "STUDY['metadata_version_date']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update metadata_version_date\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tAll required fields are present in STUDY.\n",
      "No empty or NaN values found in Required fields.\n",
      "No empty or NaN values found in Optional fields.\n",
      "\tAll Enum fields have valid values in STUDY.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(STUDY, \"STUDY\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 12/12 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*sex*_:  invalid values üí©'F', 'M'\n",
      "    - valid ‚û°Ô∏è 'Male', 'Female', 'Intersex', 'Unnown'\n",
      "- _*race*_:  invalid values üí©'B', 'W'\n",
      "    - valid ‚û°Ô∏è 'American Indian or Alaska Native', 'Asian', 'White', 'Black or African American', 'Multi-racial', 'Native Hawaiian or Other Pacific Islander', 'Other', 'Unknown', 'Not Reported'\n",
      "- _*ethnicity*_:  invalid values üí©'B', 'W'\n",
      "    - valid ‚û°Ô∏è 'Hispanic or Latino', 'Not Hispanic or Latino', 'Unknown', 'Not Reported'\n",
      "- _*primary_diagnosis*_:  invalid values üí©'Normal control', 'Idiopathic Parkinson's disease'\n",
      "    - valid ‚û°Ô∏è 'Healthy Control', 'Idiopathic PD', 'Alzheimer‚Äôs disease', 'Frontotemporal dementia', 'Corticobasal syndrome', 'Dementia with Lewy bodies', 'Dopa-responsive dystonia', 'Essential tremor', 'Hemiparkinson/hemiatrophy syndrome', 'Juvenile autosomal recessive parkinsonism', 'Motor neuron disease with parkinsonism', 'Multiple system atrophy', 'Neuroleptic-induced parkinsonism', 'Normal pressure hydrocephalus', 'Progressive supranuclear palsy', 'Psychogenic parkinsonism', 'Vascular parkinsonism', 'No PD nor other neurological disorder', 'Spinocerebellar Ataxia (SCA)', 'Prodromal non-motor PD', 'Prodromal motor PD', 'Other neurological disorder'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "# Testing the function with SUBJECT.csv and CDE.csv\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "SUBJECT = prep_table(SUBJECT, CDE_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 12/12 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*ethnicity*_:  invalid values üí©'B', 'W'\n",
      "    - valid ‚û°Ô∏è 'Hispanic or Latino', 'Not Hispanic or Latino', 'Unknown', 'Not Reported'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SUBJECT['sex'] = SUBJECT['sex'].replace({'F':\"Female\", 'M':\"Male\"})\n",
    "SUBJECT['race'] = SUBJECT['race'].replace({'W':\"White\", 'B':\"Black or African American\"})\n",
    "\n",
    "SUBJECT['primary_diagnosis'] = SUBJECT['primary_diagnosis'].replace({'Normal control':\"Healthy Control\", \"Idiopathic Parkinson's disease\":\"Idiopathic PD\"})\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "# they \"race\" was reported under ethnicity, so code as \"Not Reported\"\n",
    "SUBJECT['ethnicity'] = 'Not Reported'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 12/12 empty rows\n",
      "## Enum fields have valid values in SUBJECT. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = prep_table(SAMPLE,  CDE_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "# set time == 0 for all samples\n",
    "SAMPLE['time'] = 0\n",
    "\n",
    "SAMPLE['file_type'] = SAMPLE['file_type'].replace({\"Fastq\":\"fastq\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# need to join with subject to get \"sex\" and convert to ontology term\n",
    "SAMPLE_SUBJECT = SAMPLE.merge(SUBJECT, on='subject_id',  how='left')\n",
    "SAMPLE_og = SAMPLE.copy()\n",
    "SAMPLE['sex_ontology_term_id'] = SAMPLE_SUBJECT['sex'].replace({\"Male\":\"PATO:0000384 (male)\", \"Female\":\"PATO:0000383 (female)\" })\n",
    "\n",
    "# ignore development_stage_ontology_term_id, self_reported_ethnicity_ontology_term_id, assay_ontology_term_id, etc for now. (Check wiht Le)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix assay\n",
    "SAMPLE['assay'] = SAMPLE['assay'].replace({'v3.1 - Single Index, 10x Genomics ':\"v3.1 - Single Index\"})\n",
    "# fix assay\n",
    "SAMPLE['sequencing_length'] = SAMPLE['sequencing_length'].replace({'150bp x2':\"150\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 36/36 empty rows\n",
      "\n",
      "\t- RIN: 36/36 empty rows\n",
      "\n",
      "\t- suspension_type: 36/36 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- pm_PH: 36/36 empty rows\n",
      "## Enum fields have valid values in SAMPLE. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in CLINPATH: path_thal**\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 12/12 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 12/12 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 12/12 empty rows\n",
      "\n",
      "\t- path_year_death: 12/12 empty rows\n",
      "\n",
      "\t- brain_weight: 12/12 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 12/12 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*region_level_2*_:  invalid values üí©'Prefrontal cortex'\n",
      "    - valid ‚û°Ô∏è 'Superior frontal gyrus', 'Middle frontal gyrus', 'Inferior frontal gyrus', 'Superior temporal gyrus', 'Middle temporal gyrus', 'Inferior temporal gyrus', 'Fusiform gyrus', 'Transentorhinal region', 'Entorinal region', 'Subiculum', 'CA1-CA4', 'Amygdala', 'Periamygdala cortex', 'Anterior cingulate gyrus', 'Posterior cingulate gyrus', 'Superior parietal lobule', 'Inferior parietal lobule', 'Parastriate cortex', 'Peristriate cortex', 'Striate cortex', 'Insular cortex', 'Caudate nucleus', 'Putamen', 'Globus pallidus', 'Thalamus', 'Subthalamic nucleus', 'Substantia nigra', 'Pontine tegmentum', 'Pontine base', 'Medulla tegmentum', 'Medulla base', 'Cerebellar vermis', 'Cerebellar hemisphere', 'Dentate nucleus', 'Right colon', 'Left colon', 'Unknown'\n",
      "- _*APOE_e4_status*_:  invalid values üí©'3,3', '2,3'\n",
      "    - valid ‚û°Ô∏è '22', '23', '24', '33', '34', '44', 'Unknown'\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (<NA> or NaN):**\n",
      "- hx_dementia_mcihx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- path_autopsy_dx_main \n",
      "- path_braak_nft \n",
      "- path_braak_asyn \n",
      "- path_cerad \n",
      "- known_pathogenic_mutation \n",
      "- path_mckeith \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# CLINPATH.replace(\"Nan\", \"\", inplace=True)\n",
    "# CLINPATH.replace(\"nan\", \"\", inplace=True)\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH = prep_table(CLINPATH, CDE_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# redact \"Prefrontal Cortex\" from region_level_2 for now\n",
    "CLINPATH['region_level_2'] = CLINPATH['region_level_2'].replace({'Prefrontal Cortex':\"\"})\n",
    "\n",
    "# leave te APOE_e4_status as is for now . multiple are coded as \"2,3\" \n",
    "# but remove commas\n",
    "CLINPATH[\"APOE_e4_status\"] = CLINPATH[\"APOE_e4_status\"].str.replace(\",\",\"\")\n",
    "\n",
    "# need to fix the path_autopsy_dx_main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in CLINPATH: path_thal**\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 12/12 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 12/12 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 12/12 empty rows\n",
      "\n",
      "\t- path_year_death: 12/12 empty rows\n",
      "\n",
      "\t- brain_weight: 12/12 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 12/12 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*region_level_2*_:  invalid values üí©'Prefrontal cortex'\n",
      "    - valid ‚û°Ô∏è 'Superior frontal gyrus', 'Middle frontal gyrus', 'Inferior frontal gyrus', 'Superior temporal gyrus', 'Middle temporal gyrus', 'Inferior temporal gyrus', 'Fusiform gyrus', 'Transentorhinal region', 'Entorinal region', 'Subiculum', 'CA1-CA4', 'Amygdala', 'Periamygdala cortex', 'Anterior cingulate gyrus', 'Posterior cingulate gyrus', 'Superior parietal lobule', 'Inferior parietal lobule', 'Parastriate cortex', 'Peristriate cortex', 'Striate cortex', 'Insular cortex', 'Caudate nucleus', 'Putamen', 'Globus pallidus', 'Thalamus', 'Subthalamic nucleus', 'Substantia nigra', 'Pontine tegmentum', 'Pontine base', 'Medulla tegmentum', 'Medulla base', 'Cerebellar vermis', 'Cerebellar hemisphere', 'Dentate nucleus', 'Right colon', 'Left colon', 'Unknown'\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (<NA> or NaN):**\n",
      "- hx_dementia_mcihx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- path_autopsy_dx_main \n",
      "- path_braak_nft \n",
      "- path_braak_asyn \n",
      "- path_cerad \n",
      "- known_pathogenic_mutation \n",
      "- path_mckeith \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = data_path / \"team-hafler\"\n",
    "data_path = Path.home() / (\"Projects/ASAP/team-hafler\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fix the column order\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE)\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE)\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE)\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE)     \n",
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE)\n",
    "\n",
    "# write the clean metadata\n",
    "STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Hafler\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STUDY.to_csv( metadata_path / \"STUDY.csv\")\n",
    "# CLINPATH.to_csv( metadata_path / \"CLINPATH.csv\")\n",
    "# SUBJECT.to_csv( metadata_path / \"SUBJECT.csv\")\n",
    "# SAMPLE.to_csv( metadata_path / \"SAMPLE.csv\")\n",
    "# PROTOCOL.to_csv( metadata_path / \"PROTOCOL.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[\"22\", \"23\",\"24\",\"33\", \"34\", \"44\", \"Unknown\"]'], dtype=object)"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH[\"APOE_e4_status\"].unique()\n",
    "\n",
    "CDE[CDE[\"Field\"]== \"APOE_e4_status\"].Validation.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team Hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-hardy\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\")\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\")\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>project_name</td>\n",
       "      <td>project_dataset</td>\n",
       "      <td>project_description</td>\n",
       "      <td>ASAP_team_name</td>\n",
       "      <td>ASAP_lab_name</td>\n",
       "      <td>PI_full_name</td>\n",
       "      <td>PI_email</td>\n",
       "      <td>contributor_names</td>\n",
       "      <td>submitter_name</td>\n",
       "      <td>submitter_email</td>\n",
       "      <td>...</td>\n",
       "      <td>other_funding_source</td>\n",
       "      <td>publication_DOI</td>\n",
       "      <td>publication_PMID</td>\n",
       "      <td>number_of_brain_samples</td>\n",
       "      <td>brain_regions</td>\n",
       "      <td>types_of_samples</td>\n",
       "      <td>PI_ORCHID</td>\n",
       "      <td>PI_google_scholar_id</td>\n",
       "      <td>DUA_version</td>\n",
       "      <td>metadata_version_date</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  0                1   \\\n",
       "0                                       project_name  project_dataset   \n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "                                                  2               3   \\\n",
       "0                                project_description  ASAP_team_name   \n",
       "1  Genetic analysis has identified many risk gene...      TEAM-HARDY   \n",
       "\n",
       "              4             5                     6   \\\n",
       "0  ASAP_lab_name  PI_full_name              PI_email   \n",
       "1      Ryten Lab    Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "                                                  7                        8   \\\n",
       "0                                  contributor_names           submitter_name   \n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "                                     9   ...                    11  \\\n",
       "0                       submitter_email  ...  other_funding_source   \n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                   NaN   \n",
       "\n",
       "                12                13                       14  \\\n",
       "0  publication_DOI  publication_PMID  number_of_brain_samples   \n",
       "1              NaN               NaN                      128   \n",
       "\n",
       "                                                  15  \\\n",
       "0                                      brain_regions   \n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "                                                  16                   17  \\\n",
       "0                                   types_of_samples            PI_ORCHID   \n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "                                                  18           19  \\\n",
       "0                               PI_google_scholar_id  DUA_version   \n",
       "1  https://scholar.google.co.uk/citations?user=lt...          NaN   \n",
       "\n",
       "                      20  \n",
       "0  metadata_version_date  \n",
       "1     Version 1, 09/2023  \n",
       "\n",
       "[2 rows x 21 columns]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = STUDY[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submitter_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Version 1, 09/2023</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  project_dataset  \\\n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "1  Genetic analysis has identified many risk gene...     TEAM-HARDY   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name              PI_email  \\\n",
       "1     Ryten Lab   Mina Ryten  mina.ryten@ucl.ac.uk   \n",
       "\n",
       "0                                  contributor_names           submitter_name  \\\n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "0                       submitter_email  ... other_funding_source  \\\n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                  NaN   \n",
       "\n",
       "0 publication_DOI publication_PMID number_of_brain_samples  \\\n",
       "1             NaN              NaN                     128   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "0                                   types_of_samples            PI_ORCHID  \\\n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "0                               PI_google_scholar_id DUA_version  \\\n",
       "1  https://scholar.google.co.uk/citations?user=lt...         NaN   \n",
       "\n",
       "0 metadata_version_date  \n",
       "1    Version 1, 09/2023  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "tmp.columns = tmp.iloc[0]\n",
    "STUDY = tmp.drop([0])\n",
    "STUDY.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "STUDY = prep_table(STUDY, CDE_df)\n",
    "# update metadata_version_date\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in STUDY: submittor_email**\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in STUDY. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(STUDY, \"STUDY\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submitter_email</th>\n",
       "      <th>...</th>\n",
       "      <th>other_funding_source</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Understanding mechanisms of Parkinson's diseas...</td>\n",
       "      <td>Hardy snRNA-seq</td>\n",
       "      <td>Genetic analysis has identified many risk gene...</td>\n",
       "      <td>TEAM-HARDY</td>\n",
       "      <td>Ryten Lab</td>\n",
       "      <td>Mina Ryten</td>\n",
       "      <td>Mina.ryten@ucl.ac.uk</td>\n",
       "      <td>Aine Fairbrother-Browne, Jonathan Brenton, Mel...</td>\n",
       "      <td>Aine Fairbrother-Browne</td>\n",
       "      <td>aine.fairbrother-browne.18@ucl.ac.uk</td>\n",
       "      <td>...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>128</td>\n",
       "      <td>Inferior Parietal Lobule (IPL), Anterior Cingu...</td>\n",
       "      <td>Late stage (Braak 5-6) PD and control post-mor...</td>\n",
       "      <td>0000-0001-9520-6957</td>\n",
       "      <td>Https://scholar.google.co.uk/citations?user=lt...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>v1_20231113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "0                                       project_name  project_dataset  \\\n",
       "1  Understanding mechanisms of Parkinson's diseas...  Hardy snRNA-seq   \n",
       "\n",
       "0                                project_description ASAP_team_name  \\\n",
       "1  Genetic analysis has identified many risk gene...     TEAM-HARDY   \n",
       "\n",
       "0 ASAP_lab_name PI_full_name              PI_email  \\\n",
       "1     Ryten Lab   Mina Ryten  Mina.ryten@ucl.ac.uk   \n",
       "\n",
       "0                                  contributor_names           submitter_name  \\\n",
       "1  Aine Fairbrother-Browne, Jonathan Brenton, Mel...  Aine Fairbrother-Browne   \n",
       "\n",
       "0                       submitter_email  ... other_funding_source  \\\n",
       "1  aine.fairbrother-browne.18@ucl.ac.uk  ...                  Nan   \n",
       "\n",
       "0 publication_DOI publication_PMID number_of_brain_samples  \\\n",
       "1             Nan              Nan                     128   \n",
       "\n",
       "0                                      brain_regions  \\\n",
       "1  Inferior Parietal Lobule (IPL), Anterior Cingu...   \n",
       "\n",
       "0                                   types_of_samples            PI_ORCHID  \\\n",
       "1  Late stage (Braak 5-6) PD and control post-mor...  0000-0001-9520-6957   \n",
       "\n",
       "0                               PI_google_scholar_id DUA_version  \\\n",
       "1  Https://scholar.google.co.uk/citations?user=lt...         Nan   \n",
       "\n",
       "0 metadata_version_date  \n",
       "1           v1_20231113  \n",
       "\n",
       "[1 rows x 21 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_collection_summary</th>\n",
       "      <th>cell_extraction_summary</th>\n",
       "      <th>lib_prep_summary</th>\n",
       "      <th>data_processing_summary</th>\n",
       "      <th>github_url</th>\n",
       "      <th>protocols_io_DOI</th>\n",
       "      <th>other_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>This dataset contains cortical regions only, p...</td>\n",
       "      <td>From protocols.io: This protocol is used to is...</td>\n",
       "      <td>'Nuclei were extracted from homogenised post-m...</td>\n",
       "      <td>Cell ranger was used to convert raw sequencing...</td>\n",
       "      <td>Raw to fastq to mapped: https://github.com/RHR...</td>\n",
       "      <td>Nuclear extraction protocol: 10.17504/protocol...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "0                          sample_collection_summary  \\\n",
       "1  This dataset contains cortical regions only, p...   \n",
       "\n",
       "0                            cell_extraction_summary  \\\n",
       "1  From protocols.io: This protocol is used to is...   \n",
       "\n",
       "0                                   lib_prep_summary  \\\n",
       "1  'Nuclei were extracted from homogenised post-m...   \n",
       "\n",
       "0                            data_processing_summary  \\\n",
       "1  Cell ranger was used to convert raw sequencing...   \n",
       "\n",
       "0                                         github_url  \\\n",
       "1  Raw to fastq to mapped: https://github.com/RHR...   \n",
       "\n",
       "0                                   protocols_io_DOI other_reference  \n",
       "1  Nuclear extraction protocol: 10.17504/protocol...             NaN  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fix STUDY formatting\n",
    "tmp = pd.DataFrame()\n",
    "tmp = PROTOCOL[[\"name\",\"value\"]].transpose().reset_index().drop(columns=[\"index\"])\n",
    "tmp.columns = tmp.iloc[0]\n",
    "PROTOCOL = tmp.drop([0])\n",
    "PROTOCOL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# fix the column order\n",
    "PROTOCOL = prep_table(PROTOCOL, CDE_df)\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE_df, report)\n",
    "report.print_log()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- duration_pmi: 1/64 empty rows\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in SUBJECT. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "SUBJECT = prep_table(SUBJECT, CDE_df)\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_42890/220114886.py:3: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  SUBJECT.fillna(\"\", inplace=True)\n"
     ]
    }
   ],
   "source": [
    "SUBJECT.replace(\"Nan\", \"\", inplace=True)\n",
    "SUBJECT.replace(\"nan\", \"\", inplace=True)\n",
    "SUBJECT.fillna(\"\", inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in SAMPLE: source_sample_id**\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 3616/3616 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- pm_PH: 3616/3616 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*sequencing_length*_:  invalid values üí©'190'\n",
      "    - valid ‚û°Ô∏è '25', '50', '100', '150'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# force the right sex_ontology_term_id\n",
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "\n",
    "SAMPLE = prep_table(SAMPLE, CDE_df)\n",
    "\n",
    "# allow sequence_length == 190 for now\n",
    "\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE_df, report)\n",
    "report.print_log()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# add 'replicate' coding (nans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in CLINPATH: source_sample_id**\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 138/138 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 10/138 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 138/138 empty rows\n",
      "\n",
      "\t- path_year_death: 138/138 empty rows\n",
      "\n",
      "\t- brain_weight: 138/138 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 138/138 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*path_autopsy_dx_main*_:  invalid values üí©'Control brain', 'Pathological ageing', 'Control brain / Path ageing', 'Argyrophilic grain disease', 'Control brain, Cerebrovascular disease (small vessel)', 'Cerebrovascular disease (small vessel)', 'Control brain, Alzheimer`s disease (intermediate level AD pathological change)', 'Control brain / Path ageing, CAA'\n",
      "    - valid ‚û°Ô∏è 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder'\n",
      "- _*path_braak_nft*_:  invalid values üí©'2.0', '1.0', '3.0', '0.0', '4.0', '6.0'\n",
      "    - valid ‚û°Ô∏è '0', 'I', 'II', 'III', 'IV', 'V', 'VI', 'I/II', 'III/IV', 'IV/V', 'V/VI'\n",
      "- _*path_braak_asyn*_:  invalid values üí©'6.0', '0.0', '5.0'\n",
      "    - valid ‚û°Ô∏è '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_thal*_:  invalid values üí©'At least 4'\n",
      "    - valid ‚û°Ô∏è '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5'\n",
      "- _*path_mckeith*_:  invalid values üí©'Diffuse neocortical', 'Limbic transitional', 'Diffuse Neocortical'\n",
      "    - valid ‚û°Ô∏è 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown'\n",
      "- _*path_nia_aa_a*_:  invalid values üí©'1.0', '2.0', '0.0', '3.0'\n",
      "    - valid ‚û°Ô∏è 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values üí©'1.0', '2.0', '0.0', '3.0'\n",
      "    - valid ‚û°Ô∏è 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values üí©'0.0', '1.0', '2.0', '3.0'\n",
      "    - valid ‚û°Ô∏è 'C0', 'C1', 'C2', 'C3'\n",
      "- _*path_ad_level*_:  invalid values üí©'No evidence'\n",
      "    - valid ‚û°Ô∏è 'No evidence of Alzheimer's disease neuropathological change', 'Low level Alzheimer's disease neuropathological change', 'At least low level Alzheimer's disease neuropathological change', 'Intermediate level Alzheimer's disease neuropathological change', 'At least intermediate level Alzheimer's disease neuropathological change', 'High level Alzheimer's disease neuropathological change', 'Unknown'\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (<NA> or NaN):**\n",
      "- family_historyhx_dementia_mci \n",
      "- hx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- APOE_e4_status \n",
      "- cognitive_status \n",
      "- path_autopsy_dx_main \n",
      "- path_braak_nft \n",
      "- path_braak_asyn \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- known_pathogenic_mutation \n",
      "- path_mckeith \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")\n",
    "\n",
    "CLINPATH = prep_table(CLINPATH, CDE_df)\n",
    "\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "na_float2intstr = lambda val: str(int(float(val))) if val != 'Nan' else pd.NA \n",
    "\n",
    "# replace 'path_braak_asyn' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_braak_asyn'] = CLINPATH['path_braak_asyn'].apply(na_float2intstr)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_braak_nft'] = CLINPATH['path_braak_nft'].apply(na_float2intstr).replace({\"0\":\"0\", \n",
    "                                                                                        \"1\":\"I\", \n",
    "                                                                                        \"2\": \"II\", \n",
    "                                                                                        \"3\":\"III\", \n",
    "                                                                                        \"4\":\"IV\", \n",
    "                                                                                        \"5\":\"V\", \n",
    "                                                                                        \"6\":\"VI\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# code family_history as \"Not Reported\" (currently empty)\n",
    "CLINPATH['family_history'] = \"Not Reported\"\n",
    "\n",
    "\n",
    "\n",
    "# check APOE_e4_status ? currently empty\n",
    "\n",
    "# `path_autopsy_dx_main`  actually seems good parser might be wrong\n",
    "\n",
    "# code \"at least 4\" as \"4/5\" \n",
    "\n",
    "CLINPATH['path_thal'] = CLINPATH['path_thal'].replace({'At least 4':\"4/5\"})\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace({'Diffuse neocortical': \"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\", \n",
    "                                                        'Limbic transitional': \"Limbic (transitional)\" ,\n",
    "                                                        'Diffuse Neocortical':\"Diffuse, neocortical (brainstem, limbic and neocortical involvement)\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "CLINPATH['path_nia_aa_a'] = CLINPATH['path_nia_aa_a'].apply(na_float2intstr).replace({\"0\":\"A0\", \n",
    "                                                                                    \"1\":\"A1\", \n",
    "                                                                                    \"2\": \"A2\", \n",
    "                                                                                    \"3\":\"A3\"})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_b'] = CLINPATH['path_nia_aa_b'].apply(na_float2intstr).replace({\"0\":\"B0\", \n",
    "                                                                                    \"1\":\"B1\", \n",
    "                                                                                    \"2\": \"B2\", \n",
    "                                                                                    \"3\":\"B3\"})\n",
    "\n",
    "\n",
    "# replace 'path_braak_nft' with with string of the numeric. converte nan to \"\"\n",
    "CLINPATH['path_nia_aa_c'] = CLINPATH['path_nia_aa_c'].apply(na_float2intstr).replace({\"0\":\"C0\", \n",
    "                                                                                    \"1\":\"C1\", \n",
    "                                                                                    \"2\": \"C2\", \n",
    "                                                                                    \"3\":\"C3\"})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "CLINPATH['path_ad_level'] = CLINPATH['path_ad_level'].replace({\"No evidence\": \"No evidence of Alzheimer\\'s disease neuropathological change\"})\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_autopsy_map = { \"Parkinson's disease with dementia\": \"Parkinson's disease with dementia\", \n",
    "       \"Parkinson's disease\": \"Parkinson's disease\",\n",
    "       'Control brain':\"Control, no misfolded protein or significant vascular pathology\", \n",
    "       'Pathological ageing': 'Control, no misfolded protein or significant vascular pathology',\n",
    "       'Control brain / Path ageing': 'Control, no misfolded protein or significant vascular pathology',\n",
    "       'Argyrophilic grain disease': \"Control, Argyrophilic grain disease\",\n",
    "       'Control brain, Cerebrovascular disease (small vessel)':\"Control, Cerebrovascular disease (atherosclerosis)\",\n",
    "       'Cerebrovascular disease (small vessel)':\"Control, Cerebrovascular disease (atherosclerosis)\",\n",
    "       \"Control brain, Alzheimer`s disease (intermediate level AD pathological change)\":\"Alzheimer's disease (intermediate level neuropathological change)\",\n",
    "       'Control brain / Path ageing, CAA':\"Control, Cerebrovascular disease (cerebral amyloid angiopathy)\"}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH['path_autopsy_dx_main'] = CLINPATH['path_autopsy_dx_main'].replace(path_autopsy_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([\"Parkinson's disease with dementia\", \"Parkinson's disease\",\n",
       "       'Control, no misfolded protein or significant vascular pathology',\n",
       "       'Control, Argyrophilic grain disease',\n",
       "       'Control, Cerebrovascular disease (atherosclerosis)',\n",
       "       \"Alzheimer's disease (intermediate level neuropathological change)\",\n",
       "       'Control, Cerebrovascular disease (cerebral amyloid angiopathy)',\n",
       "       'Nan'], dtype=object)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH['path_autopsy_dx_main'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LAST TWO entries are empty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['[\"Lewy body disease nos\", \"Parkinson\\'s disease\", \"Parkinson\\'s disease with dementia\", \"Dementia with Lewy bodies\", \"Multiple system atrophy (SND>OPCA)\", \"Multiple system atrophy (OPCA<SND)\", \"Multiple system atrophy (SND=OPCA)\", \"Progressive supranuclear palsy\", \"Corticobasal degeneration\", \"Globular glial tauoapathy (GGT)\", \"Chronic traumatic encephalopathy (CTE)\", \"FTLD-Tau (Pick\\'s)\", \"FTLD-Tau (MAPT)\", \"FTLD-Tau (AGD)\", \"FTLD-TDP43, Type A\", \"FTLD-TDP43, Type B\", \"FTLD-TDP43, Type C\", \"FTLD-TDP43, Type D\", \"FTLD-TDP43, Type E\", \"Motor neurone disease-TDP43 (MND or ALS)\", \"FTLD-MND-TDP43\", \"Huntington\\'s disease\", \"Spinocerebellar ataxia, nos\", \"Prion disease, nos\", \"Alzheimer\\'s disease (high level neuropathological change)\", \"Alzheimer\\'s disease (intermediate level neuropathological change)\", \"Control, Low level AD neuropathological change\", \"Control, Limbic predominant age-related TDP43 proteinopathy (LATE)\", \"Control, Argyrophilic grain disease\", \"Control, Primary age-related tauopathy (PART)\", \"Control, Ageing-related tau astrogliopathy (ARTAG)\", \"Control, Cerebrovascular disease (atherosclerosis)\", \"Control, Cerebrovascular disease (hyaline arteriolosclerosis)\", \"Control, Cerebrovascular disease (cerebral amyloid angiopathy)\", \"Control, no misfolded protein or significant vascular pathology\", \"Other neurological disorder\"]'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CDE_df[CDE_df[\"Field\"]==\"path_autopsy_dx_main\"].Validation.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "üö®‚ö†Ô∏è‚ùó **Missing Required Fields in CLINPATH: source_sample_id**\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 138/138 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 10/138 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 138/138 empty rows\n",
      "\n",
      "\t- path_year_death: 138/138 empty rows\n",
      "\n",
      "\t- brain_weight: 138/138 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 138/138 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*path_braak_nft*_:  invalid values üí©'<NA>'\n",
      "    - valid ‚û°Ô∏è '0', 'I', 'II', 'III', 'IV', 'V', 'VI', 'I/II', 'III/IV', 'IV/V', 'V/VI'\n",
      "- _*path_braak_asyn*_:  invalid values üí©'<NA>'\n",
      "    - valid ‚û°Ô∏è '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_nia_aa_a*_:  invalid values üí©'<NA>'\n",
      "    - valid ‚û°Ô∏è 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values üí©'<NA>'\n",
      "    - valid ‚û°Ô∏è 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values üí©'<NA>'\n",
      "    - valid ‚û°Ô∏è 'C0', 'C1', 'C2', 'C3'\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (<NA> or NaN):**\n",
      "- hx_dementia_mcihx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- APOE_e4_status \n",
      "- cognitive_status \n",
      "- path_autopsy_dx_main \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- known_pathogenic_mutation \n",
      "- path_mckeith \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# # write the clean metadata\n",
    "# STUDY.to_csv(data_path / \"metadata/STUDY.csv\")\n",
    "# PROTOCOL.to_csv(data_path / \"metadata/PROTOCOL.csv\")\n",
    "# CLINPATH.to_csv(data_path / \"metadata/CLINPATH.csv\")\n",
    "# SAMPLE.to_csv(data_path / \"metadata/SAMPLE.csv\")\n",
    "# SUBJECT.to_csv(data_path / \"metadata/SUBJECT.csv\")\n",
    "\n",
    "# # also writh them to clean...\n",
    "# \n",
    "#  \n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Hardy\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team-Jacobsson\n",
    "\n",
    "Look at the v1 versions of Team-Jacobsson... they mostly seem okay.  \n",
    "\n",
    "TODO: clean them up and use to cross-validate the `update_schema` tools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-jakobsson\")\n",
    "metadata_path = data_path / \"metadata_v1\"\n",
    "\n",
    "SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\")\n",
    "CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")\n",
    "STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\")\n",
    "PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\")\n",
    "SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *STUDY* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in STUDY. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(STUDY, \"STUDY\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_name</th>\n",
       "      <th>project_dataset</th>\n",
       "      <th>project_description</th>\n",
       "      <th>ASAP_team_name</th>\n",
       "      <th>ASAP_lab_name</th>\n",
       "      <th>PI_full_name</th>\n",
       "      <th>PI_email</th>\n",
       "      <th>contributor_names</th>\n",
       "      <th>submitter_name</th>\n",
       "      <th>submittor_email</th>\n",
       "      <th>...</th>\n",
       "      <th>publication_DOI</th>\n",
       "      <th>publication_PMID</th>\n",
       "      <th>number_of_brain_samples</th>\n",
       "      <th>brain_regions</th>\n",
       "      <th>types_of_samples</th>\n",
       "      <th>PI_ORCHID</th>\n",
       "      <th>PI_google_scholar_id</th>\n",
       "      <th>DUA_version</th>\n",
       "      <th>metadata_version_date</th>\n",
       "      <th>v1, SEPT2023</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Activation of transposable elements as a trigg...</td>\n",
       "      <td>Single nuclei sequencing of brain regions from...</td>\n",
       "      <td>A range of genetic, clinical and pathological ...</td>\n",
       "      <td>TEAM-JAKOBSSON</td>\n",
       "      <td>Gale-Hammell Lab</td>\n",
       "      <td>Molly, Gale Hammell</td>\n",
       "      <td>Molly.GaleHammell@nyulangone.org</td>\n",
       "      <td>Anita, Adami; Talitha, Forcier; Raquel, Garza;...</td>\n",
       "      <td>Oliver, H, Tam</td>\n",
       "      <td>oliver.tam@nyulangone.org</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71</td>\n",
       "      <td>Substantia nigra, prefrontal cortex, amygdala,...</td>\n",
       "      <td>Late stage PD &amp; control</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Creative Commons Attribution license CC BY 4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>v1_20231113</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows √ó 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        project_name  \\\n",
       "0  Activation of transposable elements as a trigg...   \n",
       "\n",
       "                                     project_dataset  \\\n",
       "0  Single nuclei sequencing of brain regions from...   \n",
       "\n",
       "                                 project_description  ASAP_team_name  \\\n",
       "0  A range of genetic, clinical and pathological ...  TEAM-JAKOBSSON   \n",
       "\n",
       "      ASAP_lab_name         PI_full_name                          PI_email  \\\n",
       "0  Gale-Hammell Lab  Molly, Gale Hammell  Molly.GaleHammell@nyulangone.org   \n",
       "\n",
       "                                   contributor_names  submitter_name  \\\n",
       "0  Anita, Adami; Talitha, Forcier; Raquel, Garza;...  Oliver, H, Tam   \n",
       "\n",
       "             submittor_email  ... publication_DOI  publication_PMID  \\\n",
       "0  oliver.tam@nyulangone.org  ...             NaN               NaN   \n",
       "\n",
       "   number_of_brain_samples                                      brain_regions  \\\n",
       "0                       71  Substantia nigra, prefrontal cortex, amygdala,...   \n",
       "\n",
       "          types_of_samples PI_ORCHID PI_google_scholar_id  \\\n",
       "0  Late stage PD & control       NaN                  NaN   \n",
       "\n",
       "                                      DUA_version  metadata_version_date  \\\n",
       "0  Creative Commons Attribution license CC BY 4.0                    NaN   \n",
       "\n",
       "  v1, SEPT2023  \n",
       "0  v1_20231113  \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "STUDY[metadata_version_date] = METADATA_VERSION_DATE\n",
    "STUDY.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- duration_pmi: 7/25 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 25/25 empty rows\n",
      "## Enum fields have valid values in SUBJECT. ü•≥\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- source_RIN: 339/339 empty rows\n",
      "\n",
      "\t- RIN: 339/339 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- pm_PH: 339/339 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*sequencing_length*_:  invalid values üí©'28', '91', '90', '8', '10'\n",
      "    - valid ‚û°Ô∏è '25', '50', '100', '150'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE_df, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "üö®‚ö†Ô∏è‚ùó **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- age_at_onset: 71/71 empty rows\n",
      "\n",
      "\t- age_at_diagnosis: 47/71 empty rows\n",
      "\n",
      "\t- first_motor_symptom: 71/71 empty rows\n",
      "\n",
      "\t- path_year_death: 71/71 empty rows\n",
      "\n",
      "\t- brain_weight: 71/71 empty rows\n",
      "üö®‚ö†Ô∏è‚ùó **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- smoking_years: 71/71 empty rows\n",
      "## Enums\n",
      "üö®‚ö†Ô∏è‚ùó **Invalid entries**\n",
      "- _*region_level_2*_:  invalid values üí©'Prefrontal cortex'\n",
      "    - valid ‚û°Ô∏è 'Superior frontal gyrus', 'Middle frontal gyrus', 'Inferior frontal gyrus', 'Superior temporal gyrus', 'Middle temporal gyrus', 'Inferior temporal gyrus', 'Fusiform gyrus', 'Transentorhinal region', 'Entorinal region', 'Subiculum', 'CA1-CA4', 'Amygdala', 'Periamygdala cortex', 'Anterior cingulate gyrus', 'Posterior cingulate gyrus', 'Superior parietal lobule', 'Inferior parietal lobule', 'Parastriate cortex', 'Peristriate cortex', 'Striate cortex', 'Insular cortex', 'Caudate nucleus', 'Putamen', 'Globus pallidus', 'Thalamus', 'Subthalamic nucleus', 'Substantia nigra', 'Pontine tegmentum', 'Pontine base', 'Medulla tegmentum', 'Medulla base', 'Cerebellar vermis', 'Cerebellar hemisphere', 'Dentate nucleus', 'Right colon', 'Left colon', 'Unknown'\n",
      "üö®‚ö†Ô∏è‚ùó **Found unexpected NULL (<NA> or NaN):**\n",
      "- hemispherefamily_history \n",
      "- hx_dementia_mci \n",
      "- hx_melanoma \n",
      "- education_level \n",
      "- smoking_status \n",
      "- APOE_e4_status \n",
      "- cognitive_status \n",
      "- path_braak_nft \n",
      "- path_braak_asyn \n",
      "- path_cerad \n",
      "- path_thal \n",
      "- known_pathogenic_mutation \n",
      "- path_mckeith \n",
      "- sn_neuronal_loss \n",
      "- path_infarcs \n",
      "- path_nia_ri \n",
      "- path_nia_aa_a \n",
      "- path_nia_aa_b \n",
      "- path_nia_aa_c \n",
      "- TDP43 \n",
      "- arteriolosclerosis_severity_scale \n",
      "- amyloid_angiopathy_severity_scale \n",
      "- path_ad_level\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-jakobsson\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "\n",
    "CLINPATHv2 = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>region_level_1</th>\n",
       "      <th>region_level_2</th>\n",
       "      <th>region_level_3</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>family_history</th>\n",
       "      <th>...</th>\n",
       "      <th>sn_neuronal_loss</th>\n",
       "      <th>path_infarcs</th>\n",
       "      <th>path_nia_ri</th>\n",
       "      <th>path_nia_aa_a</th>\n",
       "      <th>path_nia_aa_b</th>\n",
       "      <th>path_nia_aa_c</th>\n",
       "      <th>TDP43</th>\n",
       "      <th>arteriolosclerosis_severity_scale</th>\n",
       "      <th>amyloid_angiopathy_severity_scale</th>\n",
       "      <th>path_ad_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASAP1_PD_NP16-162_SN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASAP2_PD_NP16-25_SN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASAP3_PD_P73_SN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASAP4_PD_P74_SN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASAP5_Ctl_NP16-161_SN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Midbrain</td>\n",
       "      <td>Substantia nigra</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows √ó 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample_id  source_sample_id  time_from_baseline  GP2_id  \\\n",
       "0   ASAP1_PD_NP16-162_SN               NaN                   0     NaN   \n",
       "1    ASAP2_PD_NP16-25_SN               NaN                   0     NaN   \n",
       "2        ASAP3_PD_P73_SN               NaN                   0     NaN   \n",
       "3        ASAP4_PD_P74_SN               NaN                   0     NaN   \n",
       "4  ASAP5_Ctl_NP16-161_SN               NaN                   0     NaN   \n",
       "\n",
       "   hemisphere region_level_1    region_level_2 region_level_3  AMPPD_id  \\\n",
       "0         NaN       Midbrain  Substantia nigra        Unknown       NaN   \n",
       "1         NaN       Midbrain  Substantia nigra        Unknown       NaN   \n",
       "2         NaN       Midbrain  Substantia nigra        Unknown       NaN   \n",
       "3         NaN       Midbrain  Substantia nigra        Unknown       NaN   \n",
       "4         NaN       Midbrain  Substantia nigra        Unknown       NaN   \n",
       "\n",
       "   family_history  ...  sn_neuronal_loss  path_infarcs  path_nia_ri  \\\n",
       "0             NaN  ...               NaN           NaN          NaN   \n",
       "1             NaN  ...               NaN           NaN          NaN   \n",
       "2             NaN  ...               NaN           NaN          NaN   \n",
       "3             NaN  ...               NaN           NaN          NaN   \n",
       "4             NaN  ...               NaN           NaN          NaN   \n",
       "\n",
       "   path_nia_aa_a  path_nia_aa_b  path_nia_aa_c  TDP43  \\\n",
       "0            NaN            NaN            NaN    NaN   \n",
       "1            NaN            NaN            NaN    NaN   \n",
       "2            NaN            NaN            NaN    NaN   \n",
       "3            NaN            NaN            NaN    NaN   \n",
       "4            NaN            NaN            NaN    NaN   \n",
       "\n",
       "   arteriolosclerosis_severity_scale  amyloid_angiopathy_severity_scale  \\\n",
       "0                                NaN                                NaN   \n",
       "1                                NaN                                NaN   \n",
       "2                                NaN                                NaN   \n",
       "3                                NaN                                NaN   \n",
       "4                                NaN                                NaN   \n",
       "\n",
       "   path_ad_level  \n",
       "0            NaN  \n",
       "1            NaN  \n",
       "2            NaN  \n",
       "3            NaN  \n",
       "4            NaN  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CLINPATH.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "added empty column dig_slide_avail to CLINPATH table\n",
      "added empty column quant_path_avail to CLINPATH table\n",
      "added empty column DV200 to SAMPLE table\n"
     ]
    }
   ],
   "source": [
    "# force all the tables to match the CDE\n",
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDEv1)\n",
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDEv1)\n",
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDEv1)\n",
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDEv1)\n",
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDEv1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "export_root = Path.cwd() / \"clean/team-Jakobsson\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scverse10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
