{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ASAP CRN Metadata validation - wave 2\n",
    "\n",
    "# Team Scherzer. ASAP CRN Metadata validation - wave 2\n",
    "\n",
    "28 Nov 2023\n",
    "Andy Henrie\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Streamlit NOT successfully\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from pathlib import Path\n",
    "\n",
    "from utils.io import read_meta_table, get_dtypes_dict, ReportCollector\n",
    "from utils.qcutils import prep_table, validate_table, validate_file, read_ASAP_CDE, reorder_table_to_CDE\n",
    "\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean Tables\n",
    "Some of the coding is wrong and easily fixed.  Some is not.  \n",
    "\n",
    "TODO: share back details to Team Scherzer for better metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/ergonyc/Projects/ASAP/meta-clean\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [raw-admin-scherzer@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/scherzer-credentials.json  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "WARNING: You have requested checksumming but your crcmod installation isn't\n",
      "using the module's C extension, so checksumming will run very slowly. For help\n",
      "installing the extension, please see \"gsutil help crcmod\".\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# !gcloud storage hash \"gs://asap-raw-data-team-scherzer/**/*.gz\"  --skip-crc32c --hex  --billing-project dnastack-asap-parkinsons > scherzer_hexhash.log\n",
    "\n",
    "!gsutil -u dnastack-asap-parkinsons hash -h \"gs://asap-raw-data-team-scherzer/**/*.gz\" > scherzer_hexhash.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/ASAP_CDE-ScherzerTeam.xlsx [Content-Type=application/vnd.openxmlformats-officedocument.spreadsheetml.sheet]...\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/SUBJECT.csv [Content-Type=text/csv]...\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/SAMPLE.csv [Content-Type=text/csv]...\n",
      "- [4 files][540.2 KiB/540.2 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/DATA.csv [Content-Type=text/csv]...\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023/PROTOCOL.csv [Content-Type=text/csv]...\n",
      "| [7 files][  1.1 MiB/  1.1 MiB]                                                \n",
      "Operation completed over 7 objects/1.1 MiB.                                      \n"
     ]
    }
   ],
   "source": [
    "!gsutil -u dnastack-asap-parkinsons cp -r \"/Users/ergonyc/Projects/ASAP/team-scherzer/Scherzer_raw_meta11202023\" \"gs://asap-raw-data-team-scherzer/metadata/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     35161  2023-11-22T01:47:41Z  gs://asap-raw-data-team-scherzer/metadata/CLINPATH.csv\n",
      "    595701  2023-11-22T01:47:41Z  gs://asap-raw-data-team-scherzer/metadata/DATA.csv\n",
      "      1165  2023-11-22T01:47:40Z  gs://asap-raw-data-team-scherzer/metadata/PROTOCOL.csv\n",
      "     32705  2023-11-22T01:47:40Z  gs://asap-raw-data-team-scherzer/metadata/SAMPLE.csv\n",
      "      1684  2023-11-22T01:47:40Z  gs://asap-raw-data-team-scherzer/metadata/STUDY.csv\n",
      "     15648  2023-11-22T01:47:40Z  gs://asap-raw-data-team-scherzer/metadata/SUBJECT.csv\n",
      "                                  gs://asap-raw-data-team-scherzer/metadata/Scherzer_raw_meta11202023/\n",
      "                                  gs://asap-raw-data-team-scherzer/metadata/v2/\n",
      "TOTAL: 6 objects, 682064 bytes (666.08kiB)\n"
     ]
    }
   ],
   "source": [
    "# !gcloud storage ls -l \"gs://asap-raw-data-team-scherzer/metadata\"  --billing-project dnastack-asap-parkinsons "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Team-Scherzer\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read url\n"
     ]
    }
   ],
   "source": [
    "## convert to seurat Object\n",
    "\n",
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-scherzer\")\n",
    "metadata_path = data_path / \"metadata\"\n",
    "\n",
    "metadata_version = \"v2\"\n",
    "\n",
    "CDE_df,dtypes_dict = read_ASAP_CDE(metadata_version=metadata_version)\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\"\n",
    "\n",
    "\n",
    "## convert \n",
    "data_path = Path.home() / (\"Projects/ASAP/team-scherzer\")\n",
    "metadata_path = data_path / \"Scherzer_raw_meta11202023\"\n",
    "\n",
    "SUBJECT = read_meta_table(metadata_path / \"SUBJECT.csv\", dtypes_dict)\n",
    "CLINPATH = read_meta_table(metadata_path / \"CLINPATH.csv\", dtypes_dict)\n",
    "STUDY = read_meta_table(metadata_path / \"STUDY.csv\", dtypes_dict)\n",
    "PROTOCOL = read_meta_table(metadata_path / \"PROTOCOL.csv\", dtypes_dict)\n",
    "SAMPLE = read_meta_table(metadata_path / \"SAMPLE.csv\", dtypes_dict)\n",
    "DATA = read_meta_table(metadata_path / \"DATA.csv\", dtypes_dict)\n",
    "\n",
    "\n",
    "\n",
    "# # excel_path = data_path / \"ASAP_CDE_team_Scherzer.v11152023.xlsx\"\n",
    "# excel_path = data_path / \"Scherzer_raw_meta11202023/ASAP_CDE-ScherzerTeam.xlsx\"\n",
    "# STUDY = pd.read_excel(excel_path,sheet_name=\"STUDY\")#,header=1)#.drop(columns=\"Field\")\n",
    "# CLINPATH = pd.read_excel(excel_path,sheet_name=\"CLINPATH\")#,header=1)#.drop(columns=\"Field\")\n",
    "# SUBJECT = pd.read_excel(excel_path,sheet_name=\"SUBJECT\")#,header=1)#.drop(columns=\"Field\")\n",
    "# SAMPLE = pd.read_excel(excel_path,sheet_name=\"SAMPLE\")#,header=1)#.drop(columns=\"Field\")\n",
    "# PROTOCOL = pd.read_excel(excel_path,sheet_name=\"PROTOCOL\")#,header=1)#.drop(columns=\"Field\")\n",
    "# DATA = pd.read_excel(excel_path,sheet_name=\"DATA\")#,header=1)#.drop(columns=\"Field\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "STUDY = prep_table(STUDY, CDE_df)\n",
    "\n",
    "\n",
    "STUDY['metadata_version_date'] = METADATA_VERSION_DATE\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *STUDY* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in STUDY. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "STUDY = reorder_table_to_CDE(STUDY, \"STUDY\", CDE_df)\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(STUDY, \"STUDY\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *PROTOCOL* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in PROTOCOL. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "PROTOCOL = reorder_table_to_CDE(PROTOCOL, \"PROTOCOL\", CDE_df)\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(PROTOCOL, \"PROTOCOL\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_subject_id</th>\n",
       "      <th>AMPPD_id</th>\n",
       "      <th>GP2_id</th>\n",
       "      <th>biobank_name</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>age_at_collection</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>...</th>\n",
       "      <th>hx_dementia_mci</th>\n",
       "      <th>hx_melanoma</th>\n",
       "      <th>education_level</th>\n",
       "      <th>smoking_status</th>\n",
       "      <th>smoking_years</th>\n",
       "      <th>APOE_e4_status</th>\n",
       "      <th>cognitive_status</th>\n",
       "      <th>time_from_baseline</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_diagnosis_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BN0009</td>\n",
       "      <td>00-09</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>64</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>unknown</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>30</td>\n",
       "      <td>33</td>\n",
       "      <td>Dementia</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BN0329</td>\n",
       "      <td>03-29</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>79</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>yes</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>14</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>23</td>\n",
       "      <td>Dementia</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Idiopathic PD</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BN0339</td>\n",
       "      <td>03-39</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>86</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>12</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>50</td>\n",
       "      <td>33</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BN0341</td>\n",
       "      <td>03-41</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Male</td>\n",
       "      <td>89</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>16</td>\n",
       "      <td>Never</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BN0347</td>\n",
       "      <td>03-47</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>BSHRI</td>\n",
       "      <td>Human</td>\n",
       "      <td>Female</td>\n",
       "      <td>95</td>\n",
       "      <td>White</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>...</td>\n",
       "      <td>no</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>12</td>\n",
       "      <td>Former smoker</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>33</td>\n",
       "      <td>MCI</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Healthy Control</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  subject_id source_subject_id AMPPD_id GP2_id biobank_name organism     sex  \\\n",
       "0     BN0009             00-09     <NA>   <NA>        BSHRI    Human    Male   \n",
       "1     BN0329             03-29     <NA>   <NA>        BSHRI    Human    Male   \n",
       "2     BN0339             03-39     <NA>   <NA>        BSHRI    Human    Male   \n",
       "3     BN0341             03-41     <NA>   <NA>        BSHRI    Human    Male   \n",
       "4     BN0347             03-47     <NA>   <NA>        BSHRI    Human  Female   \n",
       "\n",
       "   age_at_collection   race ethnicity  ... hx_dementia_mci hx_melanoma  \\\n",
       "0                 64  White   Unknown  ...             yes     Unknown   \n",
       "1                 79  White   Unknown  ...             yes     Unknown   \n",
       "2                 86  White   Unknown  ...              no     Unknown   \n",
       "3                 89  White   Unknown  ...              no     Unknown   \n",
       "4                 95  White   Unknown  ...              no     Unknown   \n",
       "\n",
       "  education_level smoking_status smoking_years APOE_e4_status  \\\n",
       "0         unknown  Former smoker            30             33   \n",
       "1              14        Unknown       Unknown             23   \n",
       "2              12  Former smoker            50             33   \n",
       "3              16          Never             0             34   \n",
       "4              12  Former smoker       Unknown             33   \n",
       "\n",
       "  cognitive_status time_from_baseline primary_diagnosis primary_diagnosis_text  \n",
       "0         Dementia            Unknown     Idiopathic PD                    NaN  \n",
       "1         Dementia            Unknown     Idiopathic PD                    NaN  \n",
       "2           Normal            Unknown   Healthy Control                    NaN  \n",
       "3           Normal            Unknown   Healthy Control                    NaN  \n",
       "4              MCI            Unknown   Healthy Control                    NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SUBJECT.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT = prep_table(SUBJECT, CDE_df)\n",
    "\n",
    "SUBJECT['hx_melanoma'] = SUBJECT['hx_melanoma'].replace({'Present':'Yes', 'Unknown': pd.NA})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "SUBJECT['cognitive_status'] = SUBJECT['cognitive_status'].replace({'Unknown': pd.NA })\n",
    "\n",
    "# SUBJECT.replace(\"Unknown\", pd.NA, inplace=True)\n",
    "\n",
    "# eductation levels \n",
    "SUBJECT['education_level'] = SUBJECT['education_level'].replace({'Unknown': pd.NA })\n",
    "\n",
    "# map \"Unknown\" to pd.NA\n",
    "\n",
    "# map ['14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'] ???"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SUBJECT* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- primary_diagnosis_text: 94/94 empty rows\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "- _*hx_melanoma*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*education_level*_:  invalid values 💩'<NA>', '14', '12', '16', '18', '13', '8', '15', '20', '19', '17', '11', '22', '10', '21'\n",
      "    - valid ➡️ 'High School', 'High School/GED', 'Some college without degree', 'Associate degree college', 'Bachelor's degree', 'Master's degree', 'Professional or doctoral degree', 'Refuse', 'Other'\n",
      "- _*cognitive_status*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'Normal', 'MCI', 'Dementia'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SUBJECT = reorder_table_to_CDE(SUBJECT, \"SUBJECT\", CDE_df)\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SUBJECT, \"SUBJECT\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>cellranger_outs_folder</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0329_MTG_snRNAseq_rep1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch2_ASAP_...</td>\n",
       "      <td>BN0329</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC_BN0339_MTG_snRNAseq_rep1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>BN0339</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC_BN0341_MTG_snRNAseq_rep1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...</td>\n",
       "      <td>BN0341</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC_BN0347_MTG_snRNAseq_rep1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>/data/neurogen/ASAP/scRNAseq/data/batch2_ASAP_...</td>\n",
       "      <td>BN0347</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample_id source_sample_id  \\\n",
       "0  PD_BN0009_MTG_snRNAseq_rep1             <NA>   \n",
       "1  PD_BN0329_MTG_snRNAseq_rep1             <NA>   \n",
       "2  HC_BN0339_MTG_snRNAseq_rep1             <NA>   \n",
       "3  HC_BN0341_MTG_snRNAseq_rep1             <NA>   \n",
       "4  HC_BN0347_MTG_snRNAseq_rep1             <NA>   \n",
       "\n",
       "                              cellranger_outs_folder subject_id replicate  \\\n",
       "0  /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...     BN0009      rep1   \n",
       "1  /data/neurogen/ASAP/scRNAseq/data/batch2_ASAP_...     BN0329      rep1   \n",
       "2  /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...     BN0339      rep1   \n",
       "3  /data/neurogen/ASAP/scRNAseq/data/batch1_ASAP_...     BN0341      rep1   \n",
       "4  /data/neurogen/ASAP/scRNAseq/data/batch2_ASAP_...     BN0347      rep1   \n",
       "\n",
       "   replicate_count  repeated_sample   batch tissue           brain_region  \\\n",
       "0                1                0  batch1  Brain  Middle temporal gyrus   \n",
       "1                1                0  batch2  Brain  Middle temporal gyrus   \n",
       "2                1                0  batch1  Brain  Middle temporal gyrus   \n",
       "3                1                0  batch1  Brain  Middle temporal gyrus   \n",
       "4                1                0  batch2  Brain  Middle temporal gyrus   \n",
       "\n",
       "   ...   sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0  ...    PATO:0000384 (male)                                     <NA>   \n",
       "1  ...    PATO:0000384 (male)                                     <NA>   \n",
       "2  ...    PATO:0000384 (male)                                     <NA>   \n",
       "3  ...    PATO:0000384 (male)                                     <NA>   \n",
       "4  ...  PATO:0000383 (female)                                     <NA>   \n",
       "\n",
       "  disease_ontology_term_id tissue_ontology_term_id  \\\n",
       "0                     <NA>                    <NA>   \n",
       "1                     <NA>                    <NA>   \n",
       "2                     <NA>                    <NA>   \n",
       "3                     <NA>                    <NA>   \n",
       "4                     <NA>                    <NA>   \n",
       "\n",
       "   cell_type_ontology_term_id  \\\n",
       "0                        <NA>   \n",
       "1                        <NA>   \n",
       "2                        <NA>   \n",
       "3                        <NA>   \n",
       "4                        <NA>   \n",
       "\n",
       "                              assay_ontology_term_id suspension_type  DV200  \\\n",
       "0  EFO:0030003 '10x 3' v3 transcription profiling...            <NA>    NaN   \n",
       "1  EFO:0030003 '10x 3' v3 transcription profiling...            <NA>    NaN   \n",
       "2  EFO:0030003 '10x 3' v3 transcription profiling...            <NA>    NaN   \n",
       "3  EFO:0030003 '10x 3' v3 transcription profiling...            <NA>    NaN   \n",
       "4  EFO:0030003 '10x 3' v3 transcription profiling...            <NA>    NaN   \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN     <NA>  \n",
       "1   NaN     <NA>  \n",
       "2   NaN     <NA>  \n",
       "3   NaN     <NA>  \n",
       "4   NaN     <NA>  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE = prep_table(SAMPLE, CDE_df)\n",
    "\n",
    "SAMPLE['region_level_1'] = SAMPLE['region_level_1'].replace({'Temporal Lobe':'Temporal lobe'})\n",
    "SAMPLE['molecular_source'] = SAMPLE['molecular_source'].replace({'PolyA_RNA':'PolyA RNA'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "subj_sex_mapper = dict(zip(SUBJECT['subject_id'], SUBJECT['sex']))\n",
    "SAMPLE['sex_ontology_term_id'] = SAMPLE['subject_id'].map(subj_sex_mapper).replace({\"Male\":\"PATO:0000384 (male)\", \"Female\":\"PATO:0000383 (female)\" })\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ERROR in metadata\n",
    "\n",
    "'source_sample_id' cannot be the same as 'subject_id'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "SAMPLE['source_sample_id'] = pd.NA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *SAMPLE* table.\n",
      "No empty entries (Nan) found in _Required_ fields.\n",
      "🚨⚠️❗ **Optional Fields with Empty (nan) values:**\n",
      "\n",
      "\t- DV200: 97/97 empty rows\n",
      "\n",
      "\t- pm_PH: 97/97 empty rows\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "- _*sequencing_length*_:  invalid values 💩'120'\n",
      "    - valid ➡️ '25', '50', '100', '150'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "SAMPLE = reorder_table_to_CDE(SAMPLE, \"SAMPLE\", CDE_df)\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(SAMPLE, \"SAMPLE\", CDE_df, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "CLINPATH = prep_table(CLINPATH, CDE_df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# autopsy coding is too too idosyncratic.  leave \n",
    "# the path_nia_aa_a are coded as path_nia_ri... swap them \n",
    "\n",
    "\n",
    "\n",
    "# 'Low', 'Intermediate', 'High', 'None'\n",
    "CLINPATH['path_nia_ri'] = CLINPATH['path_nia_aa_a'].replace({'Low ADNC': 'Low', \n",
    "                                                            'Not AD': 'None', \n",
    "                                                            'Intermediate ADNC': 'Intermediate'})\n",
    "CLINPATH['path_nia_aa_a'] = pd.NA\n",
    "\n",
    "\n",
    "# path_thal\n",
    "CLINPATH['path_thal'].replace({'Phase 0 (A0)':'0', \n",
    "                                'Phase 1 (A1)':'1', \n",
    "                                'Phase 2 (A1)':'2', \n",
    "                                'Phase 3 (A2)':'3', \n",
    "                                'Phase 4 (A3)':'4',\n",
    "                                'Phase 5 (A3)':'5', \n",
    "                                'Unknown':pd.NA, \n",
    "                                'Missing/unknown':pd.NA }, inplace=True)\n",
    "\n",
    "\n",
    "CLINPATH['path_infarcs'].replace('Unknown', pd.NA, inplace=True)\n",
    "\n",
    "\n",
    "# no idea what No, undx, yes maps to.  leave\n",
    "CLINPATH['TDP43'].replace('Unknown', pd.NA, inplace=True)\n",
    "\n",
    "# sn_neuronal_loss has alternate 0-3 coding.  leave"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE_df)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# autopsy coding is too too idosyncratic.  leave \n",
    "# the path_nia_aa_a are coded as path_nia_ri... swap them \n",
    "\n",
    "\n",
    "\n",
    "# 'Low', 'Intermediate', 'High', 'None'\n",
    "CLINPATH['path_nia_ri'] = CLINPATH['path_nia_aa_a'].replace({'Low ADNC': 'Low', \n",
    "                                                            'Not AD': 'None', \n",
    "                                                            'Intermediate ADNC': 'Intermediate'})\n",
    "CLINPATH['path_nia_aa_a'] = pd.NA\n",
    "\n",
    "\n",
    "# path_thal\n",
    "CLINPATH['path_thal'].replace({'Phase 0 (A0)':'0', \n",
    "                                'Phase 1 (A1)':'1', \n",
    "                                'Phase 2 (A1)':'2', \n",
    "                                'Phase 3 (A2)':'3', \n",
    "                                'Phase 4 (A3)':'4',\n",
    "                                'Phase 5 (A3)':'5', \n",
    "                                'Unknown':pd.NA, \n",
    "                                'Missing/unknown':pd.NA }, inplace=True)\n",
    "\n",
    "\n",
    "CLINPATH['path_infarcs'].replace('Unknown', pd.NA, inplace=True)\n",
    "\n",
    "\n",
    "# no idea what No, undx, yes maps to.  leave\n",
    "CLINPATH['TDP43'].replace('Unknown', pd.NA, inplace=True)\n",
    "\n",
    "# sn_neuronal_loss has alternate 0-3 coding.  leave"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mckeith_map = {'l. Olfactory Bulb-Only':\"Olfactory bulb only\", 'lla. Brainstem Predominant':\"Brainstem\",\n",
    "       'llb. Limbic Predominant':\"Limbic (transitional)\", 'lV. Neocortical':\"Neocortical\",\n",
    "       'lll. Brainstem/Limbic':\"Amygdala Predominant\", '0. No Lewy bodies':\"Absent\", \"Unknown\":pd.NA}\n",
    "\n",
    "\n",
    "CLINPATH['path_mckeith'] = CLINPATH['path_mckeith'].replace(mckeith_map)\n",
    "\n",
    "# leave path_nia_ri like this for now. not sure how to map \"criteria not met\" and \"Not AD\"\n",
    "\n",
    "# leave amyloid_angiopathy_severity_scale like this for now. not sure how to map 'Cerebral amyloid angiopathy, temporal and occipital lobe','Cerebral amyloid angiopathy, frontal lobe']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *CLINPATH* table.\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- path_year_death: 2/94 empty rows\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enums\n",
      "🚨⚠️❗ **Invalid entries**\n",
      "- _*path_autopsy_dx_main*_:  invalid values 💩'PD/Dem', 'Control', 'Control (MCI)', 'PD/ND', '<NA>', 'Control, Non-motoric', 'PDD', 'MCI', 'PD (MCI)', 'Control (history)', 'Control (diseased)', 'MCI (clinical history)', 'Control (clinical history)'\n",
      "    - valid ➡️ 'Lewy body disease nos', 'Parkinson's disease', 'Parkinson's disease with dementia', 'Dementia with Lewy bodies', 'Multiple system atrophy (SND>OPCA)', 'Multiple system atrophy (OPCA<SND)', 'Multiple system atrophy (SND=OPCA)', 'Progressive supranuclear palsy', 'Corticobasal degeneration', 'Globular glial tauoapathy (GGT)', 'Chronic traumatic encephalopathy (CTE)', 'FTLD-Tau (Pick's)', 'FTLD-Tau (MAPT)', 'FTLD-Tau (AGD)', 'FTLD-TDP43, Type A', 'FTLD-TDP43, Type B', 'FTLD-TDP43, Type C', 'FTLD-TDP43, Type D', 'FTLD-TDP43, Type E', 'Motor neurone disease-TDP43 (MND or ALS)', 'FTLD-MND-TDP43', 'Huntington's disease', 'Spinocerebellar ataxia, nos', 'Prion disease, nos', 'Alzheimer's disease (high level neuropathological change)', 'Alzheimer's disease (intermediate level neuropathological change)', 'Control, Low level AD neuropathological change', 'Control, Limbic predominant age-related TDP43 proteinopathy (LATE)', 'Control, Argyrophilic grain disease', 'Control, Primary age-related tauopathy (PART)', 'Control, Ageing-related tau astrogliopathy (ARTAG)', 'Control, Cerebrovascular disease (atherosclerosis)', 'Control, Cerebrovascular disease (hyaline arteriolosclerosis)', 'Control, Cerebrovascular disease (cerebral amyloid angiopathy)', 'Control, no misfolded protein or significant vascular pathology', 'Other neurological disorder'\n",
      "- _*path_braak_asyn*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '6', '1/2', '3/4', '4/5', '5/6'\n",
      "- _*path_thal*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ '0', '1', '2', '3', '4', '5', '1/2', '3', '4/5'\n",
      "- _*path_mckeith*_:  invalid values 💩'Llb. Limbic Predominant', 'Lll. Brainstem/Limbic', 'Lla. Brainstem Predominant', 'LV. Neocortical', 'L. Olfactory Bulb-Only', '<NA>'\n",
      "    - valid ➡️ 'Neocortical', 'Limbic (transitional)', 'Brainstem', 'Amygdala Predominant', 'Olfactory bulb only', 'Limbic, transitional (brainstem and limbic involvement)', 'Diffuse, neocortical (brainstem, limbic and neocortical involvement)', 'Olfactory Bulb-Only', 'Limbic (amygdala) predominant', 'Absent', 'Present, but extent unknown'\n",
      "- _*sn_neuronal_loss*_:  invalid values 💩'3', '0', '1', '2', '<NA>'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*path_infarcs*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*path_nia_ri*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'Low', 'Intermediate', 'High', 'None'\n",
      "- _*path_nia_aa_a*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'A0', 'A1', 'A2', 'A3'\n",
      "- _*path_nia_aa_b*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'B0', 'B1', 'B2', 'B3'\n",
      "- _*path_nia_aa_c*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'C0', 'C1', 'C2', 'C3'\n",
      "- _*TDP43*_:  invalid values 💩'<NA>', 'No', 'Undx', 'Yes'\n",
      "    - valid ➡️ 'None in medial temporal lobe', 'Present in amygdala, only', 'Present in hippocampus, only', 'Present in amygdala and hippocampus, only', 'Present in medial temporal lobe and middle frontal gyrus (not FTLD pattern)', 'Unknown'\n",
      "- _*arteriolosclerosis_severity_scale*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*amyloid_angiopathy_severity_scale*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'None', 'Mild', 'Moderate', 'Severe', 'Not assessed', 'Unknown'\n",
      "- _*dig_slide_avail*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "- _*quant_path_avail*_:  invalid values 💩'<NA>'\n",
      "    - valid ➡️ 'Yes', 'No'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "CLINPATH = reorder_table_to_CDE(CLINPATH, \"CLINPATH\", CDE_df)\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(CLINPATH, \"CLINPATH\", CDE_df, report)\n",
    "report.print_log()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>time</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz</td>\n",
       "      <td>Index 1 Lane 1</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz</td>\n",
       "      <td>Index 2 Lane 1</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz</td>\n",
       "      <td>Read 1 Lane 1</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz</td>\n",
       "      <td>Read 2 Lane 1</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz</td>\n",
       "      <td>Index 1 Lane 2</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample_id replicate  replicate_count  repeated_sample  \\\n",
       "0  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "1  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "2  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "3  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "4  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "\n",
       "    batch file_type                                         file_name  \\\n",
       "0  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz   \n",
       "1  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz   \n",
       "2  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz   \n",
       "3  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz   \n",
       "4  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz   \n",
       "\n",
       "  file_description                          file_MD5 technology omic  \\\n",
       "0   Index 1 Lane 1  bc037ea92d5359f3da1ba6da4f974e1c         SN  RNA   \n",
       "1   Index 2 Lane 1  0c8ef844b76212e82428af03d341048b         SN  RNA   \n",
       "2    Read 1 Lane 1  c6a41f29849152df2ba2391dc559149b         SN  RNA   \n",
       "3    Read 2 Lane 1  4d610fdf5a72c8206b2f1505d8fcecd7         SN  RNA   \n",
       "4   Index 1 Lane 2  e34917b3ec01cd09ea8b52c82db1dc83         SN  RNA   \n",
       "\n",
       "  adjustment content  time header annotation configuration_file  \n",
       "0        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "1        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "2        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "3        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "4        Raw   Reads   NaN   <NA>       <NA>               <NA>  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All required fields are present in *DATA* table.\n",
      "🚨⚠️❗ **Required Fields with Empty (nan) values:**\n",
      "\n",
      "\t- time: 1552/1552 empty rows\n",
      "No empty entries (Nan) found in _Optional_ fields.\n",
      "## Enum fields have valid values in DATA. 🥳\n",
      "\n"
     ]
    }
   ],
   "source": [
    "DATA_ = DATA.copy()\n",
    "DATA = reorder_table_to_CDE(DATA, \"DATA\", CDE_df)\n",
    "\n",
    "report = ReportCollector(destination=\"\")\n",
    "validate_table(DATA, \"DATA\", CDE_df, report)\n",
    "report.print_log()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One last scrub of the columns "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# export_root = metadata_path / \"clean\"\n",
    "# if not export_root.exists():\n",
    "#     export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "\n",
    "export_root = Path.cwd() / \"clean/team-Scherzer\"\n",
    "if not export_root.exists():\n",
    "    export_root.mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "STUDY.to_csv( export_root / \"STUDY.csv\")\n",
    "PROTOCOL.to_csv(export_root / \"PROTOCOL.csv\")\n",
    "SAMPLE.to_csv(export_root / \"SAMPLE.csv\")\n",
    "SUBJECT.to_csv(export_root / \"SUBJECT.csv\")\n",
    "CLINPATH.to_csv(export_root / \"CLINPATH.csv\")\n",
    "DATA.to_csv(export_root / \"DATA.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_path = export_root"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "read url\n"
     ]
    }
   ],
   "source": [
    "metadata_version = \"v2\"\n",
    "\n",
    "CDE_df,dtypes_dict = read_ASAP_CDE(metadata_version=metadata_version)\n",
    "METADATA_VERSION_DATE = f\"{metadata_version}_{pd.Timestamp.now().strftime('%Y%m%d')}\"\n",
    "\n",
    "\n",
    "## convert \n",
    "\n",
    "export_root\n",
    "\n",
    "SUBJECT = read_meta_table(metadata_path / \"SUBJECT.csv\", dtypes_dict)\n",
    "CLINPATH = read_meta_table(metadata_path / \"CLINPATH.csv\", dtypes_dict)\n",
    "STUDY = read_meta_table(metadata_path / \"STUDY.csv\", dtypes_dict)\n",
    "PROTOCOL = read_meta_table(metadata_path / \"PROTOCOL.csv\", dtypes_dict)\n",
    "SAMPLE = read_meta_table(metadata_path / \"SAMPLE.csv\", dtypes_dict)\n",
    "DATA = read_meta_table(metadata_path / \"DATA.csv\", dtypes_dict)\n",
    "\n",
    "# SUBJECT = pd.read_csv(f\"{metadata_path}/SUBJECT.csv\")\n",
    "# CLINPATH = pd.read_csv(f\"{metadata_path}/CLINPATH.csv\")\n",
    "# STUDY = pd.read_csv(f\"{metadata_path}/STUDY.csv\")\n",
    "# PROTOCOL = pd.read_csv(f\"{metadata_path}/PROTOCOL.csv\")\n",
    "# SAMPLE = pd.read_csv(f\"{metadata_path}/SAMPLE.csv\")\n",
    "# DATA = pd.read_csv(f\"{metadata_path}/DATA.csv\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer')"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>source_sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>tissue</th>\n",
       "      <th>brain_region</th>\n",
       "      <th>hemisphere</th>\n",
       "      <th>...</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>cell_type_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>DV200</th>\n",
       "      <th>pm_PH</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0009</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Left</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0329_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0329</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Left</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC_BN0339_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0339</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Left</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC_BN0341_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0341</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Left</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000384 (male)</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC_BN0347_MTG_snRNAseq_rep1</td>\n",
       "      <td>BN0347</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch2</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Middle temporal gyrus</td>\n",
       "      <td>Left</td>\n",
       "      <td>...</td>\n",
       "      <td>PATO:0000383 (female)</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>Nan</td>\n",
       "      <td>EFO:0030003 '10x 3' v3 transcription profiling...</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample_id subject_id source_sample_id replicate  \\\n",
       "0  PD_BN0009_MTG_snRNAseq_rep1     BN0009             <NA>      rep1   \n",
       "1  PD_BN0329_MTG_snRNAseq_rep1     BN0329             <NA>      rep1   \n",
       "2  HC_BN0339_MTG_snRNAseq_rep1     BN0339             <NA>      rep1   \n",
       "3  HC_BN0341_MTG_snRNAseq_rep1     BN0341             <NA>      rep1   \n",
       "4  HC_BN0347_MTG_snRNAseq_rep1     BN0347             <NA>      rep1   \n",
       "\n",
       "   replicate_count  repeated_sample   batch tissue           brain_region  \\\n",
       "0                1                0  batch1  Brain  Middle temporal gyrus   \n",
       "1                1                0  batch2  Brain  Middle temporal gyrus   \n",
       "2                1                0  batch1  Brain  Middle temporal gyrus   \n",
       "3                1                0  batch1  Brain  Middle temporal gyrus   \n",
       "4                1                0  batch2  Brain  Middle temporal gyrus   \n",
       "\n",
       "  hemisphere  ...   sex_ontology_term_id  \\\n",
       "0       Left  ...    PATO:0000384 (male)   \n",
       "1       Left  ...    PATO:0000384 (male)   \n",
       "2       Left  ...    PATO:0000384 (male)   \n",
       "3       Left  ...    PATO:0000384 (male)   \n",
       "4       Left  ...  PATO:0000383 (female)   \n",
       "\n",
       "  self_reported_ethnicity_ontology_term_id disease_ontology_term_id  \\\n",
       "0                                      Nan                      Nan   \n",
       "1                                      Nan                      Nan   \n",
       "2                                      Nan                      Nan   \n",
       "3                                      Nan                      Nan   \n",
       "4                                      Nan                      Nan   \n",
       "\n",
       "   tissue_ontology_term_id  cell_type_ontology_term_id  \\\n",
       "0                      Nan                         Nan   \n",
       "1                      Nan                         Nan   \n",
       "2                      Nan                         Nan   \n",
       "3                      Nan                         Nan   \n",
       "4                      Nan                         Nan   \n",
       "\n",
       "                              assay_ontology_term_id  suspension_type DV200  \\\n",
       "0  EFO:0030003 '10x 3' v3 transcription profiling...              Nan   NaN   \n",
       "1  EFO:0030003 '10x 3' v3 transcription profiling...              Nan   NaN   \n",
       "2  EFO:0030003 '10x 3' v3 transcription profiling...              Nan   NaN   \n",
       "3  EFO:0030003 '10x 3' v3 transcription profiling...              Nan   NaN   \n",
       "4  EFO:0030003 '10x 3' v3 transcription profiling...              Nan   NaN   \n",
       "\n",
       "  pm_PH donor_id  \n",
       "0   NaN      Nan  \n",
       "1   NaN      Nan  \n",
       "2   NaN      Nan  \n",
       "3   NaN      Nan  \n",
       "4   NaN      Nan  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SAMPLE.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "\n",
    "# ## convert \n",
    "# data_path2 = Path.home() / (\"Projects/ASAP/meta-clean/clean/team-Jakobsson\")\n",
    "# metadata_path2 = data_path2 / \"v2_20231114\"\n",
    "\n",
    "# SUBJECT_ = read_meta_table(metadata_path2 / \"SUBJECT.csv\", dtypes_dict)\n",
    "# CLINPATH_ = read_meta_table(metadata_path2 / \"CLINPATH.csv\", dtypes_dict)\n",
    "# STUDY_ = read_meta_table(metadata_path2 / \"STUDY.csv\", dtypes_dict)\n",
    "# PROTOCOL_ = read_meta_table(metadata_path2 / \"PROTOCOL.csv\", dtypes_dict)\n",
    "# SAMPLE_ = read_meta_table(metadata_path2 / \"SAMPLE.csv\", dtypes_dict)\n",
    "# DATA_ = read_meta_table(metadata_path2 / \"DATA.csv\", dtypes_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# # fix SAMPLE replicate, replicate_count, batch with SAMPLE_\n",
    "\n",
    "# batch_mapper = dict(zip(DATA['sample_id'], DATA['batch']))\n",
    "                        \n",
    "# batch = \"batch_\" + SAMPLE['sample_id'].map(batch_mapper)\n",
    "# batch\n",
    "\n",
    "# SAMPLE['batch']  = batch\n",
    "# SAMPLE['replicate'] = 'rep1'\n",
    "# SAMPLE['replicate_count'] = 1\n",
    "# SAMPLE['repeated_sample'] = 0\n",
    "\n",
    "# # also fix \"replicate\" coding in DATA \"rep1\"\n",
    "# DATA['replicate'] = 'rep1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>replicate</th>\n",
       "      <th>replicate_count</th>\n",
       "      <th>repeated_sample</th>\n",
       "      <th>batch</th>\n",
       "      <th>file_type</th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_description</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>technology</th>\n",
       "      <th>omic</th>\n",
       "      <th>adjustment</th>\n",
       "      <th>content</th>\n",
       "      <th>time</th>\n",
       "      <th>header</th>\n",
       "      <th>annotation</th>\n",
       "      <th>configuration_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz</td>\n",
       "      <td>Index 1 Lane 1</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz</td>\n",
       "      <td>Index 2 Lane 1</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz</td>\n",
       "      <td>Read 1 Lane 1</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz</td>\n",
       "      <td>Read 2 Lane 1</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>batch1</td>\n",
       "      <td>fastq</td>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz</td>\n",
       "      <td>Index 1 Lane 2</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>SN</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Raw</td>\n",
       "      <td>Reads</td>\n",
       "      <td>NaN</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     sample_id replicate  replicate_count  repeated_sample  \\\n",
       "0  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "1  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "2  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "3  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "4  PD_BN0009_MTG_snRNAseq_rep1      rep1                1                0   \n",
       "\n",
       "    batch file_type                                         file_name  \\\n",
       "0  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz   \n",
       "1  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz   \n",
       "2  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz   \n",
       "3  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz   \n",
       "4  batch1     fastq  PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz   \n",
       "\n",
       "  file_description                          file_MD5 technology omic  \\\n",
       "0   Index 1 Lane 1  bc037ea92d5359f3da1ba6da4f974e1c         SN  RNA   \n",
       "1   Index 2 Lane 1  0c8ef844b76212e82428af03d341048b         SN  RNA   \n",
       "2    Read 1 Lane 1  c6a41f29849152df2ba2391dc559149b         SN  RNA   \n",
       "3    Read 2 Lane 1  4d610fdf5a72c8206b2f1505d8fcecd7         SN  RNA   \n",
       "4   Index 1 Lane 2  e34917b3ec01cd09ea8b52c82db1dc83         SN  RNA   \n",
       "\n",
       "  adjustment content  time header annotation configuration_file  \n",
       "0        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "1        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "2        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "3        Raw   Reads   NaN   <NA>       <NA>               <NA>  \n",
       "4        Raw   Reads   NaN   <NA>       <NA>               <NA>  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATA.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# checksum = DATA[['file_name','file_MD5']]\n",
    "# checksum['check'] = checksum['file_name'].map(bucket_files_md5)\n",
    "# checksum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from utils.checksums import extract_md5_from_details, extract_md5_from_details2,  extract_crc32c_from_details2, extract_hashes_from_gcloudstorage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_hash = Path.home() / \"Projects/ASAP/team-scherzer/scherzer_hash.log\"\n",
    "\n",
    "source_file_crc = extract_crc32c_from_details2(source_hash)\n",
    "source_file_md5 = extract_md5_from_details2(source_hash)\n",
    "\n",
    "source_hash.exists()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_file_crc,source_file_md5 = extract_hashes_from_gcloudstorage(source_hash)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## crc32c hashes \n",
    "\n",
    "TODO: write hash tools for checking..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket_files_md5 = extract_md5_from_details2(\"scherzer_hexhash.log\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket_files_crc = extract_crc32c_from_details2(\"scherzer_hexhash.log\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_src_crc'] = checksum['file_name'].map(source_file_crc)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_bucket_crc'] = checksum['file_name'].map(bucket_files_crc)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['meta_MD5'] = checksum['file_MD5'].str.strip()\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_src_md5'] = checksum['file_name'].map(source_file_md5)\n",
      "/var/folders/0z/hvx6j8cn6yj3bqbyq6rtsxm40000gn/T/ipykernel_16885/3456434893.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  checksum['check_bucket_md5'] = checksum['file_name'].map(bucket_files_md5)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "checksum = DATA[['file_name','file_MD5']]\n",
    "\n",
    "\n",
    "checksum['check_src_crc'] = checksum['file_name'].map(source_file_crc)\n",
    "checksum['check_bucket_crc'] = checksum['file_name'].map(bucket_files_crc)\n",
    "\n",
    "\n",
    "checksum['meta_MD5'] = checksum['file_MD5'].str.strip()\n",
    "checksum['check_src_md5'] = checksum['file_name'].map(source_file_md5)\n",
    "checksum['check_bucket_md5'] = checksum['file_name'].map(bucket_files_md5)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>file_MD5</th>\n",
       "      <th>check_src_crc</th>\n",
       "      <th>check_bucket_crc</th>\n",
       "      <th>meta_MD5</th>\n",
       "      <th>check_src_md5</th>\n",
       "      <th>check_bucket_md5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>c99915b4</td>\n",
       "      <td>c99915b4</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>bc037ea92d5359f3da1ba6da4f974e1c</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>f0af5e35</td>\n",
       "      <td>f0af5e35</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>0c8ef844b76212e82428af03d341048b</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>151d9d97</td>\n",
       "      <td>151d9d97</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>c6a41f29849152df2ba2391dc559149b</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>58bc244e</td>\n",
       "      <td>58bc244e</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>4d610fdf5a72c8206b2f1505d8fcecd7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>600d0703</td>\n",
       "      <td>600d0703</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>e34917b3ec01cd09ea8b52c82db1dc83</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1547</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L003_R2_001.fastq.gz</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>93e1775e</td>\n",
       "      <td>93e1775e</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>887e3d000587b2774ce3cd7022078577</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_I1_001.fastq.gz</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>a361f850</td>\n",
       "      <td>a361f850</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>44ac0c6d4f7f3b4502eead39564c8723</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_I2_001.fastq.gz</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>b617330f</td>\n",
       "      <td>b617330f</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>dc114df490e1857072462ca090799005</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1550</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>e49ff1eb</td>\n",
       "      <td>e49ff1eb</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>e24eddb1ac0b1f4d2bdd32b5f69dae27</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551</th>\n",
       "      <td>HC_BN1762_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>55d83b85</td>\n",
       "      <td>55d83b85</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>1d316a45079ccb5213fee401535beb02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1552 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             file_name  \\\n",
       "0     PD_BN0009_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz   \n",
       "1     PD_BN0009_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz   \n",
       "2     PD_BN0009_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz   \n",
       "3     PD_BN0009_MTG_snRNAseq_rep1_L001_R2_001.fastq.gz   \n",
       "4     PD_BN0009_MTG_snRNAseq_rep1_L002_I1_001.fastq.gz   \n",
       "...                                                ...   \n",
       "1547  HC_BN1762_MTG_snRNAseq_rep1_L003_R2_001.fastq.gz   \n",
       "1548  HC_BN1762_MTG_snRNAseq_rep1_L004_I1_001.fastq.gz   \n",
       "1549  HC_BN1762_MTG_snRNAseq_rep1_L004_I2_001.fastq.gz   \n",
       "1550  HC_BN1762_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz   \n",
       "1551  HC_BN1762_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz   \n",
       "\n",
       "                              file_MD5 check_src_crc check_bucket_crc  \\\n",
       "0     bc037ea92d5359f3da1ba6da4f974e1c      c99915b4         c99915b4   \n",
       "1     0c8ef844b76212e82428af03d341048b      f0af5e35         f0af5e35   \n",
       "2     c6a41f29849152df2ba2391dc559149b      151d9d97         151d9d97   \n",
       "3     4d610fdf5a72c8206b2f1505d8fcecd7      58bc244e         58bc244e   \n",
       "4     e34917b3ec01cd09ea8b52c82db1dc83      600d0703         600d0703   \n",
       "...                                ...           ...              ...   \n",
       "1547  887e3d000587b2774ce3cd7022078577      93e1775e         93e1775e   \n",
       "1548  44ac0c6d4f7f3b4502eead39564c8723      a361f850         a361f850   \n",
       "1549  dc114df490e1857072462ca090799005      b617330f         b617330f   \n",
       "1550  e24eddb1ac0b1f4d2bdd32b5f69dae27      e49ff1eb         e49ff1eb   \n",
       "1551  1d316a45079ccb5213fee401535beb02      55d83b85         55d83b85   \n",
       "\n",
       "                              meta_MD5                     check_src_md5  \\\n",
       "0     bc037ea92d5359f3da1ba6da4f974e1c  bc037ea92d5359f3da1ba6da4f974e1c   \n",
       "1     0c8ef844b76212e82428af03d341048b  0c8ef844b76212e82428af03d341048b   \n",
       "2     c6a41f29849152df2ba2391dc559149b  c6a41f29849152df2ba2391dc559149b   \n",
       "3     4d610fdf5a72c8206b2f1505d8fcecd7  4d610fdf5a72c8206b2f1505d8fcecd7   \n",
       "4     e34917b3ec01cd09ea8b52c82db1dc83  e34917b3ec01cd09ea8b52c82db1dc83   \n",
       "...                                ...                               ...   \n",
       "1547  887e3d000587b2774ce3cd7022078577  887e3d000587b2774ce3cd7022078577   \n",
       "1548  44ac0c6d4f7f3b4502eead39564c8723  44ac0c6d4f7f3b4502eead39564c8723   \n",
       "1549  dc114df490e1857072462ca090799005  dc114df490e1857072462ca090799005   \n",
       "1550  e24eddb1ac0b1f4d2bdd32b5f69dae27  e24eddb1ac0b1f4d2bdd32b5f69dae27   \n",
       "1551  1d316a45079ccb5213fee401535beb02  1d316a45079ccb5213fee401535beb02   \n",
       "\n",
       "     check_bucket_md5  \n",
       "0                 NaN  \n",
       "1                 NaN  \n",
       "2                 NaN  \n",
       "3                 NaN  \n",
       "4                 NaN  \n",
       "...               ...  \n",
       "1547              NaN  \n",
       "1548              NaN  \n",
       "1549              NaN  \n",
       "1550              NaN  \n",
       "1551              NaN  \n",
       "\n",
       "[1552 rows x 7 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checksum"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "transfer clean metadata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [raw-admin-scherzer@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/scherzer-credentials.json  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                  gs://asap-raw-data-team-scherzer/metadata/Scherzer_raw_meta11202023/\n",
      "                                  gs://asap-raw-data-team-scherzer/metadata/v2/\n",
      "TOTAL: 0 objects, 0 bytes (0B)\n"
     ]
    }
   ],
   "source": [
    "!gcloud storage ls -l \"gs://asap-raw-data-team-scherzer/metadata/\"  --billing-project dnastack-asap-parkinsons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file:///Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer/CLINPATH.csv to gs://asap-raw-data-team-scherzer/metadata/team-Scherzer/CLINPATH.csv\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer/SUBJECT.csv to gs://asap-raw-data-team-scherzer/metadata/team-Scherzer/SUBJECT.csv\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer/SAMPLE.csv to gs://asap-raw-data-team-scherzer/metadata/team-Scherzer/SAMPLE.csv\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer/DATA.csv to gs://asap-raw-data-team-scherzer/metadata/team-Scherzer/DATA.csv\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer/STUDY.csv to gs://asap-raw-data-team-scherzer/metadata/team-Scherzer/STUDY.csv\n",
      "Copying file:///Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer/PROTOCOL.csv to gs://asap-raw-data-team-scherzer/metadata/team-Scherzer/PROTOCOL.csv\n",
      "  Completed files 6/6 | 351.3kiB/351.3kiB                                      \n",
      "\n",
      "Average throughput: 2.8MiB/s\n"
     ]
    }
   ],
   "source": [
    "!gcloud storage cp -r \"/Users/ergonyc/Projects/ASAP/meta-clean/clean/team-Scherzer\" \"gs://asap-raw-data-team-scherzer/metadata\"  --billing-project dnastack-asap-parkinsons "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Scherzer/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Scherzer/SUBJECT.csv [Content-Type=text/csv]...     \n",
      "Copying file://./clean/team-Scherzer/SAMPLE.csv [Content-Type=text/csv]...      \n",
      "Copying file://./clean/team-Scherzer/DATA.csv [Content-Type=text/csv]...        \n",
      "- [4 files][348.6 KiB/348.6 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Scherzer/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Scherzer/PROTOCOL.csv [Content-Type=text/csv]...    \n",
      "\\ [6 files][351.3 KiB/351.3 KiB]                                                \n",
      "Operation completed over 6 objects/351.3 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "!gsutil -u dnastack-asap-parkinsons cp -r \"./clean/team-Scherzer/*.csv\"  \"gs://asap-raw-data-team-scherzer/metadata/v2\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>crc32c_checksum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ILB_BN9947_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz</td>\n",
       "      <td>904383552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ILB_BN9947_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz</td>\n",
       "      <td>3515453253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HC_BN0855_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz</td>\n",
       "      <td>4034171979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HC_BN0855_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz</td>\n",
       "      <td>1461765174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HC_BN0855_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz</td>\n",
       "      <td>2768035529</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            filename  crc32c_checksum\n",
       "0  ILB_BN9947_MTG_snRNAseq_rep1_L004_R1_001.fastq.gz        904383552\n",
       "1  ILB_BN9947_MTG_snRNAseq_rep1_L004_R2_001.fastq.gz       3515453253\n",
       "2   HC_BN0855_MTG_snRNAseq_rep1_L001_I1_001.fastq.gz       4034171979\n",
       "3   HC_BN0855_MTG_snRNAseq_rep1_L001_I2_001.fastq.gz       1461765174\n",
       "4   HC_BN0855_MTG_snRNAseq_rep1_L001_R1_001.fastq.gz       2768035529"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# checksum_path = \"scherzer_crc32c_checksums.csv\"\n",
    "\n",
    "# chksum = pd.read_csv(checksum_path)\n",
    "# chksum.head()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy to workflow-dev bucket\n",
    "\n",
    "First copy each set of metadata locally ..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Activated service account credentials for: [admin-workflow-dev@dnastack-asap-parkinsons.iam.gserviceaccount.com]\n"
     ]
    }
   ],
   "source": [
    "!gcloud auth activate-service-account --key-file=/Users/ergonyc/Projects/ASAP/wf-credentials.json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Copying file://./clean/team-Scherzer/CLINPATH.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Scherzer/SUBJECT.csv [Content-Type=text/csv]...     \n",
      "Copying file://./clean/team-Scherzer/SAMPLE.csv [Content-Type=text/csv]...      \n",
      "Copying file://./clean/team-Scherzer/DATA.csv [Content-Type=text/csv]...        \n",
      "- [4 files][348.6 KiB/348.6 KiB]                                                \n",
      "==> NOTE: You are performing a sequence of gsutil operations that may\n",
      "run significantly faster if you instead use gsutil -m cp ... Please\n",
      "see the -m section under \"gsutil help options\" for further information\n",
      "about when gsutil -m can be advantageous.\n",
      "\n",
      "Copying file://./clean/team-Scherzer/STUDY.csv [Content-Type=text/csv]...\n",
      "Copying file://./clean/team-Scherzer/PROTOCOL.csv [Content-Type=text/csv]...    \n",
      "\\ [6 files][351.3 KiB/351.3 KiB]                                                \n",
      "Operation completed over 6 objects/351.3 KiB.                                    \n"
     ]
    }
   ],
   "source": [
    "!gsutil cp \"./clean/team-Scherzer/*.csv\" \"gs://asap-workflow-dev/metadata/v2/scherzer\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      6243  2023-11-30T21:20:08Z  gs://asap-workflow-dev/metadata/v2/lee/CLINPATH.csv#1701379208316153  metageneration=1\n",
      "     24214  2023-11-30T21:20:08Z  gs://asap-workflow-dev/metadata/v2/lee/DATA.csv#1701379208935791  metageneration=1\n",
      "       968  2023-11-30T21:20:09Z  gs://asap-workflow-dev/metadata/v2/lee/PROTOCOL.csv#1701379209349939  metageneration=1\n",
      "     23092  2023-11-30T21:20:08Z  gs://asap-workflow-dev/metadata/v2/lee/SAMPLE.csv#1701379208725066  metageneration=1\n",
      "      1054  2023-11-30T21:20:09Z  gs://asap-workflow-dev/metadata/v2/lee/STUDY.csv#1701379209140495  metageneration=1\n",
      "      4277  2023-11-30T21:20:08Z  gs://asap-workflow-dev/metadata/v2/lee/SUBJECT.csv#1701379208528070  metageneration=1\n",
      "                                 gs://asap-workflow-dev/metadata/v2/lee/v2_20231128/\n",
      "TOTAL: 6 objects, 59848 bytes (58.45 KiB)\n"
     ]
    }
   ],
   "source": [
    "!gsutil ls -al \"gs://asap-workflow-dev/metadata/v2/lee\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "lw10",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
